<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Notes about SCP data analysis • scp</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><meta property="og:title" content="Notes about SCP data analysis"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="articles/scp.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul></li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Notes about SCP data analysis</h1>
    </div>

<div id="notes-about-scp-data-analysis" class="section level1">

<p>This document summarises thoughts about improving data analysis workflows for mass spectrometry-based single-cell proteomics. The effort here is in response to our recent articles. In the overview of data processing workflows (<a href="http://dx.doi.org/10.1002/cpz1.658" class="external-link">Vanderaa et Gatto 2023</a>), we highlighted the lack of consensus when processing data leading to a “one lab, one workflow” situation. Although this might be the consequence of heterogeneous data structures, we could show this is not the case and is rather due to the lack of a principled approach. This document provides future tracks for <code>scp</code> and beyond to harmonise the current practices.</p>
<div class="section level2">
<h2 id="spectral-data-processing">Spectral data processing<a class="anchor" aria-label="anchor" href="#spectral-data-processing"></a></h2>
<p><a href="http://dx.doi.org/10.1101/2021.08.17.456676" class="external-link">Van der Watt et al. 2021</a> showed it is important to benchmark spectrum identification tools. Furthermore, the same lab later reported that the properties of the spectra from SCP experiment do not fulfill the assumption of identification tools designed on bulk proteomics samples (<a href="http://dx.doi.org/10.1021/acs.jproteome.1c00670" class="external-link">Boekweg et al. 2021</a>).</p>
<p>There is definitely room for improvement, here are some thoughts about SCP spectrum identification:</p>
<ul><li>Is there a rigourous platform to test/benchmark different identification tools or to have a standardised workflow (cf snakemake)? Maybe <a href="https://www.nature.com/articles/s41592-020-0955-0" class="external-link">MassIVE.quant</a>. or Ryan Kelly’s <a href="https://github.com/RTKlab-BYU/Proteomic-Data-Manager" class="external-link">data management software</a>.</li>
<li>
<a href="http://dx.doi.org/10.1101/2021.07.02.450686" class="external-link">IonBot</a> seems a promising tool to improve spectrum identification, especially when including PTMs. It was not tested on SCP data.</li>
<li>
<a href="http://dx.doi.org/10.1038/s41592-019-0426-7" class="external-link">Prosit</a> and <a href="http://dx.doi.org/10.1038/s41592-021-01301-5" class="external-link">deepLC</a> (both python) seem very promising for peptide identifications and validation. These software are based n deep-learning and may adapted/learn the features of SCP if we provide a “good” dataset.</li>
</ul></div>
<div class="section level2">
<h2 id="data-quality-controls">Data quality controls<a class="anchor" aria-label="anchor" href="#data-quality-controls"></a></h2>
<p>Low-quality features (PSMs) and low-quality samples (single-cells) may introduce noise (eg because of low signal due to sample prep issues) or bias (eg because a PSMs is wrongly matched).</p>
<p>It seems sensible to identify and remove low-quality features and samples before they propagate through later processing steps. Data quality control should therefore be the very first step. I find it surprising that many feature or sample quality controls are performed at a later stage. From all the surveyed QC metrics, none require upstream processing and can be applied immediately.</p>
<div class="section level3">
<h3 id="feature-qc">Feature QC<a class="anchor" aria-label="anchor" href="#feature-qc"></a></h3>
<p>The most difficult part of sample preparation is to get rid of the background signal/contamination such as FACS carry-over (cf discussion with Harrison Specht and Andrew Leduc). Negative controls may be used in a way to model systematic background contamination. This modelling should take into account that the background changes with time.</p>
<p>Checking for protein coverage is barely explored in SCP, except in <a href="http://dx.doi.org/10.1038/s41467-022-34919-w" class="external-link">Orsburn et al. 2022</a>.</p>
</div>
<div class="section level3">
<h3 id="sample-qc">Sample QC<a class="anchor" aria-label="anchor" href="#sample-qc"></a></h3>
<p>Identify failed cells and failed runs based on:</p>
<ul><li>Number of detected features, number of proteins with &gt;3 precursors</li>
<li>Intensity distributions. For some experiments it is difficult as the intensity distribution in single cells is similar to blank samples (<a href="http://arxiv.org/abs/2310.13598" class="external-link">Grégoire et al. 2O23</a>).</li>
<li>Coefficient of variation. One should also be mindful of not computing CV in log-scale, or at least use the adapted formula as described by <a href="http://dx.doi.org/10.15406/mojpb.2017.06.00200" class="external-link">Canchola et al. 2017</a>. I compiled some thoughts about CV in <a href="https://github.com/UCLouvain-CBIO/scp/issues/38" class="external-link">this issue</a>:</li>
</ul></div>
</div>
<div class="section level2">
<h2 id="log-transformation">Log-transformation<a class="anchor" aria-label="anchor" href="#log-transformation"></a></h2>
<ul><li>Transform as soon as possible</li>
<li>You cannot use pseudo-counts when log-transforming MS data! MS data are intensities and not counts. Zeros should be NAs anyway.</li>
<li>Note that according to <a href="http://dx.doi.org/10.1101/2023.04.07.535980" class="external-link">Harris et al. 2023</a>, the main objective of log-transformation is variance stabilisation, and they observed that log-transformation may not achieve this goal since the data seems overdispersed.</li>
<li>
<a href="http://dx.doi.org/10.1038/s41592-023-01814-1" class="external-link">Ahlmann-Eltze and Huber 2023</a> offer an interesting discussion and comparison about different variance-stabilising approaches for scRNA-Seq.</li>
</ul></div>
<div class="section level2">
<h2 id="normalisation">Normalisation<a class="anchor" aria-label="anchor" href="#normalisation"></a></h2>
<div class="section level3">
<h3 id="normalisation-with-reference-channels">Normalisation with reference channels<a class="anchor" aria-label="anchor" href="#normalisation-with-reference-channels"></a></h3>
<p>A reference in single-cell is about 5-cell equivalents and is generated by dilution of bulk lysates. This sample is then used to subtract or divide the peptide intensities of the sample in the same batch. Therefore, we should call this batch correction rather than normalisation. I dislike this approach since noise is associated to each reference acquisitions. Olga Vitek suggested including at least 2 channels to better estimate the “normalisation” factor. We also refrained from “normalising” in the linear scale (division by reference), but rather perform it in the log scale (subtraction).</p>
</div>
<div class="section level3">
<h3 id="is-normalisation-really-what-we-want">Is normalisation really what we want?<a class="anchor" aria-label="anchor" href="#is-normalisation-really-what-we-want"></a></h3>
<p>Only applicable for multiplexed experiments: since we perform batch correction for MS acquisition runs and for multiplexing labels (TMT, mTRAQ), is normalisation still required?</p>
<p>I answered this question with scplainer, and yes normalisation is very important and captures cell-specific variations that are not captured by combined modelling of label and MS batch. scplainer currently uses the median peptide abundance as a normalisation variable during modelling. In this case, normalisation captures both technical variation, such as small differences in sample prep, and biological variation, for instance differences in cell size or metabolic activity. Whether the later should be included or excluded from downstream analyses is a conceptual consideration. My intuition is that:</p>
<ul><li>Relative abundances in a cell (concentrations) matter more for proteins such as enzymes, ligand-receptors</li>
<li>Absolute abundances in a cell (copy numbers) matter ore for proteins such as structural proteins</li>
</ul><p>If indeed these two types of abundance matter, the model should be able to decouple cell-specific technical variation from cell-specific biological variation. To do so, we could model the normalisation factor as a function of biological variables such as the cell size. For instance, FACS experiments measure SSC and FSC that provide proxies for granularity and size, respectively. Alternatively, CellenOne measures cell sizes during dispensing. I however noticed in the schoof2021 dataset a poor correlation between FSC/SSC and the median intensity. One explanation is that the factors that drive the need for normalisation are more technical than biological and happen after cell dispensing. <a href="http://dx.doi.org/10.1186/s13059-022-02817-5" class="external-link">Leduc et al. 2022</a> observed good correlation between total protein intensities and cell sizes, as measured by the CellenOne. I could confirm this finding with our modelling approach using minimal data processing. However, the relationship is not perfect, indicating that technical variability is influencing the intensity. The modelling the normalisation factor as a function of cell size would enable an effective decoupling of biological effects and technical effects. The estimated trend would represent biological normalisation while the residuals would represent technical normalisation, each being included in the scplainer modelling approach.</p>
</div>
<div class="section level3">
<h3 id="how-to-define-normalisation-factors">How to define normalisation factors?<a class="anchor" aria-label="anchor" href="#how-to-define-normalisation-factors"></a></h3>
<p>Common practice to estimate sample normalisation factors is to compute median, mean or total intensity for each cell. I would choose median over mean as it is more robust against outliers, but is this the best solution?</p>
<p>Another approach is to align quantiles. While it makes sense for bulk samples, I’m not convinced that different cells should share the same feature intensity distributions, especially in the presence of highly missing data.</p>
<p>Another approach would be to get inspiration from the deconvolution approach implemented in the <code>scran</code> package by <a href="http://dx.doi.org/10.1186/s13059-016-0947-7" class="external-link">Lun et al. 2016</a>, but this would require some efforts to explore the suitability and to adapt the algorithm.</p>
</div>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a></h2>
<div class="section level3">
<h3 id="include-batch-correction-in-the-statistical-modelling">Include batch correction in the statistical modelling<a class="anchor" aria-label="anchor" href="#include-batch-correction-in-the-statistical-modelling"></a></h3>
<p>Replacing value with batch corrected values does not include estimation uncertainty.</p>
</div>
<div class="section level3">
<h3 id="include-the-label-factor-in-the-model">Include the label factor in the model<a class="anchor" aria-label="anchor" href="#include-the-label-factor-in-the-model"></a></h3>
<p>It was clear from the Jupiter Notebooks from <a href="http://dx.doi.org/10.1038/s41467-021-23667-y" class="external-link">Schoof et al. 2021</a> that the TMT label has an impact on quantification. We further confirmed these finding for most multiplexed experiments that the label has an impact on quantifications, but this impact can be modelled.</p>
</div>
<div class="section level3">
<h3 id="batch-correction-for-label-free-data">Batch correction for label-free data<a class="anchor" aria-label="anchor" href="#batch-correction-for-label-free-data"></a></h3>
<ul><li>Ratio compression is often mentioned as a drawback when using TMT labelling. However, LFQ data is not able to model the MS acquisition level bias.</li>
<li>One way to estimate the bias of MS acquisition in LFQ is to model it with respect to acquisition time (eg using as a function of spline basis), or include LC/MS maintenance or calibration as discrete batches.</li>
</ul></div>
</div>
<div class="section level2">
<h2 id="imputation">Imputation<a class="anchor" aria-label="anchor" href="#imputation"></a></h2>
<div class="section level3">
<h3 id="missing-mechanisms">Missing mechanisms<a class="anchor" aria-label="anchor" href="#missing-mechanisms"></a></h3>
<p>SCP data seem not to be MNAR, but I couldn’t find a convincing framework to demonstrate this. Instead, I rely on the relationship between average feature intensity and feature detection rate, as suggested by <a href="http://dx.doi.org/10.1093/bioinformatics/btad200" class="external-link">Li and Smyth 2023</a>. However, their approach does not account for the presence of batch effect and may be biased for SCP.</p>
<p>I think SCP data is therefore mostly MCAR. I find the MCAR definition bears only little information. It says that the data is missing unrelated to the underlying value, but there must be other reasons for values to be missing for which we could model the mechanism. For instance, in DDA, we could model the detection rate of a peptide with respect to the MS1 rank of its precursor in each run. Is a peptide not detected because it was not selected for MS2? For DIA, we could adapt the approach and model the detection rate with the expected spectral complexity determined in MS1. Other variables could be peptide properties. Could we predict peptide ionisation based on its physico-chemical properties?</p>
</div>
<div class="section level3">
<h3 id="is-imputation-really-what-we-want">Is imputation really what we want?<a class="anchor" aria-label="anchor" href="#is-imputation-really-what-we-want"></a></h3>
<p>We discussed this in <a href="http://dx.doi.org/10.1021/acs.jproteome.3c00227" class="external-link">Vanderaa et Gatto 2023</a>.</p>
<p>Some alternative approaches that serves as inspiration:</p>
<ul><li>The RUV-III-C model is a batch correction model where the quantification data is decomposed in a component with known biological factors and component with unknown factors that are considered as batch correction. This model can handle missing data by fitting the model to each peptide separately, ignoring samples with missing values. This model removes the need for imputation, but I don’t think the method will scale to single-cell data and the unknown variation is considered as unwanted whereas unknown biological variation is certainly present in SCP data. Furthermore, this model does not support statistical inference and hence requires to transform the data (cf unaccount uncertainty stated above)</li>
<li>ZINB-Wave is a model used for scRNA-Seq that extends the RUV model. It models the RNA count data using the NB error family (should be replaced by Guassian error for proteomics) as well as the probability of detection. Both component are modeled as gene-wise known factors, sample-wise known factors and gene- and sample-wise unknown factors. The latter could represent wanted biological variation such as undocumented cell types or cell states. The issue is that this method does not scale well (but there is the newwave upgrade), and it does not perform statistical inference.</li>
<li>scVI is also taken from the scRNA-Seq field, but it models the data using variational auto-encoders. It uses the learned latent space to optimise a generative model to predict batch corrected counts and expected dropout, but it can also directly perform Bayesian inference on the parameters. The method scales well but requires that the number of cells &gt;&gt; number of features and it needs to be adapted to proteomics data.</li>
<li>The hurdle model is a two-component model with one focusing on differential abundance (MSqRob model) and the other on differential detection (binomial model). However, further research is needed to speed-up the optimisation (cf Lieven Clement).</li>
<li>Philippe Hauchamps did his master’s thesis about an improvement of proDA to enable statistical inference from peptide data without the need for imputation.</li>
</ul></div>
<div class="section level3">
<h3 id="multiple-imputation-as-a-proxy-to-imputation-uncertainty">Multiple imputation as a proxy to imputation uncertainty<a class="anchor" aria-label="anchor" href="#multiple-imputation-as-a-proxy-to-imputation-uncertainty"></a></h3>
<ul><li>Multiple imputation with different methods and parameters allow to derive distribution statistics (hence uncertainty metrics) for the imputed values that could be used for downstream analyses.</li>
<li>The variance of imputated values could be used as weights for downstream statistical modelling.</li>
</ul></div>
<div class="section level3">
<h3 id="benchmarking-imputation-for-scp">Benchmarking imputation for SCP<a class="anchor" aria-label="anchor" href="#benchmarking-imputation-for-scp"></a></h3>
<ul><li>Get inspiration from <a href="http://dx.doi.org/10.1021/acs.jproteome.0c00123" class="external-link">Bramer et al. 2021</a>, <a href="http://dx.doi.org/10.1021/acs.jproteome.5b00981" class="external-link">Lazar et al. 2016</a>, <a href="http://dx.doi.org/10.1002/pmic.202200092" class="external-link">Kong et al. 2022</a>.</li>
</ul></div>
</div>
<div class="section level2">
<h2 id="aggregation">Aggregation<a class="anchor" aria-label="anchor" href="#aggregation"></a></h2>
<ul><li>Multiple PSMs (~ precursor ions) can match a peptide and multiple peptides can match a protein -&gt; cf recent development by Laurent MsCoreUtils::aggregate_by_matrix()</li>
<li>PSMs should not be seen as independent observations (cf discussion with Thomas Burger)</li>
<li>With MaxQuant, some PSMs match the same sequence, modification and charge but display different retention times (cf Lieven Clement)</li>
</ul><p><code>{r} library(scpdata) specht &lt;- specht2019v3() featAnn &lt;- rowData(specht[[1]]) featAnn$ID &lt;- paste0(featAnn$Set, featAnn$Sequence, featAnn$Charge,                       featAnn$Modifications) dupl &lt;- featAnn$ID[duplicated(featAnn$ID)] featAnn %&gt;%      data.frame %&gt;%      filter(Set == Set[[1]],            ID %in% dupl) %&gt;%      ggplot() +     aes(y = dart_qval,          x = Retention.time,         colour = ID,         group = ID) +     geom_point() +     geom_line() +     theme(legend.position = "none")</code></p>
<div class="section level3">
<h3 id="include-aggregation-in-the-model">Include aggregation in the model<a class="anchor" aria-label="anchor" href="#include-aggregation-in-the-model"></a></h3>
<p>Perform protein-level infers using peptide-level data, as performed by the MsqRob model.</p>
</div>
</div>
<div class="section level2">
<h2 id="modelling-scp-data">Modelling SCP data<a class="anchor" aria-label="anchor" href="#modelling-scp-data"></a></h2>
<div class="section level3">
<h3 id="include-latent-variables">Include latent variables<a class="anchor" aria-label="anchor" href="#include-latent-variables"></a></h3>
<p>For most datasets, linear regression is sufficient and we can retrieve unknown sources of variation from the residuals. This two-step approach could be formalized through the optimisation of latent variables, as performed in RUV-III-C (<a href="http://dx.doi.org/10.1038/s41467-020-17641-3" class="external-link">Poulos et al. 2020</a>) or ZINB-WaVE (<a href="http://dx.doi.org/10.1038/s41467-017-02554-5" class="external-link">Risso et al. 2018</a>).</p>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer></div>






  </body></html>

