<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Minimally processed single-cell proteomics data set — leduc_minimal • scp</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Minimally processed single-cell proteomics data set — leduc_minimal"><meta property="og:description" content="A SingleCellExperiment object that has been minimally processed.
The data set is published by Leduc et al. 2022 (see references)
and retrieved using scpdata::leduc2022_pSCoPE(). The data
processing was conducted with QFeatures and scp. Quality control
was performed, followed by building the peptide data and
log2-transformation. To limit the size of the data, only cells
associated to the 3 first and 3 last MS acquisition runs were
kept. For the same reason, 200 peptides were randomly
sampled. Therefore, the data set consists of 200 peptides and
73 cells. Peptide annotations can be retrieved from the rowData
and cell annotations can be retrieved from the colData."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="../articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="../articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Minimally processed single-cell proteomics data set</h1>
    <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/R/data.R" class="external-link"><code>R/data.R</code></a></small>
    <div class="hidden name"><code>leduc_minimal.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A <code>SingleCellExperiment</code> object that has been minimally processed.
The data set is published by Leduc et al. 2022 (see references)
and retrieved using <code><a href="https://rdrr.io/pkg/scpdata/man/leduc2022_pSCoPE.html" class="external-link">scpdata::leduc2022_pSCoPE()</a></code>. The data
processing was conducted with <code>QFeatures</code> and <code>scp</code>. Quality control
was performed, followed by building the peptide data and
log2-transformation. To limit the size of the data, only cells
associated to the 3 first and 3 last MS acquisition runs were
kept. For the same reason, 200 peptides were randomly
sampled. Therefore, the data set consists of 200 peptides and
73 cells. Peptide annotations can be retrieved from the <code>rowData</code>
and cell annotations can be retrieved from the <code>colData</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"leduc_minimal"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>An object of class <code>SingleCellExperiment</code> with 200 rows and 73 columns.</p>
    </div>
    <div id="quality-control">
    <h2>Quality control</h2>



<p>Any zero value has been replaced by NA.</p>
<p>A peptide was removed from the data set if:</p><ul><li><p>it matched to a decoy or contaminant peptide</p></li>
<li><p>it had an parental ion fraction below 60 \<!-- % --></p></li>
<li><p>it had a DART-ID adjusted q-value superior to 1\<!-- % --></p></li>
<li><p>it had an average sample to carrier ratio above 0.05</p></li>
</ul><p>A cell was removed from the data set if:</p><ul><li><p>it had a median coefficient of variation superior to 0.6</p></li>
<li><p>it had a log2 median intensity outside (6, 8)</p></li>
<li><p>it contained less than 750 peptides</p></li>
</ul></div>
    <div id="building-the-peptide-matrix">
    <h2>Building the peptide matrix</h2>



<p>PSMs belonging to the same peptide were aggregating using the
median value. Some peptides were mapped to a different protein
depending on the MS acquisition run. To solve this issue, a
majority vote was applied to assign a single protein to each
peptide. Protein IDs were translated into gene symbols using the
<code>ensembldb</code> package.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Leduc, Andrew, R. Gray Huffman, Joshua Cantlon, Saad Khan, and
Nikolai Slavov. 2022. “Exploring Functional Protein Covariation
across Single Cells Using nPOP.” Genome Biology 23 (1): 261.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Christophe Vanderaa, Laurent Gatto</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer></div>






  </body></html>

