<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute the median coefficient of variation (CV) per cell — medianCVperCell • scp</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute the median coefficient of variation (CV) per cell — medianCVperCell"><meta property="og:description" content="The function computes for each cell the median CV and stores them
accordingly in the colData of the QFeatures object. The CVs in
each cell are computed from a group of features. The grouping is
defined by a variable in the rowData. The function can be
applied to one or more assays, as long as the samples (column
names) are not duplicated. Also, the user can supply a minimal
number of observations required to compute a CV to avoid that CVs
computed on too few observations influence the distribution within
a cell. The quantification matrix can be optionally normalized
before computing the CVs. Multiple normalizations are possible."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="../articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="../articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute the median coefficient of variation (CV) per cell</h1>
    <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/R/compute_metrics.R" class="external-link"><code>R/compute_metrics.R</code></a></small>
    <div class="hidden name"><code>medianCVperCell.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function computes for each cell the median CV and stores them
accordingly in the <code>colData</code> of the <code>QFeatures</code> object. The CVs in
each cell are computed from a group of features. The grouping is
defined by a variable in the <code>rowData</code>. The function can be
applied to one or more assays, as long as the samples (column
names) are not duplicated. Also, the user can supply a minimal
number of observations required to compute a CV to avoid that CVs
computed on too few observations influence the distribution within
a cell. The quantification matrix can be optionally normalized
before computing the CVs. Multiple normalizations are possible.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">medianCVperCell</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">i</span>,</span>
<span>  <span class="va">groupBy</span>,</span>
<span>  nobs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  na.rm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>,</span>
<span>  norm <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code>QFeatures</code> object</p></dd>


<dt id="arg-i">i<a class="anchor" aria-label="anchor" href="#arg-i"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric()</a></code> or <code><a href="https://rdrr.io/r/base/character.html" class="external-link">character()</a></code> vector indicating from
which assays the <code>rowData</code> should be taken.</p></dd>


<dt id="arg-groupby">groupBy<a class="anchor" aria-label="anchor" href="#arg-groupby"></a></dt>
<dd><p>A <code>character(1)</code> indicating the variable name in
the <code>rowData</code> that contains the feature grouping.</p></dd>


<dt id="arg-nobs">nobs<a class="anchor" aria-label="anchor" href="#arg-nobs"></a></dt>
<dd><p>An <code>integer(1)</code> indicating how many observations
(features) should at least be considered for computing the CV.
Since no CV can be computed for less than 2 observations,
<code>nobs</code> should at least be 2.</p></dd>


<dt id="arg-na-rm">na.rm<a class="anchor" aria-label="anchor" href="#arg-na-rm"></a></dt>
<dd><p>A <code>logical(1)</code> indicating whether missing data should
be removed before computation.</p></dd>


<dt id="arg-coldataname">colDataName<a class="anchor" aria-label="anchor" href="#arg-coldataname"></a></dt>
<dd><p>A <code>character(1)</code> giving the name of the new
variable in the <code>colData</code> where the computed CVs will be
stored. The name cannot already exist in the <code>colData</code>.</p></dd>


<dt id="arg-norm">norm<a class="anchor" aria-label="anchor" href="#arg-norm"></a></dt>
<dd><p>A <code><a href="https://rdrr.io/r/base/character.html" class="external-link">character()</a></code> of normalization methods that will be
sequentially applied to each feature (row) in each assay.
Available methods and additional
information about normalization can be found in
MsCoreUtils::normalizeMethods. You can also specify
<code>norm = "SCoPE2"</code> to reproduce the normalization performed
before computing the CVs as suggested by Specht et al.
<code>norm = "none"</code> will not normalize the data (default)</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments that are passed to the
normalization method.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>QFeatures</code> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>A new column is added to the <code>colData</code> of the object. The samples
(columns) that are not present in the selection <code>i</code> will get
assigned an NA.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Specht, Harrison, Edward Emmott, Aleksandra A. Petelski,
R. Gray Huffman, David H. Perlman, Marco Serra, Peter Kharchenko,
Antonius Koller, and Nikolai Slavov. 2021. “Single-Cell Proteomic
and Transcriptomic Analysis of Macrophage Heterogeneity Using
SCoPE2.” Genome Biology 22 (1): 50.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scp1"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu">filterFeatures</span><span class="op">(</span><span class="va">scp1</span>, <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Proteins</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'Proteins' found in 4 out of 5 assay(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No filter applied to the following assay(s) because one or more filtering variables are missing in the rowData: proteins.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You can control whether to remove or keep the features using the 'keep' argument (see '?filterFeature').</span>
<span class="r-in"><span><span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu">medianCVperCell</span><span class="op">(</span><span class="va">scp1</span>, </span></span>
<span class="r-in"><span>                        i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,</span></span>
<span class="r-in"><span>                        groupBy <span class="op">=</span> <span class="st">"Proteins"</span>,</span></span>
<span class="r-in"><span>                        nobs <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                        na.rm <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                        colDataName <span class="op">=</span> <span class="st">"MedianCV"</span>,</span></span>
<span class="r-in"><span>                        norm <span class="op">=</span> <span class="st">"div.median"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The median CV could not be computed for one or more samples. You may want to try a smaller value for 'nobs'.</span>
<span class="r-in"><span><span class="co">## Check results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">$</span><span class="va">MedianCV</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="medianCVperCell-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer></div>






  </body></html>

