<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced usage of `scp` • scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced usage of `scp`">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="../articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="../articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Advanced usage of <code>scp</code>
</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
                                    <h4 data-toc-skip class="author">Christophe
Vanderaa</h4>
                        
            <h4 data-toc-skip class="date">31 July 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/advanced.Rmd" class="external-link"><code>vignettes/advanced.Rmd</code></a></small>
      <div class="hidden name"><code>advanced.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="about-this-vignette">About this vignette<a class="anchor" aria-label="anchor" href="#about-this-vignette"></a>
</h2>
<p>This vignette is dedicated to advanced users and to method
developers. It assumes that you are already familiar with
<code>QFeatures</code> and <code>scp</code> and that you are looking for
more flexibility in the analysis of your single-cell proteomics (SCP)
data. In fact, <code>scp</code> provides wrapper functions around
generic functions and metrics. However, advanced users may want to apply
or develop their own features. The <code>QFeatures</code> class offers a
flexible data container while guaranteeing data consistency.</p>
<p>In this vignette, you will learn how to:</p>
<ul>
<li>Modify the quantitative data</li>
<li>Modify the sample annotations</li>
<li>Modify the feature annotations</li>
<li>Create a new function for <code>scp</code>
</li>
</ul>
<p>As a general guideline, you can add/remove/update data in a
<code>QFeatures</code> in 4 main steps:</p>
<ol style="list-style-type: decimal">
<li>Gather the data to change and other required data involved in the
processing.</li>
<li>Apply the transformation/computation.</li>
<li>Insert the changes in the <code>QFeatures</code> object.</li>
<li>Make sure the updated <code>QFeatures</code> object is still
valid.</li>
</ol>
<p>To illustrate the different topics, we will load the
<code>scp1</code> example data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scp1"</span><span class="op">)</span></span>
<span><span class="va">scp1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span></span>
<span><span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span></span>
<span><span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span></span>
<span><span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="modify-the-quantitative-data">Modify the quantitative data<a class="anchor" aria-label="anchor" href="#modify-the-quantitative-data"></a>
</h2>
<p>To illustrate how to modify quantitative data, we will implement a
<code>normByType</code> function that will normalize the feature (row)
for each cell type separately. This function is probably not relevant
for a real case analysis, but it provides a good example of a custom
data processing. The process presented in this section is applicable to
any custom function that takes at least a <strong>matrix-like
object</strong> as input and returns a matrix-like object as output.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normByType</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">type</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Check argument</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co">## Normalize for each type separately</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co">## Get normalization factor</span></span>
<span>        <span class="va">nf</span> <span class="op">&lt;-</span> <span class="fu">rowMedians</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>, <span class="va">type</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>        <span class="co">## Perform normalization</span></span>
<span>        <span class="va">x</span><span class="op">[</span>, <span class="va">type</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span>, <span class="va">type</span> <span class="op">==</span> <span class="va">i</span><span class="op">]</span> <span class="op">/</span> <span class="va">nf</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co">## Return normalized data</span></span>
<span>    <span class="va">x</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Suppose we want to apply the function to the <code>proteins</code>
assay, we need to first extract that assay. We here need to transfer the
sample annotations from the <code>QFeatures</code> object to the
extracted <code>SingleCellExperiment</code> in order to get the sample
types required by the <code>normByType</code> function. We therefore use
<code>getWithColData</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">getWithColData</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 292 38 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(2): assay aggcounts</span></span>
<span><span class="co">#&gt; rownames(292): A1A519 A5D8V6 ... REV__CON__Q32PI4 REV__CON__Q3MHN5</span></span>
<span><span class="co">#&gt; rowData names(9): protein Match.time.difference ...</span></span>
<span><span class="co">#&gt;   Potential.contaminant .n</span></span>
<span><span class="co">#&gt; colnames(38): 190321S_LCA10_X_FP97AG_RI1 190321S_LCA10_X_FP97AG_RI2 ...</span></span>
<span><span class="co">#&gt;   190914S_LCB3_X_16plex_Set_21_RI15 190914S_LCB3_X_16plex_Set_21_RI16</span></span>
<span><span class="co">#&gt; colData names(7): Set Channel ... sortday digest</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>Next, we can apply the data transformation to the quantitative data.
As mentioned above, our function expects a matrix-like object as an
input, so we use the <code>assay</code> function. We then update the
<code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mnorm</span> <span class="op">&lt;-</span> <span class="fu">normByType</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, type <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">mnorm</span></span></code></pre></div>
<p>We are now faced with 2 possibilities: either we want to create a new
assay or we want to overwrite an existing assay. In both cases we need
to make sure your data is still valid after data transformation.</p>
<div class="section level3">
<h3 id="create-a-new-assay">Create a new assay<a class="anchor" aria-label="anchor" href="#create-a-new-assay"></a>
</h3>
<p>Creating a new assay has the advantage that you don’t modify an
existing assay and hence limit the risk of introducing inconsistency in
the data and avoid losing intermediate steps of the data processing.</p>
<p>We add the transformed assay using the <code>addAssay</code>
function, then link the parent assay to the transformed assay using
<code>addAssayLinkOneToOne</code>. Note that if each row name in the
parent assay does not match exactly one row in the child assay, you can
also use <code>addAssayLink</code> that will require a linking variable
in the <code>rowData</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">addAssay</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="va">sce</span>, name <span class="op">=</span> <span class="st">"proteinsNorm"</span><span class="op">)</span></span>
<span><span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/AssayLinks.html" class="external-link">addAssayLinkOneToOne</a></span><span class="op">(</span><span class="va">scp1</span>, from <span class="op">=</span> <span class="st">"proteins"</span>, to <span class="op">=</span> <span class="st">"proteinsNorm"</span><span class="op">)</span></span>
<span><span class="va">scp1</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span></span>
<span><span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span></span>
<span><span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span></span>
<span><span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns </span></span>
<span><span class="co">#&gt;  [6] proteinsNorm: SingleCellExperiment with 292 rows and 38 columns</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="overwrite-an-existing-assay">Overwrite an existing assay<a class="anchor" aria-label="anchor" href="#overwrite-an-existing-assay"></a>
</h3>
<p>Overwriting an existing assay has the advantage to limit the memory
consumption as compared to adding a new assay. You can overwrite an
assay simply by replacing the target assay in its corresponding
slot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp1</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">sce</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="check-for-validity">Check for validity<a class="anchor" aria-label="anchor" href="#check-for-validity"></a>
</h3>
<p>Applying custom changes to the data increases the risk for data
inconsistencies. You can however verify that a <code>QFeatures</code>
object is still valid after some processing thanks to the
<code>validObject</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">validObject</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>If the function detects no issues in the data, it will return
<code>TRUE</code>. Otherwise the function will throw an informative
error that should guide the user to identifying the issue.</p>
</div>
</div>
<div class="section level2">
<h2 id="modify-the-sample-annotations">Modify the sample annotations<a class="anchor" aria-label="anchor" href="#modify-the-sample-annotations"></a>
</h2>
<p>To illustrate how to modify the sample annotations, we will compute
the average expression in each sample and include to the
<code>colData</code> of the <code>QFeatures</code> object. This is
typically performed when computing QC metrics for sample filtering. So,
we first extract the required data, in this case the quantitative
values, and compute the sample-wise average protein expression.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span><span class="op">)</span></span>
<span><span class="va">meanExprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">m</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">meanExprs</span></span>
<span><span class="co">#&gt;        190321S_LCA10_X_FP97AG_RI1        190321S_LCA10_X_FP97AG_RI2 </span></span>
<span><span class="co">#&gt;                        45368.3628                         7566.8782 </span></span>
<span><span class="co">#&gt;        190321S_LCA10_X_FP97AG_RI3        190321S_LCA10_X_FP97AG_RI4 </span></span>
<span><span class="co">#&gt;                         3271.0402                         2158.4230 </span></span>
<span><span class="co">#&gt;        190321S_LCA10_X_FP97AG_RI5        190321S_LCA10_X_FP97AG_RI6 </span></span>
<span><span class="co">#&gt;                         1162.3936                         1835.7939 </span></span>
<span><span class="co">#&gt;        190321S_LCA10_X_FP97AG_RI7        190321S_LCA10_X_FP97AG_RI8 </span></span>
<span><span class="co">#&gt;                         1293.4738                         1415.4205 </span></span>
<span><span class="co">#&gt;        190321S_LCA10_X_FP97AG_RI9       190321S_LCA10_X_FP97AG_RI10 </span></span>
<span><span class="co">#&gt;                         1965.6531                         1264.4064 </span></span>
<span><span class="co">#&gt;       190321S_LCA10_X_FP97AG_RI11         190222S_LCA9_X_FP94BM_RI1 </span></span>
<span><span class="co">#&gt;                         1238.5979                        48901.4654 </span></span>
<span><span class="co">#&gt;         190222S_LCA9_X_FP94BM_RI2         190222S_LCA9_X_FP94BM_RI3 </span></span>
<span><span class="co">#&gt;                         5189.5912                         2802.1850 </span></span>
<span><span class="co">#&gt;         190222S_LCA9_X_FP94BM_RI4         190222S_LCA9_X_FP94BM_RI5 </span></span>
<span><span class="co">#&gt;                         1204.2350                          715.7667 </span></span>
<span><span class="co">#&gt;         190222S_LCA9_X_FP94BM_RI6         190222S_LCA9_X_FP94BM_RI7 </span></span>
<span><span class="co">#&gt;                         1032.7122                          403.8267 </span></span>
<span><span class="co">#&gt;         190222S_LCA9_X_FP94BM_RI8         190222S_LCA9_X_FP94BM_RI9 </span></span>
<span><span class="co">#&gt;                         1089.0548                         1076.2613 </span></span>
<span><span class="co">#&gt;        190222S_LCA9_X_FP94BM_RI10        190222S_LCA9_X_FP94BM_RI11 </span></span>
<span><span class="co">#&gt;                         1572.3337                         1186.4431 </span></span>
<span><span class="co">#&gt;  190914S_LCB3_X_16plex_Set_21_RI1  190914S_LCB3_X_16plex_Set_21_RI2 </span></span>
<span><span class="co">#&gt;                        96881.8350                         2731.9430 </span></span>
<span><span class="co">#&gt;  190914S_LCB3_X_16plex_Set_21_RI3  190914S_LCB3_X_16plex_Set_21_RI4 </span></span>
<span><span class="co">#&gt;                         8208.1958                          208.2857 </span></span>
<span><span class="co">#&gt;  190914S_LCB3_X_16plex_Set_21_RI5  190914S_LCB3_X_16plex_Set_21_RI6 </span></span>
<span><span class="co">#&gt;                         3209.7730                         2014.4889 </span></span>
<span><span class="co">#&gt;  190914S_LCB3_X_16plex_Set_21_RI7  190914S_LCB3_X_16plex_Set_21_RI8 </span></span>
<span><span class="co">#&gt;                         2897.1103                         3006.6741 </span></span>
<span><span class="co">#&gt;  190914S_LCB3_X_16plex_Set_21_RI9 190914S_LCB3_X_16plex_Set_21_RI10 </span></span>
<span><span class="co">#&gt;                         2760.8564                         2549.9348 </span></span>
<span><span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI11 190914S_LCB3_X_16plex_Set_21_RI12 </span></span>
<span><span class="co">#&gt;                         2388.5289                         2533.8006 </span></span>
<span><span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI13 190914S_LCB3_X_16plex_Set_21_RI14 </span></span>
<span><span class="co">#&gt;                         2134.5615                         3125.1980 </span></span>
<span><span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI15 190914S_LCB3_X_16plex_Set_21_RI16 </span></span>
<span><span class="co">#&gt;                         3276.5618                         3195.8827</span></span></code></pre></div>
<p>Next, we insert the computed averages into the <code>colData</code>.
You need to make sure to match sample names because an extracted assay
may not contain all samples and may be in a different order compared to
the <code>colData</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">meanExprs</span><span class="op">)</span>, <span class="st">"meanProtExprs"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">meanExprs</span></span></code></pre></div>
<p>The new sample variable <code>meanProtExprs</code> is now accessible
for filtering or plotting. The <code>$</code> operator makes it
straightforward to access the new data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">$</span><span class="va">meanProtExprs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="advanced_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>To make sure that the process did not corrupt the
<code>colData</code>, we advise to verify the data is still valid.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">validObject</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="modify-the-feature-annotations">Modify the feature annotations<a class="anchor" aria-label="anchor" href="#modify-the-feature-annotations"></a>
</h2>
<p>We will again illustrate how to modify the feature annotations with
an example. We here demonstrate how to add the number of samples in
which each feature is detected for the three first assays (PSM assays).
The challenge here is that the metric needs to be computed for each
assay separately and inserted in the corresponding assay.</p>
<p>We will take advantage of the replacement function for
<code>rowData</code> as implemented in <code>QFeatures</code>. It
expects a list-like object where names indicate in which assays we want
to modify the <code>rowData</code> and each element contains a table
with the replacement values.</p>
<p>We therefore compute the metrics for each assay separately and store
the results in a named <code>List</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Initialize the List object that will store the computed values</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">List</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## We compute the metric for the first 3 assays</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## We get the quantitative values for the current assay</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="co">## We compute the number of samples in which each features is detected</span></span>
<span>    <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">m</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="co">## We store the result as a DataFrame in the List</span></span>
<span>    <span class="va">res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>nbSamples <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; List of length 3</span></span>
<span><span class="co">#&gt; names(3): 190321S_LCA10_X_FP97AG 190222S_LCA9_X_FP94BM 190914S_LCB3_X_16plex_Set_21</span></span>
<span><span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 166 rows and 1 column</span></span>
<span><span class="co">#&gt;           nbSamples</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; PSM3773          11</span></span>
<span><span class="co">#&gt; PSM9078          11</span></span>
<span><span class="co">#&gt; PSM9858          11</span></span>
<span><span class="co">#&gt; PSM11744         11</span></span>
<span><span class="co">#&gt; PSM21752          0</span></span>
<span><span class="co">#&gt; ...             ...</span></span>
<span><span class="co">#&gt; PSM732069        11</span></span>
<span><span class="co">#&gt; PSM735396        11</span></span>
<span><span class="co">#&gt; PSM744756        10</span></span>
<span><span class="co">#&gt; PSM745037        11</span></span>
<span><span class="co">#&gt; PSM745130        11</span></span></code></pre></div>
<p>Now that we have a <code>List</code> of <code>DataFrame</code>s, we
can update the object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span></span></code></pre></div>
<p>The new feature variable <code>nbSamples</code> is now accessible for
filtering or plotting. The <code>rbindRowData</code> function
facilitates the access the new data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nbSamples</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">16</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">assay</span><span class="op">)</span></span></code></pre></div>
<p><img src="advanced_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>To make sure that the process did not corrupt the
<code>rowData</code> in any assay, we advise to verify the data is still
valid.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">validObject</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-new-function-for-scp">Create a new function for <code>scp</code><a class="anchor" aria-label="anchor" href="#create-a-new-function-for-scp"></a>
</h2>
<p>The modifying data in a <code>QFeatures</code> involves a
multiple-step process. Creating a wrapper function that would take care
of those different steps in a single line of code is a good habit to
reduce the length of analysis scripts and hence making it easier to
understand and less error-prone.</p>
<p>We will wrap the last example in a new function that we call
<code>computeNbDetectedSamples</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">computeNbDetectedSamples</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">object</span>, <span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">List</span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">ii</span> <span class="kw">in</span> <span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">object</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">m</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>        <span class="va">res</span><span class="op">[[</span><span class="va">ii</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>nbSamples <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">res</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu">validObject</span><span class="op">(</span><span class="va">object</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">object</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Thanks to this new function, the previous section now simply boils
down to running:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu">computeNbDetectedSamples</span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>Keep in mind a few recommendations when creating a new function for
<code>scp</code>:</p>
<ul>
<li>The function should take a <code>QFeatures</code> object as input
(and more arguments if needed) and return a <code>QFeatures</code>
object as output. This will make workflows much easier to
understand.</li>
<li>Allow user to select assays (if required) either as numeric,
character, or logical.</li>
<li>Use conventional argument names: when naming an argument, try to
match the names that already exist. For instance, selecting assays is
passed through the <code>i</code> argument, selecting
<code>rowData</code> variables is passed through <code>rowvars</code>
and selecting <code>colData</code> variables is passed through
<code>colvars</code>.</li>
<li>Follow the <code>rformassspectrometry</code> <a href="https://rformassspectrometry.github.io/RforMassSpectrometry/articles/RforMassSpectrometry.html#coding-style" class="external-link">coding
style</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="whats-next">What’s next?<a class="anchor" aria-label="anchor" href="#whats-next"></a>
</h2>
<p>So you developed a new metric or method and believe it might benefit
the community? We would love to hear about your improvements and
eventually include your new functionality into <code>scp</code> or
associate your new package to our documentation. Please, raise an <a href="https://github.com/UCLouvain-CBIO/scp/issues/new/choose" class="external-link">issue</a>
in our Github repository to suggest your improvements or, better, submit
your code as a <a href="https://github.com/UCLouvain-CBIO/scp/compare" class="external-link">pull
request</a>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggplot2_3.5.1               scp_1.15.1                 
 [3] QFeatures_1.15.2            MultiAssayExperiment_1.31.4
 [5] SummarizedExperiment_1.35.1 Biobase_2.65.0             
 [7] GenomicRanges_1.57.1        GenomeInfoDb_1.41.1        
 [9] IRanges_2.39.2              S4Vectors_0.43.2           
[11] BiocGenerics_0.51.0         MatrixGenerics_1.17.0      
[13] matrixStats_1.3.0           BiocStyle_2.33.1           

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1            farver_2.1.2               
 [3] dplyr_1.1.4                 bitops_1.0-8               
 [5] fastmap_1.2.0               SingleCellExperiment_1.27.2
 [7] lazyeval_0.2.2              RCurl_1.98-1.16            
 [9] nipals_0.8                  digest_0.6.36              
[11] lifecycle_1.0.4             cluster_2.1.6              
[13] ProtGenerics_1.37.0         magrittr_2.0.3             
[15] compiler_4.4.1              rlang_1.1.4                
[17] sass_0.4.9                  tools_4.4.1                
[19] igraph_2.0.3                utf8_1.2.4                 
[21] yaml_2.3.10                 knitr_1.48                 
[23] labeling_0.4.3              S4Arrays_1.5.5             
[25] htmlwidgets_1.6.4           DelayedArray_0.31.10       
[27] plyr_1.8.9                  RColorBrewer_1.1-3         
[29] abind_1.4-5                 withr_3.0.0                
[31] purrr_1.0.2                 desc_1.4.3                 
[33] grid_4.4.1                  fansi_1.0.6                
[35] colorspace_2.1-1            scales_1.3.0               
[37] MASS_7.3-61                 cli_3.6.3                  
[39] rmarkdown_2.27              crayon_1.5.3               
[41] ragg_1.3.2                  generics_0.1.3             
[43] metapod_1.13.0              httr_1.4.7                 
[45] reshape2_1.4.4              BiocBaseUtils_1.7.0        
[47] cachem_1.1.0                stringr_1.5.1              
[49] zlibbioc_1.51.1             AnnotationFilter_1.29.0    
[51] BiocManager_1.30.23         XVector_0.45.0             
[53] vctrs_0.6.5                 Matrix_1.7-0               
[55] slam_0.1-51                 jsonlite_1.8.8             
[57] bookdown_0.40               IHW_1.33.0                 
[59] ggrepel_0.9.5               clue_0.3-65                
[61] systemfonts_1.1.0           jquerylib_0.1.4            
[63] tidyr_1.3.1                 glue_1.7.0                 
[65] pkgdown_2.1.0.9000          stringi_1.8.4              
[67] gtable_0.3.5                UCSC.utils_1.1.0           
[69] munsell_0.5.1               lpsymphony_1.33.0          
[71] tibble_3.2.1                pillar_1.9.0               
[73] htmltools_0.5.8.1           GenomeInfoDbData_1.2.12    
[75] R6_2.5.1                    textshaping_0.4.0          
[77] evaluate_0.24.0             lattice_0.22-6             
[79] highr_0.11                  bslib_0.8.0                
[81] fdrtool_1.2.17              Rcpp_1.0.13                
[83] SparseArray_1.5.27          xfun_0.46                  
[85] MsCoreUtils_1.17.0          fs_1.6.4                   
[87] pkgconfig_2.0.3            </code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
