<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>QFeatures in a nutshell â€¢ scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="QFeatures in a nutshell">
<meta property="og:description" content="scp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="QFeatures_nutshell_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>QFeatures in a nutshell</h1>
                        <h4 data-toc-skip class="author">Laurent Gatto</h4>
                                    <h4 data-toc-skip class="author">Christophe Vanderaa</h4>
                        
            <h4 data-toc-skip class="date">20 July 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/QFeatures_nutshell.Rmd" class="external-link"><code>vignettes/QFeatures_nutshell.Rmd</code></a></small>
      <div class="hidden name"><code>QFeatures_nutshell.Rmd</code></div>

    </div>

    
    
<p>This vignette briefly recaps the main concepts of <code>QFeatures</code> on which <code>scp</code> relies. More in depth information is to be found in the <code>QFeatures</code> <a href="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures.html" class="external-link">vignettes</a>.</p>
<div id="the-qfeatures-class" class="section level1">
<h1 class="hasAnchor">
<a href="#the-qfeatures-class" class="anchor" aria-hidden="true"></a>The <code>QFeatures</code> class</h1>
<p>The <code>QFeatures</code> class is based on the <code>MultiAssayExperiment</code> class that holds a collection of <code>SummarizedExperiment</code> (or other classes that inherits from it) objects termed <strong>assays</strong>. The assays in a <code>QFeatures</code> object have a hierarchical relation: proteins are composed of peptides, themselves produced by spectra, as depicted in figure below.</p>
<p><img src="https://rformassspectrometry.github.io/QFeatures/articles/QFeatures_files/figure-html/featuresplot-1.png"></p>
<p>Those links are stored as part as the <code>QFeatures</code> object and connect the assays together. We load an example dataset from the <code>scp</code> package that is formatted as an <code>QFeatures</code> object and plot those connection.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"scp1"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span></code></pre></div>
<p><img src="QFeatures_nutshell_files/figure-html/hierarchy-1.png" width="600px"></p>
</div>
<div id="accessing-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#accessing-the-data" class="anchor" aria-hidden="true"></a>Accessing the data</h1>
<p>The <code>QFeatures</code> class contains all the available and metadata. We here show how to retrieve those different pieces of information.</p>
<div id="quantitative-data" class="section level2">
<h2 class="hasAnchor">
<a href="#quantitative-data" class="anchor" aria-hidden="true"></a>Quantitative data</h2>
<p>The quantitative data, stored as matrix-like objects, can be accessed using the <code>assay</code> function. For example, we here extract the quantitative data for the first MS batch (and show a subset of it):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"190321S_LCA10_X_FP97AG"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI1 190321S_LCA10_X_FP97AG_RI2</span>
<span class="co">#&gt; PSM3773                       57895                     603.73</span>
<span class="co">#&gt; PSM9078                       64889                    1481.30</span>
<span class="co">#&gt; PSM9858                       58993                     489.85</span>
<span class="co">#&gt; PSM11744                      75711                     539.02</span>
<span class="co">#&gt; PSM21752                          0                       0.00</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI3 190321S_LCA10_X_FP97AG_RI4</span>
<span class="co">#&gt; PSM3773                      2787.9                     757.17</span>
<span class="co">#&gt; PSM9078                      4891.6                     597.53</span>
<span class="co">#&gt; PSM9858                      2899.4                     882.37</span>
<span class="co">#&gt; PSM11744                     7292.7                     357.90</span>
<span class="co">#&gt; PSM21752                        0.0                       0.00</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI5 190321S_LCA10_X_FP97AG_RI6</span>
<span class="co">#&gt; PSM3773                      862.08                    1118.80</span>
<span class="co">#&gt; PSM9078                     1140.30                    1300.10</span>
<span class="co">#&gt; PSM9858                      296.60                     977.15</span>
<span class="co">#&gt; PSM11744                    1091.30                     736.87</span>
<span class="co">#&gt; PSM21752                       0.00                       0.00</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI7 190321S_LCA10_X_FP97AG_RI8</span>
<span class="co">#&gt; PSM3773                      640.10                    1446.10</span>
<span class="co">#&gt; PSM9078                     1092.50                    1309.40</span>
<span class="co">#&gt; PSM9858                      498.60                    1437.90</span>
<span class="co">#&gt; PSM11744                     712.74                     590.75</span>
<span class="co">#&gt; PSM21752                       0.00                       0.00</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI9 190321S_LCA10_X_FP97AG_RI10</span>
<span class="co">#&gt; PSM3773                      968.49                      648.56</span>
<span class="co">#&gt; PSM9078                     1538.40                     1014.50</span>
<span class="co">#&gt; PSM9858                      857.40                      888.01</span>
<span class="co">#&gt; PSM11744                   15623.00                      298.60</span>
<span class="co">#&gt; PSM21752                       0.00                        0.00</span>
<span class="co">#&gt;          190321S_LCA10_X_FP97AG_RI11</span>
<span class="co">#&gt; PSM3773                       742.53</span>
<span class="co">#&gt; PSM9078                      1062.80</span>
<span class="co">#&gt; PSM9858                       768.61</span>
<span class="co">#&gt; PSM11744                      481.38</span>
<span class="co">#&gt; PSM21752                        0.00</span></code></pre></div>
<p>Note that you can retrieve the list of available assays in a <code>QFeatures</code> object using the <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span>
<span class="co">#&gt; [1] "190321S_LCA10_X_FP97AG"       "190222S_LCA9_X_FP94BM"       </span>
<span class="co">#&gt; [3] "190914S_LCB3_X_16plex_Set_21" "peptides"                    </span>
<span class="co">#&gt; [5] "proteins"</span></code></pre></div>
</div>
<div id="feature-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-metadata" class="anchor" aria-hidden="true"></a>Feature metadata</h2>
<p>For each individual assay, there is feature metadata available. We extract the list of metadata tables by using <code>rowData()</code> on the <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">rowData</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span>
<span class="co">#&gt; DataFrameList of length 5</span>
<span class="co">#&gt; names(5): 190321S_LCA10_X_FP97AG 190222S_LCA9_X_FP94BM 190914S_LCB3_X_16plex_Set_21 peptides proteins</span>
<span class="fu">rowData</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span><span class="op">[[</span><span class="st">"proteins"</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; DataFrame with 292 rows and 9 columns</span>
<span class="co">#&gt;                                            protein Match.time.difference</span>
<span class="co">#&gt;                                        &lt;character&gt;             &lt;logical&gt;</span>
<span class="co">#&gt; A1A519                                      A1A519                    NA</span>
<span class="co">#&gt; A5D8V6                                      A5D8V6                    NA</span>
<span class="co">#&gt; A5PLK6                                      A5PLK6                    NA</span>
<span class="co">#&gt; A5PLL1                                      A5PLL1                    NA</span>
<span class="co">#&gt; A6NC97                                      A6NC97                    NA</span>
<span class="co">#&gt; ...                                            ...                   ...</span>
<span class="co">#&gt; REV__CON__ENSEMBL:ENSBTAP00000038253 REV__CON__...                    NA</span>
<span class="co">#&gt; REV__CON__P06868                     REV__CON__...                    NA</span>
<span class="co">#&gt; REV__CON__Q05443                     REV__CON__...                    NA</span>
<span class="co">#&gt; REV__CON__Q32PI4                     REV__CON__...                    NA</span>
<span class="co">#&gt; REV__CON__Q3MHN5                     REV__CON__...                    NA</span>
<span class="co">#&gt;                                      Match.m.z.difference Match.q.value</span>
<span class="co">#&gt;                                                 &lt;logical&gt;     &lt;logical&gt;</span>
<span class="co">#&gt; A1A519                                                 NA            NA</span>
<span class="co">#&gt; A5D8V6                                                 NA            NA</span>
<span class="co">#&gt; A5PLK6                                                 NA            NA</span>
<span class="co">#&gt; A5PLL1                                                 NA            NA</span>
<span class="co">#&gt; A6NC97                                                 NA            NA</span>
<span class="co">#&gt; ...                                                   ...           ...</span>
<span class="co">#&gt; REV__CON__ENSEMBL:ENSBTAP00000038253                   NA            NA</span>
<span class="co">#&gt; REV__CON__P06868                                       NA            NA</span>
<span class="co">#&gt; REV__CON__Q05443                                       NA            NA</span>
<span class="co">#&gt; REV__CON__Q32PI4                                       NA            NA</span>
<span class="co">#&gt; REV__CON__Q3MHN5                                       NA            NA</span>
<span class="co">#&gt;                                      Match.score Reporter.PIF Reporter.fraction</span>
<span class="co">#&gt;                                        &lt;logical&gt;    &lt;logical&gt;         &lt;logical&gt;</span>
<span class="co">#&gt; A1A519                                        NA           NA                NA</span>
<span class="co">#&gt; A5D8V6                                        NA           NA                NA</span>
<span class="co">#&gt; A5PLK6                                        NA           NA                NA</span>
<span class="co">#&gt; A5PLL1                                        NA           NA                NA</span>
<span class="co">#&gt; A6NC97                                        NA           NA                NA</span>
<span class="co">#&gt; ...                                          ...          ...               ...</span>
<span class="co">#&gt; REV__CON__ENSEMBL:ENSBTAP00000038253          NA           NA                NA</span>
<span class="co">#&gt; REV__CON__P06868                              NA           NA                NA</span>
<span class="co">#&gt; REV__CON__Q05443                              NA           NA                NA</span>
<span class="co">#&gt; REV__CON__Q32PI4                              NA           NA                NA</span>
<span class="co">#&gt; REV__CON__Q3MHN5                              NA           NA                NA</span>
<span class="co">#&gt;                                      Potential.contaminant        .n</span>
<span class="co">#&gt;                                                &lt;character&gt; &lt;integer&gt;</span>
<span class="co">#&gt; A1A519                                                             1</span>
<span class="co">#&gt; A5D8V6                                                             1</span>
<span class="co">#&gt; A5PLK6                                                             1</span>
<span class="co">#&gt; A5PLL1                                                             1</span>
<span class="co">#&gt; A6NC97                                                             1</span>
<span class="co">#&gt; ...                                                    ...       ...</span>
<span class="co">#&gt; REV__CON__ENSEMBL:ENSBTAP00000038253                     +         1</span>
<span class="co">#&gt; REV__CON__P06868                                         +         1</span>
<span class="co">#&gt; REV__CON__Q05443                                         +         1</span>
<span class="co">#&gt; REV__CON__Q32PI4                                         +         1</span>
<span class="co">#&gt; REV__CON__Q3MHN5                                         +         1</span></code></pre></div>
<p>You can also retrieve the names of each <code>rowData</code> column for all assays with <code>rowDataNames</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rowDataNames</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span>
<span class="co">#&gt; CharacterList of length 5</span>
<span class="co">#&gt; [["190321S_LCA10_X_FP97AG"]] uid Sequence ... peptide Leading.razor.protein</span>
<span class="co">#&gt; [["190222S_LCA9_X_FP94BM"]] uid Sequence ... peptide Leading.razor.protein</span>
<span class="co">#&gt; [["190914S_LCB3_X_16plex_Set_21"]] uid Sequence ... Leading.razor.protein</span>
<span class="co">#&gt; [["peptides"]] Sequence Length Modifications ... .n Leading.razor.protein</span>
<span class="co">#&gt; [["proteins"]] protein Match.time.difference ... Potential.contaminant .n</span></code></pre></div>
<p>You can also get the <code>rowData</code> from different assays in a single table using the <code>rbindRowData</code> function. It will keep the common <code>rowData</code> variables to all selected assays (provided through <code>i</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 1388 rows and 10 columns</span>
<span class="co">#&gt;              assay       rowname       protein Match.time.difference</span>
<span class="co">#&gt;        &lt;character&gt;   &lt;character&gt;   &lt;character&gt;             &lt;logical&gt;</span>
<span class="co">#&gt; 1    190321S_LC...       PSM3773        P61981                    NA</span>
<span class="co">#&gt; 2    190321S_LC...       PSM9078        Q8WVN8                    NA</span>
<span class="co">#&gt; 3    190321S_LC...       PSM9858        P55084                    NA</span>
<span class="co">#&gt; 4    190321S_LC...      PSM11744        P19099                    NA</span>
<span class="co">#&gt; 5    190321S_LC...      PSM21752        P52952                    NA</span>
<span class="co">#&gt; ...            ...           ...           ...                   ...</span>
<span class="co">#&gt; 1384      proteins REV__CON__... REV__CON__...                    NA</span>
<span class="co">#&gt; 1385      proteins REV__CON__... REV__CON__...                    NA</span>
<span class="co">#&gt; 1386      proteins REV__CON__... REV__CON__...                    NA</span>
<span class="co">#&gt; 1387      proteins REV__CON__... REV__CON__...                    NA</span>
<span class="co">#&gt; 1388      proteins REV__CON__... REV__CON__...                    NA</span>
<span class="co">#&gt;      Match.m.z.difference Match.q.value Match.score Reporter.PIF</span>
<span class="co">#&gt;                 &lt;logical&gt;     &lt;logical&gt;   &lt;logical&gt;    &lt;logical&gt;</span>
<span class="co">#&gt; 1                      NA            NA          NA           NA</span>
<span class="co">#&gt; 2                      NA            NA          NA           NA</span>
<span class="co">#&gt; 3                      NA            NA          NA           NA</span>
<span class="co">#&gt; 4                      NA            NA          NA           NA</span>
<span class="co">#&gt; 5                      NA            NA          NA           NA</span>
<span class="co">#&gt; ...                   ...           ...         ...          ...</span>
<span class="co">#&gt; 1384                   NA            NA          NA           NA</span>
<span class="co">#&gt; 1385                   NA            NA          NA           NA</span>
<span class="co">#&gt; 1386                   NA            NA          NA           NA</span>
<span class="co">#&gt; 1387                   NA            NA          NA           NA</span>
<span class="co">#&gt; 1388                   NA            NA          NA           NA</span>
<span class="co">#&gt;      Reporter.fraction Potential.contaminant</span>
<span class="co">#&gt;              &lt;logical&gt;           &lt;character&gt;</span>
<span class="co">#&gt; 1                   NA                      </span>
<span class="co">#&gt; 2                   NA                      </span>
<span class="co">#&gt; 3                   NA                      </span>
<span class="co">#&gt; 4                   NA                      </span>
<span class="co">#&gt; 5                   NA                      </span>
<span class="co">#&gt; ...                ...                   ...</span>
<span class="co">#&gt; 1384                NA                     +</span>
<span class="co">#&gt; 1385                NA                     +</span>
<span class="co">#&gt; 1386                NA                     +</span>
<span class="co">#&gt; 1387                NA                     +</span>
<span class="co">#&gt; 1388                NA                     +</span></code></pre></div>
</div>
<div id="sample-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-metadata" class="anchor" aria-hidden="true"></a>Sample metadata</h2>
<p>The sample metadata is retrieved using <code>colData</code> on the <code>QFeatures</code> object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">scp1</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 38 rows and 7 columns</span>
<span class="co">#&gt;                                             Set     Channel SampleAnnotation</span>
<span class="co">#&gt;                                     &lt;character&gt; &lt;character&gt;      &lt;character&gt;</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI1         190222S_LC...         RI1    carrier_mi...</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI2         190222S_LC...         RI2             norm</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI3         190222S_LC...         RI3           unused</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI4         190222S_LC...         RI4             sc_u</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI5         190222S_LC...         RI5             sc_0</span>
<span class="co">#&gt; ...                                         ...         ...              ...</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI12 190914S_LC...        RI12            sc_m0</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI13 190914S_LC...        RI13            sc_m0</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI14 190914S_LC...        RI14            sc_m0</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI15 190914S_LC...        RI15            sc_m0</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI16 190914S_LC...        RI16            sc_m0</span>
<span class="co">#&gt;                                    SampleType     lcbatch     sortday</span>
<span class="co">#&gt;                                   &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI1             Carrier        LCA9          s8</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI2           Reference        LCA9          s8</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI3              Unused        LCA9          s8</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI4            Monocyte        LCA9          s8</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI5               Blank        LCA9          s8</span>
<span class="co">#&gt; ...                                       ...         ...         ...</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI12  Macrophage        LCB3          s9</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI13  Macrophage        LCB3          s9</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI14  Macrophage        LCB3          s9</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI15  Macrophage        LCB3          s9</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI16  Macrophage        LCB3          s9</span>
<span class="co">#&gt;                                        digest</span>
<span class="co">#&gt;                                   &lt;character&gt;</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI1                   N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI2                   N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI3                   N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI4                   N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_RI5                   N</span>
<span class="co">#&gt; ...                                       ...</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI12           R</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI13           R</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI14           R</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI15           R</span>
<span class="co">#&gt; 190914S_LCB3_X_16plex_Set_21_RI16           R</span></code></pre></div>
<p>Note that you can easily access a <code>colData</code> column using the <code>$</code> operator. See here how we extract the sample types from the <code>colData</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp1</span><span class="op">$</span><span class="va">SampleType</span>
<span class="co">#&gt;  [1] "Carrier"    "Reference"  "Unused"     "Monocyte"   "Blank"     </span>
<span class="co">#&gt;  [6] "Monocyte"   "Macrophage" "Macrophage" "Macrophage" "Macrophage"</span>
<span class="co">#&gt; [11] "Macrophage" "Carrier"    "Reference"  "Unused"     "Macrophage"</span>
<span class="co">#&gt; [16] "Monocyte"   "Macrophage" "Macrophage" "Macrophage" "Macrophage"</span>
<span class="co">#&gt; [21] "Macrophage" "Macrophage" "Carrier"    "Reference"  "Unused"    </span>
<span class="co">#&gt; [26] "Unused"     "Macrophage" "Macrophage" "Blank"      "Monocyte"  </span>
<span class="co">#&gt; [31] "Macrophage" "Monocyte"   "Blank"      "Macrophage" "Macrophage"</span>
<span class="co">#&gt; [36] "Macrophage" "Macrophage" "Macrophage"</span></code></pre></div>
</div>
</div>
<div id="subsetting-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#subsetting-the-data" class="anchor" aria-hidden="true"></a>Subsetting the data</h1>
<p>There are three dimensions we want to subset for:</p>
<ul>
<li>Assays</li>
<li>Samples</li>
<li>Features</li>
</ul>
<p>Therefore, <code>QFeatures</code> support a three-index subsetting. This is performed through the simple bracket method <code>[feature, sample, assay]</code>.</p>
<div id="subset-assays" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-assays" class="anchor" aria-hidden="true"></a>Subset assays</h2>
<p>Suppose that we want to focus only on the first MS batch (<code>190321S_LCA10_X_FP97AG</code>) for separate processing of the data. Subsetting the <code>QFeatures</code> object for that assay is simply:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp1</span><span class="op">[</span>, , <span class="st">"190321S_LCA10_X_FP97AG"</span><span class="op">]</span>
<span class="co">#&gt; Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 103 sampleMap rows not in names(experiments)</span>
<span class="co">#&gt;   removing 27 colData rownames not in sampleMap 'primary'</span>
<span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns</span></code></pre></div>
<p>An alternative that results in exactly the same output is using the <code>subsetByAssay</code> method.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">subsetByAssay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"190321S_LCA10_X_FP97AG"</span><span class="op">)</span>
<span class="co">#&gt; Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 103 sampleMap rows not in names(experiments)</span>
<span class="co">#&gt;   removing 27 colData rownames not in sampleMap 'primary'</span>
<span class="co">#&gt; An instance of class QFeatures containing 1 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns</span></code></pre></div>
</div>
<div id="subset-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-samples" class="anchor" aria-hidden="true"></a>Subset samples</h2>
<p>Subsetting samples is often performed after sample QC where we want to keep only quality samples and sample of interest. In our example, the different samples are either technical controls or single-cells (macrophages and monocytes). Suppose we are only interested in macrophages, we can subset the data as follows:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp1</span><span class="op">[</span>, <span class="va">scp1</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Macrophage"</span>, <span class="op">]</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 7 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 5 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 8 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 20 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 20 columns</span></code></pre></div>
<p>An alternative that results in exactly the same output is using the <code>subsetByColData</code> method.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">subsetByColData</span><span class="op">(</span><span class="va">scp1</span>, <span class="va">scp1</span><span class="op">$</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Macrophage"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 7 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 5 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 8 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 20 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 20 columns</span></code></pre></div>
</div>
<div id="subset-features" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-features" class="anchor" aria-hidden="true"></a>Subset features</h2>
<p>Subsetting for features does more than simply subsetting for the features of interest, it will also take the features that are linked to that feature. Here is an example, suppose we are interested in the <code>Q02878</code> protein.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp1</span><span class="op">[</span><span class="st">"Q02878"</span>, , <span class="op">]</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 9 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 10 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 0 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 11 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 1 rows and 38 columns</span></code></pre></div>
<p>You can see it indeed retrieved that protein from the <code>proteins</code> assay, but it also retrieved 11 associated peptides in the <code>peptides</code> assay and 19 associated PSMs in 2 different MS runs.</p>
<p>An alternative that results in exactly the same output is using the <code>subsetByColData</code> method.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-subsetBy.html" class="external-link">subsetByFeature</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"Q02878"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 9 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 10 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 0 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 11 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 1 rows and 38 columns</span></code></pre></div>
<p>You can also subset features based on the <code>rowData</code>. This is performed by <code>filterFeatures</code>. For example, we want to remove features that are associated to reverse sequence hits.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-filtering.html" class="external-link">filterFeatures</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="op">~</span> <span class="va">Reverse</span> <span class="op">!=</span> <span class="st">"+"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 126 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 132 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 176 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 422 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 0 rows and 38 columns</span></code></pre></div>
<p>Note however that if an assay is missing the variable that is used to filter the data (in this case the <code>proteins</code> assay), then all features for that assay are removed.</p>
<p>You can also subset the data based on the feature missingness using <code>filterNA</code>. In this example, we filter out proteins with more than 70 % missing data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">filterNA</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="st">"proteins"</span>, pNA <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 5 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 105 rows and 38 columns</span></code></pre></div>
</div>
</div>
<div id="common-processing-steps" class="section level1">
<h1 class="hasAnchor">
<a href="#common-processing-steps" class="anchor" aria-hidden="true"></a>Common processing steps</h1>
<p>We here provide a list of common processing steps that are encountered in single-cell proteomics data processing and that are already available in the <code>QFeatures</code> package.</p>
<p>All functions below require the user to select one or more assays from the <code>QFeatures</code> object. This is passed through the <code>i</code> argument. Note that some datasets may contain hundreds of assays and providing the assay selection manually can become cumbersome. We therefore suggest the user to use regular expression (aka regex) to chose from the <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> of the <code>QFeautres</code> object. A detailed cheatsheet about regex in R can be found <a href="https://rstudio-pubs-static.s3.amazonaws.com/74603_76cd14d5983f47408fdf0b323550b846.html" class="external-link">here</a>.</p>
<div id="missing-data-assignment" class="section level2">
<h2 class="hasAnchor">
<a href="#missing-data-assignment" class="anchor" aria-hidden="true"></a>Missing data assignment</h2>
<p>It often occurs that in MS experiements, 0 values are not true zeros but rather signal that is too weak to be detected. Therefore, it is advised to consider 0 values as missing data (<code>NA</code>). You can use <code>zeroIsNa</code> to automatically convert 0 values to <code>NA</code> in assays of interest. For instance, we here replace missing data in the <code>peptides</code> assay.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"peptides"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt;  5611  1509</span>
<span class="va">scp1</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-missing-data.html" class="external-link">zeroIsNA</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"peptides"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"peptides"</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; FALSE </span>
<span class="co">#&gt;  5611</span></code></pre></div>
</div>
<div id="feature-aggregation" class="section level2">
<h2 class="hasAnchor">
<a href="#feature-aggregation" class="anchor" aria-hidden="true"></a>Feature aggregation</h2>
<p>Shotgun proteomics analyses, bulk as well as single-cell, acquire and quantify peptides. However, biological inference is often performed at protein level. Protein quantitations can be estimated through feature aggregation. This is performed by <code>aggregateFeatures</code>, a function that takes an assay from the <code>Qfeatures</code> object and that aggregates its features with respect to a grouping variable in the <code>rowData</code> (<code>fcol</code>) and an aggregation function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-aggregate.html" class="external-link">aggregateFeatures</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="st">"190321S_LCA10_X_FP97AG"</span>, fcol <span class="op">=</span> <span class="st">"protein"</span>,
                  name <span class="op">=</span> <span class="st">"190321S_LCA10_X_FP97AG_aggr"</span>, 
                  fun <span class="op">=</span> <span class="fu">MsCoreUtils</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/MsCoreUtils/man/robustSummary.html" class="external-link">robustSummary</a></span><span class="op">)</span>
<span class="co">#&gt; Your row data contain missing values. Please read the relevant</span>
<span class="co">#&gt; section(s) in the aggregateFeatures manual page regarding the effects</span>
<span class="co">#&gt; of missing values on data aggregation.</span>
<span class="co">#&gt; Warning in rlm.default(X, expression, ...): 'rlm' failed to converge in 20 steps</span>
<span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns </span>
<span class="co">#&gt;  [6] 190321S_LCA10_X_FP97AG_aggr: SingleCellExperiment with 100 rows and 11 columns</span></code></pre></div>
<p>You can see that the aggregated function is added as a new assay to the <code>QFeatures</code> object. Note also that, under the hood, <code>aggregateFeatures</code> keeps track of the relationship between the features of the newly aggregated assay and its parent.</p>
</div>
<div id="normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization" class="anchor" aria-hidden="true"></a>Normalization</h2>
<p>An ubiquituous step that is performed in biological data analysis is normalization that is meant to remove undesired variability and to make different samples comparable. The <code>normalize</code> function offers an interface to a wide variety of normalization methods. See <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/normalize.html" class="external-link">?MsCoreUtils::normalize_matrix</a></code> for more details about the available normalization methods. Below, we normalize the samples so that they are mean centered.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">normalize</a></span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span>, method <span class="op">=</span> <span class="st">"center.mean"</span>, 
          name <span class="op">=</span> <span class="st">"proteins_mcenter"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns </span>
<span class="co">#&gt;  [6] proteins_mcenter: SingleCellExperiment with 292 rows and 38 columns</span></code></pre></div>
<p>Other custom normalization can be applied using the <code>sweep</code> method, where normalization factors have to be supplied manually. As an example, we here normalize the samples using a scaled size factor.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">/</span> <span class="fl">1E4</span>
<span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="st">"proteins"</span>, 
      MARGIN <span class="op">=</span> <span class="fl">2</span>, <span class="co">## 1 = by feature; 2 = by sample</span>
      STATS <span class="op">=</span> <span class="va">sf</span>, FUN <span class="op">=</span> <span class="st">"/"</span>,
      name <span class="op">=</span> <span class="st">"proteins_sf"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns </span>
<span class="co">#&gt;  [6] proteins_sf: SingleCellExperiment with 292 rows and 38 columns</span></code></pre></div>
</div>
<div id="log-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#log-transformation" class="anchor" aria-hidden="true"></a>Log transformation</h2>
<p>The <code>QFeatures</code> package also provide the <code>logTransform</code> function to facilitate the transformation of the quantitative data. We here show its usage by transforming the protein data using a base 2 logarithm with a pseudo-count of one.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html" class="external-link">logTransform</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="st">"proteins"</span>, base <span class="op">=</span> <span class="fl">2</span>, pc <span class="op">=</span> <span class="fl">1</span>, 
             name <span class="op">=</span> <span class="st">"proteins_log"</span><span class="op">)</span>
<span class="co">#&gt; An instance of class QFeatures containing 6 assays:</span>
<span class="co">#&gt;  [1] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 166 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 176 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 215 rows and 16 columns </span>
<span class="co">#&gt;  [4] peptides: SingleCellExperiment with 539 rows and 38 columns </span>
<span class="co">#&gt;  [5] proteins: SingleCellExperiment with 292 rows and 38 columns </span>
<span class="co">#&gt;  [6] proteins_log: SingleCellExperiment with 292 rows and 38 columns</span></code></pre></div>
</div>
<div id="imputation" class="section level2">
<h2 class="hasAnchor">
<a href="#imputation" class="anchor" aria-hidden="true"></a>Imputation</h2>
<p>Finally, <code>QFeatures</code> offers an interface to a wide variety of imputation methods to replace missing data by estimated values. The list of available methods is given by <code><a href="https://rdrr.io/pkg/MsCoreUtils/man/imputation.html" class="external-link">?MsCoreUtils::impute_matrix</a></code>. We demonstrate the use of this function by replacing missing data using KNN imputation.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">anyNA</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span>
<span class="va">scp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/impute.html" class="external-link">impute</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="st">"proteins"</span>, method <span class="op">=</span><span class="st">"knn"</span>, k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: impute</span>
<span class="co">#&gt; Warning in knnimp(x, k, maxmiss = rowmax, maxp = maxp): 284 rows with more than 50 % entries missing;</span>
<span class="co">#&gt;  mean imputation used for these rows</span>
<span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">anyNA</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp1</span>, <span class="st">"proteins"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] FALSE</span></code></pre></div>
</div>
</div>
<div id="data-visualization" class="section level1">
<h1 class="hasAnchor">
<a href="#data-visualization" class="anchor" aria-hidden="true"></a>Data visualization</h1>
<p>Visualization of the feature and sample metadata is rather straightforward since those are stored as tables (see section <em>Accessing the data</em>). From those tables, any visualization tool can be applied. Note however that using <code>ggplot2</code> require <code>data.frame</code>s or <code>tibble</code>s but <code>rowData</code> and <code>colData</code> are stored as <code>DFrames</code> objects. You can easily convert one data format to another. For example, we plot the parental ion fraction (measure of spectral purity) for each of the three MS batches.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">rbindRowData</a></span><span class="op">(</span><span class="va">scp1</span>, i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">rd</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">PIF</span>, 
        x <span class="op">=</span> <span class="va">assay</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 64 rows containing non-finite values (stat_boxplot).</span></code></pre></div>
<p><img src="QFeatures_nutshell_files/figure-html/vis1-1.png" width="624"></p>
<p>Combining the metadata and the quantitative data is more challenging since the risk of data mismatch is increased. The <code>QFeatures</code> package therefore provides th <code>longFormat</code> function to transform a <code>QFeatures</code> object in a long <code>DFrame</code> table. For instance, we plot the quantitative data distribution for the first assay according to the acquisition channel index and colour with respect to the sample type. Both pieces of information are taken from the <code>colData</code>, so we provide them as <code>colvars</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-class.html" class="external-link">longFormat</a></span><span class="op">(</span><span class="va">scp1</span><span class="op">[</span>, , <span class="fl">1</span><span class="op">]</span>, 
                 colvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleType"</span>, <span class="st">"Channel"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: 'experiments' dropped; see 'metadata'</span>
<span class="co">#&gt; harmonizing input:</span>
<span class="co">#&gt;   removing 103 sampleMap rows not in names(experiments)</span>
<span class="co">#&gt;   removing 27 colData rownames not in sampleMap 'primary'</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">lf</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Channel</span>, 
        y <span class="op">=</span> <span class="va">value</span>,
        colour <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="QFeatures_nutshell_files/figure-html/longFormat-1.png" width="700"></p>
<p>A more in-depth tutorial about data visualization from a <code>QFeatures</code> object is provided in the <code>QFeautres</code> visualization <a href="https://rformassspectrometry.github.io/QFeatures/articles/Visualization.html" class="external-link">vignette</a>.</p>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor" aria-hidden="true"></a>Session information</h1>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ggplot2_3.3.5               scp_1.3.1                  
 [3] QFeatures_1.3.5             MultiAssayExperiment_1.19.1
 [5] SummarizedExperiment_1.23.1 Biobase_2.53.0             
 [7] GenomicRanges_1.45.0        GenomeInfoDb_1.29.3        
 [9] IRanges_2.27.0              S4Vectors_0.31.0           
[11] BiocGenerics_0.39.1         MatrixGenerics_1.5.1       
[13] matrixStats_0.59.0          BiocStyle_2.21.3           

loaded via a namespace (and not attached):
 [1] sass_0.4.0                  jsonlite_1.7.2             
 [3] bslib_0.2.5.1               BiocManager_1.30.16        
 [5] highr_0.9                   GenomeInfoDbData_1.2.6     
 [7] yaml_2.2.1                  impute_1.67.0              
 [9] pillar_1.6.1                lattice_0.20-44            
[11] glue_1.4.2                  digest_0.6.27              
[13] XVector_0.33.0              colorspace_2.0-2           
[15] htmltools_0.5.1.1           Matrix_1.3-4               
[17] pkgconfig_2.0.3             bookdown_0.22              
[19] zlibbioc_1.39.0             purrr_0.3.4                
[21] scales_1.1.1                tibble_3.1.2               
[23] farver_2.1.0                generics_0.1.0             
[25] AnnotationFilter_1.17.0     ellipsis_0.3.2             
[27] cachem_1.0.5                withr_2.4.2                
[29] lazyeval_0.2.2              magrittr_2.0.1             
[31] crayon_1.4.1                memoise_2.0.0              
[33] evaluate_0.14               fs_1.5.0                   
[35] fansi_0.5.0                 MASS_7.3-54                
[37] textshaping_0.3.5           tools_4.1.0                
[39] lifecycle_1.0.0             stringr_1.4.0              
[41] munsell_0.5.0               cluster_2.1.2              
[43] DelayedArray_0.19.1         compiler_4.1.0             
[45] pkgdown_1.6.1.9001          jquerylib_0.1.4            
[47] systemfonts_1.0.2           rlang_0.4.11               
[49] grid_4.1.0                  RCurl_1.98-1.3             
[51] MsCoreUtils_1.5.0           SingleCellExperiment_1.15.1
[53] igraph_1.2.6                labeling_0.4.2             
[55] bitops_1.0-7                rmarkdown_2.9              
[57] gtable_0.3.0                R6_2.5.0                   
[59] knitr_1.33                  dplyr_1.0.7                
[61] fastmap_1.1.0               utf8_1.2.1                 
[63] clue_0.3-59                 rprojroot_2.0.2            
[65] ragg_1.1.3                  ProtGenerics_1.25.1        
[67] desc_1.3.0                  stringi_1.7.3              
[69] vctrs_0.3.8                 tidyselect_1.1.1           
[71] xfun_0.24                  </code></pre>
</div>
<div id="reference" class="section level1">
<h1 class="hasAnchor">
<a href="#reference" class="anchor" aria-hidden="true"></a>Reference</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
