<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load Single-Cell Proteomics data using `readSCP` • scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Load Single-Cell Proteomics data using `readSCP`">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="../articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="../articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Load Single-Cell Proteomics data using
<code>readSCP</code>
</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
                                    <h4 data-toc-skip class="author">Christophe
Vanderaa</h4>
                        
            <h4 data-toc-skip class="date">26 February 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/read_scp.Rmd" class="external-link"><code>vignettes/read_scp.Rmd</code></a></small>
      <div class="hidden name"><code>read_scp.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="the-scp-data-framework">The <code>scp</code> data framework<a class="anchor" aria-label="anchor" href="#the-scp-data-framework"></a>
</h2>
<p>Our data structure is relying on two curated data classes:
<code>QFeatures</code> (<span class="citation">Gatto and Vanderaa
(2023)</span>) and <code>SingleCellExperiment</code> (<span class="citation">Amezquita et al. (2020)</span>). <code>QFeatures</code>
is dedicated to the manipulation and processing of MS-based quantitative
data. It explicitly records the successive steps to allow users to
navigate up and down the different MS levels.
<code>SingleCellExperiment</code> is another class designed as an
efficient data container that serves as an interface to state-of-the-art
methods and algorithms for single-cell data. Our framework combines the
two classes to inherit from their respective advantages.</p>
<p>Because mass spectrometry (MS)-based single-cell proteomics (SCP)
only captures the proteome of between one and a few tens of single-cells
in a single run, the data is usually acquired across many MS batches.
Therefore, the data for each run should conceptually be stored in its
own container, that we here call a <em>set</em>. The expected input for
working with the <code>scp</code> package is quantification data of
peptide to spectrum matches (PSM). These data can then be processed to
reconstruct peptide and protein data. The links between related features
across different sets are stored to facilitate manipulation and
visualization of of PSM, peptide and protein data. This is conceptually
shown below.</p>
<div class="figure" style="text-align: center">
<img src="figures/SCP_framework.png" alt="The `scp` framework relies on `SingleCellExperiment` and `QFeatures` objects" width="100%"><p class="caption">
The <code>scp</code> framework relies on
<code>SingleCellExperiment</code> and <code>QFeatures</code> objects
</p>
</div>
<p>The main input table required for starting an analysis with
<code>scp</code> is called the <code>assayData</code>.</p>
</div>
<div class="section level2">
<h2 id="assaydata-table">
<code>assayData</code> table<a class="anchor" aria-label="anchor" href="#assaydata-table"></a>
</h2>
<p>The <code>assayData</code> table is generated after the
identification and quantification of the MS spectra by a pre-processing
software such as MaxQuant, ProteomeDiscoverer or MSFragger (the <a href="https://en.wikipedia.org/wiki/List_of_mass_spectrometry_software" class="external-link">list</a>
of available software is actually much longer). We will here use as an
example a data table that has been generated by MaxQuant. The table is
available from the <code>scp</code> package and is called
<code>mqScpData</code> (for MaxQuant generated SCP data).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mqScpData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1361  149</span></span></code></pre></div>
<p>In this toy example, there are 1361 rows corresponding to features
(quantified PSMs) and 149 columns corresponding to different data fields
recorded by MaxQuant during the processing of the MS spectra. There are
three types of columns:</p>
<ul>
<li>Quantification columns (<code>quantCols</code>): 1 to n (depending
on technology)</li>
<li>Run identifier column (<code>runCol</code>): <em>e.g.</em> file
name</li>
<li>Feature annotations: <em>e.g.</em> peptide sequence, ion charge,
protein name</li>
</ul>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_inputTable.png" alt="Conceptual representation of the `assayData` input table" width="450px"><p class="caption">
Conceptual representation of the <code>assayData</code> input table
</p>
</div>
<div class="section level4">
<h4 id="quantification-columns-quantcols">Quantification columns (<code>quantCols</code>)<a class="anchor" aria-label="anchor" href="#quantification-columns-quantcols"></a>
</h4>
<p>The quantification data can be composed of one (in case of label-free
acquisition) or multiple columns (in case of multiplexing). In the
example data set, the columns holding the quantification, the
<code>quantCols</code>, start with <code>Reporter.intensity.</code>
followed by a number.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">quantCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Reporter.intensity.\\d"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span>,</span>
<span>                  value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Reporter.intensity.1"  "Reporter.intensity.2"  "Reporter.intensity.3" </span></span>
<span><span class="co">#&gt;  [4] "Reporter.intensity.4"  "Reporter.intensity.5"  "Reporter.intensity.6" </span></span>
<span><span class="co">#&gt;  [7] "Reporter.intensity.7"  "Reporter.intensity.8"  "Reporter.intensity.9" </span></span>
<span><span class="co">#&gt; [10] "Reporter.intensity.10" "Reporter.intensity.11" "Reporter.intensity.12"</span></span>
<span><span class="co">#&gt; [13] "Reporter.intensity.13" "Reporter.intensity.14" "Reporter.intensity.15"</span></span>
<span><span class="co">#&gt; [16] "Reporter.intensity.16"</span></span></code></pre></div>
<p>As you may notice, the example data was acquired using a TMT-16
protocol since we retrieve 16 quantification columns. Actually, some
runs were acquired using a TMT-11 protocol (11 labels) but we will come
back to this later.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="va">quantCols</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Reporter.intensity.1 Reporter.intensity.2 Reporter.intensity.3</span></span>
<span><span class="co">#&gt; 1                61251               501.71               3731.3</span></span>
<span><span class="co">#&gt; 2                58648              1099.80               2837.7</span></span>
<span><span class="co">#&gt; 3               150350              3705.00               9361.0</span></span>
<span><span class="co">#&gt; 4                27347               405.90               1525.2</span></span>
<span><span class="co">#&gt; 5                84035               583.09               4092.3</span></span>
<span><span class="co">#&gt; 6                44895               700.23               2283.0</span></span>
<span><span class="co">#&gt;   Reporter.intensity.4 Reporter.intensity.5 Reporter.intensity.6</span></span>
<span><span class="co">#&gt; 1              1643.30               871.84               981.87</span></span>
<span><span class="co">#&gt; 2               494.32               349.26              1030.50</span></span>
<span><span class="co">#&gt; 3                 0.00              1945.40              1188.60</span></span>
<span><span class="co">#&gt; 4                 0.00                 0.00               318.74</span></span>
<span><span class="co">#&gt; 5               530.13               718.13              2204.50</span></span>
<span><span class="co">#&gt; 6              1109.60                 0.00               675.79</span></span>
<span><span class="co">#&gt;   Reporter.intensity.7 Reporter.intensity.8 Reporter.intensity.9</span></span>
<span><span class="co">#&gt; 1              1200.10               939.06              1457.50</span></span>
<span><span class="co">#&gt; 2                 0.00              1214.10               800.58</span></span>
<span><span class="co">#&gt; 3              1574.00              2302.10              2176.10</span></span>
<span><span class="co">#&gt; 4                 0.00               519.81                 0.00</span></span>
<span><span class="co">#&gt; 5               960.51               453.77              1188.40</span></span>
<span><span class="co">#&gt; 6                 0.00               809.38               668.88</span></span>
<span><span class="co">#&gt;   Reporter.intensity.10 Reporter.intensity.11 Reporter.intensity.12</span></span>
<span><span class="co">#&gt; 1               1329.80                981.83                    NA</span></span>
<span><span class="co">#&gt; 2                807.79                391.38                    NA</span></span>
<span><span class="co">#&gt; 3               1399.50               1307.50                2192.4</span></span>
<span><span class="co">#&gt; 4                507.23                370.79                    NA</span></span>
<span><span class="co">#&gt; 5                740.99                  0.00                    NA</span></span>
<span><span class="co">#&gt; 6               1467.50                901.38                    NA</span></span>
<span><span class="co">#&gt;   Reporter.intensity.13 Reporter.intensity.14 Reporter.intensity.15</span></span>
<span><span class="co">#&gt; 1                    NA                    NA                    NA</span></span>
<span><span class="co">#&gt; 2                    NA                    NA                    NA</span></span>
<span><span class="co">#&gt; 3                1791.4                1727.5                2157.3</span></span>
<span><span class="co">#&gt; 4                    NA                    NA                    NA</span></span>
<span><span class="co">#&gt; 5                    NA                    NA                    NA</span></span>
<span><span class="co">#&gt; 6                    NA                    NA                    NA</span></span>
<span><span class="co">#&gt;   Reporter.intensity.16</span></span>
<span><span class="co">#&gt; 1                    NA</span></span>
<span><span class="co">#&gt; 2                    NA</span></span>
<span><span class="co">#&gt; 3                  1398</span></span>
<span><span class="co">#&gt; 4                    NA</span></span>
<span><span class="co">#&gt; 5                    NA</span></span>
<span><span class="co">#&gt; 6                    NA</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-identifier-column-runcol">Run identifier column (<code>runCol</code>)<a class="anchor" aria-label="anchor" href="#run-identifier-column-runcol"></a>
</h4>
<p>This column provides the identifier of the MS runs in which each PSM
was acquired. MaxQuant uses the raw file name to identify the run.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">$</span><span class="va">Raw.file</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "190321S_LCA10_X_FP97AG"        "190222S_LCA9_X_FP94BM"        </span></span>
<span><span class="co">#&gt; [3] "190914S_LCB3_X_16plex_Set_21"  "190321S_LCA10_X_FP97_blank_01"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="feature-annotations">Feature annotations<a class="anchor" aria-label="anchor" href="#feature-annotations"></a>
</h4>
<p>The remaining columns in the <code>mqScpData</code> table contain
information used or generated during the identification of the MS
spectra. For instance, you may find the charge of the parent ion, the
score and probability of a correct match between the MS spectrum and a
peptide sequence, the sequence of the best matching peptide, its length,
its modifications, the retention time of the peptide on the LC, the
protein(s) the peptide originates from and much more.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Charge"</span>, <span class="st">"Score"</span>, <span class="st">"PEP"</span>, <span class="st">"Sequence"</span>, <span class="st">"Length"</span>,</span>
<span>                   <span class="st">"Retention.time"</span>, <span class="st">"Proteins"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Charge  Score        PEP    Sequence Length Retention.time</span></span>
<span><span class="co">#&gt; 1      2 41.029 5.2636e-04   ATNFLAHEK      9         65.781</span></span>
<span><span class="co">#&gt; 2      2 44.349 5.8789e-04   ATNFLAHEK      9         63.787</span></span>
<span><span class="co">#&gt; 3      2 51.066 4.0315e-24 SHTILLVQPTK     11         71.884</span></span>
<span><span class="co">#&gt; 4      2 63.816 4.7622e-06 SHTILLVQPTK     11         68.633</span></span>
<span><span class="co">#&gt; 5      2 74.464 6.8709e-09 SHTILLVQPTK     11         71.946</span></span>
<span><span class="co">#&gt; 6      2 41.502 5.3705e-02     SLVIPEK      7         76.204</span></span>
<span><span class="co">#&gt;               Proteins</span></span>
<span><span class="co">#&gt; 1 sp|P29692|EF1D_HUMAN</span></span>
<span><span class="co">#&gt; 2 sp|P29692|EF1D_HUMAN</span></span>
<span><span class="co">#&gt; 3  sp|P84090|ERH_HUMAN</span></span>
<span><span class="co">#&gt; 4  sp|P84090|ERH_HUMAN</span></span>
<span><span class="co">#&gt; 5  sp|P84090|ERH_HUMAN</span></span>
<span><span class="co">#&gt; 6 sp|P62269|RS18_HUMAN</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="coldata-table">
<code>colData</code> table<a class="anchor" aria-label="anchor" href="#coldata-table"></a>
</h2>
<p>The <code>colData</code> table contains the experimental design
generated by the researcher. The rows of the sample table correspond to
a sample in the experiment and the columns correspond to the available
annotations about the sample. We will here use the second example
table:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sampleAnnotation"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sampleAnnotation</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  runCol            quantCols SampleType lcbatch sortday digest</span></span>
<span><span class="co">#&gt; 1 190222S_LCA9_X_FP94BM Reporter.intensity.1    Carrier    LCA9      s8      N</span></span>
<span><span class="co">#&gt; 2 190222S_LCA9_X_FP94BM Reporter.intensity.2  Reference    LCA9      s8      N</span></span>
<span><span class="co">#&gt; 3 190222S_LCA9_X_FP94BM Reporter.intensity.3     Unused    LCA9      s8      N</span></span>
<span><span class="co">#&gt; 4 190222S_LCA9_X_FP94BM Reporter.intensity.4   Monocyte    LCA9      s8      N</span></span>
<span><span class="co">#&gt; 5 190222S_LCA9_X_FP94BM Reporter.intensity.5      Blank    LCA9      s8      N</span></span>
<span><span class="co">#&gt; 6 190222S_LCA9_X_FP94BM Reporter.intensity.6   Monocyte    LCA9      s8      N</span></span></code></pre></div>
<p>The <code>colData</code> table may contain any information about the
samples. For example, useful information could be the type of sample
that is analysed, a phenotype known from the experimental design, the MS
batch, the acquisition date, MS settings used to acquire the sample, the
LC batch, the sample preparation batch, etc… However, <code>scp</code>
<strong>requires</strong> 2 specific columns in the <code>colData</code>
table:</p>
<ol style="list-style-type: decimal">
<li>
<code>runCol</code>: this column provides the MS run names (that
match the <code>Raw.file</code> column in the <code>assayData</code>
table).</li>
<li>
<code>quantCols</code>: this column tells <code>scp</code> the names
of the columns in the feature data holds the quantification of the
corresponding sample.</li>
</ol>
<p>These two columns allow <code>scp</code> to correctly split and match
data that were acquired across multiple acquisition runs.</p>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_sampleTable.png" alt="Conceptual representation of the sample table" width="450px"><p class="caption">
Conceptual representation of the sample table
</p>
</div>
</div>
<div class="section level2">
<h2 id="readscp">
<code>readSCP()</code><a class="anchor" aria-label="anchor" href="#readscp"></a>
</h2>
<p><code>readSCP</code> is the function that converts the
<code>assayData</code> and the <code>colData</code> into a
<code>QFeatures</code> object following the data structure described
above, that is storing the data belonging to each MS batch in a separate
<code>SingleCellExperiment</code> object.</p>
<div class="section level3">
<h3 id="sample-names">Sample names<a class="anchor" aria-label="anchor" href="#sample-names"></a>
</h3>
<p><code><a href="../reference/readSCP.html">readSCP()</a></code> automatically assigns names that are unique
across all samples in all sets. This is performed by appending the name
of the MS run where a given sample is found with the name of the
quantification column for that sample. Suppose a sample belongs to batch
<code>190222S_LCA9_X_FP94BM</code> and the quantification values in the
<code>assayData</code> table are found in the column called
<code>Reporter.intensity.3</code>, then the sample name will become
<code>190222S_LCA9_X_FP94BM_Reporter.intensity.3</code>.</p>
</div>
<div class="section level3">
<h3 id="special-case-empty-samples">Special case: empty samples<a class="anchor" aria-label="anchor" href="#special-case-empty-samples"></a>
</h3>
<p>In some rare cases, it can be beneficial to remove empty samples (all
quantifications are <code>NA</code>) from the sets. Such samples can
occur when samples that were acquired with different multiplexing labels
are merged in a single table. For instance, the SCoPE2 data we provide
as an example contains runs that were acquired with two TMT protocols.
The 3 first sets were acquired using the TMT-11 protocol and the last
set was acquired using a TMT-16 protocol. The missing label channels in
the TMT-11 data are filled with <code>NA</code>s. When setting
<code>removeEmptyCols = TRUE</code>, <code><a href="../reference/readSCP.html">readSCP()</a></code>
automatically detects and removes columns containing only
<code>NA</code>s,</p>
</div>
<div class="section level3">
<h3 id="running-readscp">Running <code>readSCP</code><a class="anchor" aria-label="anchor" href="#running-readscp"></a>
</h3>
<p>We convert the sample and the feature data into a
<code>QFeatures</code> object in a single command thanks to
<code>readSCP</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCP.html">readSCP</a></span><span class="op">(</span>assayData <span class="op">=</span> <span class="va">mqScpData</span>,</span>
<span>                colData <span class="op">=</span> <span class="va">sampleAnnotation</span>,</span>
<span>                runCol <span class="op">=</span> <span class="st">"Raw.file"</span>,</span>
<span>                removeEmptyCols <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking arguments.</span></span>
<span><span class="co">#&gt; Loading data as a 'SummarizedExperiment' object.</span></span>
<span><span class="co">#&gt; Splitting data in runs.</span></span>
<span><span class="co">#&gt; Formatting sample annotations (colData).</span></span>
<span><span class="co">#&gt; Formatting data as a 'QFeatures' object.</span></span>
<span><span class="co">#&gt; An instance of class QFeatures containing 4 set(s):</span></span>
<span><span class="co">#&gt;  [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 395 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [2] 190321S_LCA10_X_FP97_blank_01: SingleCellExperiment with 109 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [3] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 487 rows and 11 columns </span></span>
<span><span class="co">#&gt;  [4] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 370 rows and 16 columns</span></span></code></pre></div>
<p>The object returned by <code><a href="../reference/readSCP.html">readSCP()</a></code> is a
<code>QFeatures</code> object containing 4
<code>SingleCellExperiment</code> sets that have been named after the 4
MS batches. Each set contains either 11 or 16 columns (samples)
depending on the TMT labelling strategy and a variable number of rows
(quantified PSMs). Each piece of information can easily be retrieved
thanks to the <code>QFeatures</code> architectures. As mentioned in
another <a href="https://uclouvain-cbio.github.io/scp/articles/QFeatures_nutshell.html" class="external-link">vignette</a>,
the <code>colData</code> is retrieved using its dedicated function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">#&gt;                                                   runCol     quantCols</span></span>
<span><span class="co">#&gt;                                              &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.1 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.2 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.3 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.4 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.5 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.6 190222S_LC... Reporter.i...</span></span>
<span><span class="co">#&gt;                                             SampleType     lcbatch     sortday</span></span>
<span><span class="co">#&gt;                                            &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.1     Carrier        LCA9          s8</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.2   Reference        LCA9          s8</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.3      Unused        LCA9          s8</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.4    Monocyte        LCA9          s8</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.5       Blank        LCA9          s8</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.6    Monocyte        LCA9          s8</span></span>
<span><span class="co">#&gt;                                                 digest</span></span>
<span><span class="co">#&gt;                                            &lt;character&gt;</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.1           N</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.2           N</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.3           N</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.4           N</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.5           N</span></span>
<span><span class="co">#&gt; 190222S_LCA9_X_FP94BM_Reporter.intensity.6           N</span></span></code></pre></div>
<p>The feature annotations are retrieved from the <code>rowData</code>.
Since the feature annotations are specific to each set, we need to tell
from which set we want to get the <code>rowData</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 5 columns</span></span>
<span><span class="co">#&gt;              uid      Sequence    Length Modifications Modified.sequence</span></span>
<span><span class="co">#&gt;      &lt;character&gt;   &lt;character&gt; &lt;integer&gt;   &lt;character&gt;       &lt;character&gt;</span></span>
<span><span class="co">#&gt; 2  _(Acetyl (...     ATNFLAHEK         9 Acetyl (Pr...     _(Acetyl (...</span></span>
<span><span class="co">#&gt; 4  _(Acetyl (... SHTILLVQPT...        11 Acetyl (Pr...     _(Acetyl (...</span></span>
<span><span class="co">#&gt; 6  _(Acetyl (...       SLVIPEK         7 Acetyl (Pr...     _(Acetyl (...</span></span>
<span><span class="co">#&gt; 9  _AAGLALK_ ...       AAGLALK         7    Unmodified         _AAGLALK_</span></span>
<span><span class="co">#&gt; 12 _AALSAGK_ ...       AALSAGK         7    Unmodified         _AALSAGK_</span></span>
<span><span class="co">#&gt; 15 _AAPEASGTP... AAPEASGTPS...        16    Unmodified     _AAPEASGTP...</span></span></code></pre></div>
<p>Finally, we can also retrieve the quantification matrix for a set of
interest:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">scp</span>, <span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.1</span></span>
<span><span class="co">#&gt; 2                                     58648.0</span></span>
<span><span class="co">#&gt; 4                                     27347.0</span></span>
<span><span class="co">#&gt; 6                                     44895.0</span></span>
<span><span class="co">#&gt; 9                                    122070.0</span></span>
<span><span class="co">#&gt; 12                                    58605.0</span></span>
<span><span class="co">#&gt; 15                                     5006.5</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.2</span></span>
<span><span class="co">#&gt; 2                                     1099.80</span></span>
<span><span class="co">#&gt; 4                                      405.90</span></span>
<span><span class="co">#&gt; 6                                      700.23</span></span>
<span><span class="co">#&gt; 9                                     1153.50</span></span>
<span><span class="co">#&gt; 12                                     895.25</span></span>
<span><span class="co">#&gt; 15                                     517.86</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.3</span></span>
<span><span class="co">#&gt; 2                                     2837.70</span></span>
<span><span class="co">#&gt; 4                                     1525.20</span></span>
<span><span class="co">#&gt; 6                                     2283.00</span></span>
<span><span class="co">#&gt; 9                                     7361.90</span></span>
<span><span class="co">#&gt; 12                                    2763.80</span></span>
<span><span class="co">#&gt; 15                                     446.19</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.4</span></span>
<span><span class="co">#&gt; 2                                      494.32</span></span>
<span><span class="co">#&gt; 4                                        0.00</span></span>
<span><span class="co">#&gt; 6                                     1109.60</span></span>
<span><span class="co">#&gt; 9                                     1732.30</span></span>
<span><span class="co">#&gt; 12                                     867.82</span></span>
<span><span class="co">#&gt; 15                                     458.17</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.5</span></span>
<span><span class="co">#&gt; 2                                      349.26</span></span>
<span><span class="co">#&gt; 4                                        0.00</span></span>
<span><span class="co">#&gt; 6                                        0.00</span></span>
<span><span class="co">#&gt; 9                                     1515.60</span></span>
<span><span class="co">#&gt; 12                                    1050.30</span></span>
<span><span class="co">#&gt; 15                                     467.90</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.6</span></span>
<span><span class="co">#&gt; 2                                     1030.50</span></span>
<span><span class="co">#&gt; 4                                      318.74</span></span>
<span><span class="co">#&gt; 6                                      675.79</span></span>
<span><span class="co">#&gt; 9                                     2252.00</span></span>
<span><span class="co">#&gt; 12                                    1268.70</span></span>
<span><span class="co">#&gt; 15                                     649.50</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.7</span></span>
<span><span class="co">#&gt; 2                                        0.00</span></span>
<span><span class="co">#&gt; 4                                        0.00</span></span>
<span><span class="co">#&gt; 6                                        0.00</span></span>
<span><span class="co">#&gt; 9                                      444.48</span></span>
<span><span class="co">#&gt; 12                                     532.30</span></span>
<span><span class="co">#&gt; 15                                     259.84</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.8</span></span>
<span><span class="co">#&gt; 2                                     1214.10</span></span>
<span><span class="co">#&gt; 4                                      519.81</span></span>
<span><span class="co">#&gt; 6                                      809.38</span></span>
<span><span class="co">#&gt; 9                                     2343.80</span></span>
<span><span class="co">#&gt; 12                                    1073.10</span></span>
<span><span class="co">#&gt; 15                                     533.55</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.9</span></span>
<span><span class="co">#&gt; 2                                      800.58</span></span>
<span><span class="co">#&gt; 4                                        0.00</span></span>
<span><span class="co">#&gt; 6                                      668.88</span></span>
<span><span class="co">#&gt; 9                                     3100.20</span></span>
<span><span class="co">#&gt; 12                                     911.30</span></span>
<span><span class="co">#&gt; 15                                     393.53</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.10</span></span>
<span><span class="co">#&gt; 2                                       807.79</span></span>
<span><span class="co">#&gt; 4                                       507.23</span></span>
<span><span class="co">#&gt; 6                                      1467.50</span></span>
<span><span class="co">#&gt; 9                                      1825.20</span></span>
<span><span class="co">#&gt; 12                                     1300.00</span></span>
<span><span class="co">#&gt; 15                                      463.26</span></span>
<span><span class="co">#&gt;    190222S_LCA9_X_FP94BM_Reporter.intensity.11</span></span>
<span><span class="co">#&gt; 2                                       391.38</span></span>
<span><span class="co">#&gt; 4                                       370.79</span></span>
<span><span class="co">#&gt; 6                                       901.38</span></span>
<span><span class="co">#&gt; 9                                      2372.50</span></span>
<span><span class="co">#&gt; 12                                     1185.90</span></span>
<span><span class="co">#&gt; 15                                      353.04</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="under-the-hood">Under the hood<a class="anchor" aria-label="anchor" href="#under-the-hood"></a>
</h3>
<p><code>readSCP</code> proceeds as follows:</p>
<ol style="list-style-type: decimal">
<li>The <code>assayData</code> table must be provided as a
<code>data.frame</code>. <code><a href="../reference/readSCP.html">readSCP()</a></code> converts the table to a
<code>SingleCellExperiment</code> object but it needs to know which
column(s) store the quantitative data. Those column name(s) is/are
provided by the <code>quantCols</code> field in the annotation table
(<code>colData</code> argument).</li>
</ol>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_step1.png" alt="Step1: Convert the input table to a `SingleCellExperiment` object" width="450px"><p class="caption">
Step1: Convert the input table to a <code>SingleCellExperiment</code>
object
</p>
</div>
<ol start="2" style="list-style-type: decimal">
<li>The <code>SingleCellExperiment</code> object is then split according
to the acquisition run. The split is performed depending on the
<code>runCol</code> field in <code>assayData</code>. It is also
indicated in the <code>runCol</code> argument. In this case, the data
will be split according to the <code>Raw.file</code> column in
<code>mqScpData</code>. <code>Raw.file</code> contains the names of the
acquisition runs that was used by MaxQuant to retrieve the raw data
files.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_step2.png" alt="Step2: Split by acquisition run" width="500px"><p class="caption">
Step2: Split by acquisition run
</p>
</div>
<ol start="3" style="list-style-type: decimal">
<li>The sample annotations is generated from the supplied sample table
(<code>colData</code> argument). Note that in order for
<code><a href="../reference/readSCP.html">readSCP()</a></code> to correctly match the feature data with the
annotations, <code>colData</code> must contain a <code>runCol</code>
column with run names and a <code>quantCols</code> column with the names
of the quantitative columns in <code>assayData</code>.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_step3.png" alt="Step3: Adding and matching the sample annotations" width="700px"><p class="caption">
Step3: Adding and matching the sample annotations
</p>
</div>
<ol start="4" style="list-style-type: decimal">
<li>Finally, the <code>SingleCellExperiment</code> sets and the
<code>colData</code> are converted to a <code>QFeatures</code>
object.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="figures/readSCP_step4.png" alt="Step4: Converting to a `QFeatures`" width="600px"><p class="caption">
Step4: Converting to a <code>QFeatures</code>
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="what-about-label-free-scp">What about label-free SCP?<a class="anchor" aria-label="anchor" href="#what-about-label-free-scp"></a>
</h2>
<p>The <code>scp</code> package is meant for both label-free and
multiplexed SCP data. Like in the example above, the label-free data
should contain the batch names in both the feature data and the sample
data. The sample data must also contain a column that points to the
columns of the feature data that contains the quantifications. Since
label-free SCP acquires one single-cell per run, this sample data column
should point the same column for all samples. Moreover, this means that
each PSM set will contain a single column.</p>
</div>
<div class="section level2">
<h2 id="what-about-other-input-formats">What about other input formats?<a class="anchor" aria-label="anchor" href="#what-about-other-input-formats"></a>
</h2>
<p><code><a href="../reference/readSCP.html">readSCP()</a></code> should work with any PSM quantification table
that is output by a pre-processing software. For instance, you can
easily import the PSM tables generated by ProteomeDiscoverer. The batch
names are contained in the <code>File ID</code> column (that should be
supplied as the <code>batchCol</code> argument to
<code><a href="../reference/readSCP.html">readSCP()</a></code>). The quantification columns are contained in the
columns starting with <code>Abundance</code>, eventually followed by a
multiplexing tag name. These columns should be stored in a dedicated
column of the sample data to be supplied as <code>runCol</code> to
<code><a href="../reference/readSCP.html">readSCP()</a></code>.</p>
<p>If your input cannot be loaded using the procedure described in this
vignette, you can submit a feature request (see next section).</p>
<p>The <code><a href="../reference/readSCP.html">readSCPfromDIANN()</a></code> function is adapted to import
label-free and plexDIA/mTRAQ <code>Report.tsv</code> files generated by
DIA-NN.</p>
<p>For more information, see the <code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeatures.html" class="external-link">readQFeatures()</a></code> and
<code><a href="https://rdrr.io/pkg/QFeatures/man/readQFeaturesFromDIANN.html" class="external-link">readQFeaturesFromDIANN()</a></code> manual pages, that described the
main principle that concern the data import and formatting.</p>
</div>
<div class="section level2">
<h2 id="need-help">Need help?<a class="anchor" aria-label="anchor" href="#need-help"></a>
</h2>
<p>You can open an issue on the <a href="https://github.com/UCLouvain-CBIO/scp/issues" class="external-link">GitHub
repository</a> in case of troubles when loading your SCP data with
<code><a href="../reference/readSCP.html">readSCP()</a></code>. Any suggestion or feature request about the
function or the documentation are also warmly welcome.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R Under development (unstable) (2025-02-15 r87726)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scp_1.17.1                  QFeatures_1.17.4           
 [3] MultiAssayExperiment_1.33.9 SummarizedExperiment_1.37.0
 [5] Biobase_2.67.0              GenomicRanges_1.59.1       
 [7] GenomeInfoDb_1.43.4         IRanges_2.41.3             
 [9] S4Vectors_0.45.4            BiocGenerics_0.53.6        
[11] generics_0.1.3              MatrixGenerics_1.19.1      
[13] matrixStats_1.5.0           BiocStyle_2.35.0           

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1            dplyr_1.1.4                
 [3] fastmap_1.2.0               SingleCellExperiment_1.29.1
 [5] lazyeval_0.2.2              nipals_1.0                 
 [7] digest_0.6.37               lifecycle_1.0.4            
 [9] cluster_2.1.8               ProtGenerics_1.39.2        
[11] magrittr_2.0.3              compiler_4.5.0             
[13] rlang_1.1.5                 sass_0.4.9                 
[15] tools_4.5.0                 igraph_2.1.4               
[17] yaml_2.3.10                 knitr_1.49                 
[19] S4Arrays_1.7.3              htmlwidgets_1.6.4          
[21] DelayedArray_0.33.6         plyr_1.8.9                 
[23] RColorBrewer_1.1-3          abind_1.4-8                
[25] purrr_1.0.4                 desc_1.4.3                 
[27] grid_4.5.0                  colorspace_2.1-1           
[29] ggplot2_3.5.1               scales_1.3.0               
[31] MASS_7.3-64                 cli_3.6.4                  
[33] rmarkdown_2.29              crayon_1.5.3               
[35] ragg_1.3.3                  metapod_1.15.0             
[37] httr_1.4.7                  reshape2_1.4.4             
[39] BiocBaseUtils_1.9.0         cachem_1.1.0               
[41] stringr_1.5.1               AnnotationFilter_1.31.0    
[43] BiocManager_1.30.25         XVector_0.47.2             
[45] vctrs_0.6.5                 Matrix_1.7-2               
[47] jsonlite_1.9.0              slam_0.1-55                
[49] bookdown_0.42               IHW_1.35.0                 
[51] ggrepel_0.9.6               clue_0.3-66                
[53] systemfonts_1.2.1           tidyr_1.3.1                
[55] jquerylib_0.1.4             glue_1.8.0                 
[57] pkgdown_2.1.1.9000          stringi_1.8.4              
[59] gtable_0.3.6                UCSC.utils_1.3.1           
[61] munsell_0.5.1               lpsymphony_1.35.0          
[63] tibble_3.2.1                pillar_1.10.1              
[65] htmltools_0.5.8.1           GenomeInfoDbData_1.2.13    
[67] R6_2.6.1                    textshaping_1.0.0          
[69] evaluate_1.0.3              lattice_0.22-6             
[71] bslib_0.9.0                 Rcpp_1.0.14                
[73] fdrtool_1.2.18              SparseArray_1.7.6          
[75] xfun_0.51                   MsCoreUtils_1.19.0         
[77] fs_1.6.5                    pkgconfig_2.0.3            </code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Amezquita2019-bf" class="csl-entry">
Amezquita, Robert A, Aaron T L Lun, Etienne Becht, Vince J Carey,
Lindsay N Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020.
<span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span>
<em>Nat. Methods</em> 17 (2): 137–45.
</div>
<div id="ref-Gatto2023-ry" class="csl-entry">
Gatto, Laurent, and Christophe Vanderaa. 2023. <span>“QFeatures:
Quantitative Features for Mass Spectrometry Data.”</span> <a href="https://doi.org/10.18129/B9.bioc.QFeatures" class="external-link">https://doi.org/10.18129/B9.bioc.QFeatures</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
