<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Load Single-Cell Proteomics data using `readSCP` • scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Load Single-Cell Proteomics data using `readSCP`">
<meta property="og:description" content="scp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="read_scp_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Load Single-Cell Proteomics data using <code>readSCP</code>
</h1>
                        <h4 data-toc-skip class="author">Laurent Gatto</h4>
                                    <h4 data-toc-skip class="author">Christophe Vanderaa</h4>
                        
            <h4 data-toc-skip class="date">10 August 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/read_scp.Rmd" class="external-link"><code>vignettes/read_scp.Rmd</code></a></small>
      <div class="hidden name"><code>read_scp.Rmd</code></div>

    </div>

    
    
<div id="the-scp-data-framework" class="section level1">
<h1 class="hasAnchor">
<a href="#the-scp-data-framework" class="anchor" aria-hidden="true"></a>The <code>scp</code> data framework</h1>
<p>Our data structure is relying on two curated data classes: <code>QFeatures</code> (<span class="citation">Gatto (2020)</span>) and <code>SingleCellExperiment</code> (<span class="citation">Amezquita et al. (2019)</span>). <code>QFeatures</code> is dedicated to the manipulation and processing of MS-based quantitative data. It explicitly records the successive steps to allow users to navigate up and down the different MS levels. <code>SingleCellExperiment</code> is another class designed as an efficient data container that serves as an interface to state-of-the-art methods and algorithms for single-cell data. Our framework combines the two classes to inherit from their respective advantages.</p>
<p>Because mass spectrometry (MS)-based single-cell proteomics (SCP) only captures the proteome of between one and a few tens of single-cells in a single run, the data is usually acquired across many MS batches. Therefore, the data for each run should conceptually be stored in its own container, that we here call an <em>assay</em>. The expected input for working with the <code>scp</code> package is quantification data of peptide to spectrum matches (PSM). These data can then be processed to reconstruct peptide and protein data. The links between related features across different assays are stored to facilitate manipulation and visualization of of PSM, peptide and protein data. This is conceptually shown below.</p>
<div class="figure" style="text-align: center">
<img src="figures/SCP_framework.png" alt="`scp` relies on `SingleCellExperiment` and `QFeatures` objects." width="100%"><p class="caption">
<code>scp</code> relies on <code>SingleCellExperiment</code> and <code>QFeatures</code> objects.
</p>
</div>
<p>There are two input tables required for starting an analysis with <code>scp</code>:</p>
<ol style="list-style-type: decimal">
<li>The feature data</li>
<li>The sample data</li>
</ol>
</div>
<div id="feature-data" class="section level1">
<h1 class="hasAnchor">
<a href="#feature-data" class="anchor" aria-hidden="true"></a>Feature data</h1>
<p>The feature data are generated after the identification and quantification of the MS spectra by a pre-processing software such as MaxQuant, ProteomeDiscoverer or MSFragger (the <a href="https://en.wikipedia.org/wiki/List_of_mass_spectrometry_software" class="external-link">list</a> of available software is actually much longer). We will here use as an example a data table that has been generated by MaxQuant. The table is available from the <code>scp</code> package and is called <code>mqScpData</code> (for MaxQuant generated SCP data).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">scp</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mqScpData"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span>
<span class="co">#&gt; [1] 1361  149</span></code></pre></div>
<p>In this toy example, there are 1361 rows corresponding to features (quantified PSMs) and 149 columns corresponding to different data fields recorded by MaxQuant during the processing of the MS spectra. The columns can be divided into three categories:</p>
<ul>
<li>Columns holding feature quantifications</li>
<li>Columns holding feature metadata</li>
<li>Columns holding MS run metadata</li>
</ul>
<div id="feature-quantifications" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-quantifications" class="anchor" aria-hidden="true"></a>Feature quantifications</h3>
<p>The quantification data can be composed of one (in case of label-free acquisition) up to 16 columns (in case of TMT-16 multiplexing). The columns holding the quantification start with <code>Reporter.intensity.</code> followed by a number.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="op">(</span><span class="va">quantCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Reporter.intensity.\\d"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">)</span>,
                  value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;  [1] "Reporter.intensity.1"  "Reporter.intensity.2"  "Reporter.intensity.3" </span>
<span class="co">#&gt;  [4] "Reporter.intensity.4"  "Reporter.intensity.5"  "Reporter.intensity.6" </span>
<span class="co">#&gt;  [7] "Reporter.intensity.7"  "Reporter.intensity.8"  "Reporter.intensity.9" </span>
<span class="co">#&gt; [10] "Reporter.intensity.10" "Reporter.intensity.11" "Reporter.intensity.12"</span>
<span class="co">#&gt; [13] "Reporter.intensity.13" "Reporter.intensity.14" "Reporter.intensity.15"</span>
<span class="co">#&gt; [16] "Reporter.intensity.16"</span></code></pre></div>
<p>As you may notice, the example data was acquired using a TMT-16 protocol since we retrieve 16 quantification columns. Actually, some runs were acquired using a TMT-11 protocol (11 labels) but we will come back to this later.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="va">quantCols</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;   Reporter.intensity.1 Reporter.intensity.2 Reporter.intensity.3</span>
<span class="co">#&gt; 1                61251               501.71               3731.3</span>
<span class="co">#&gt; 2                58648              1099.80               2837.7</span>
<span class="co">#&gt; 3               150350              3705.00               9361.0</span>
<span class="co">#&gt; 4                27347               405.90               1525.2</span>
<span class="co">#&gt; 5                84035               583.09               4092.3</span>
<span class="co">#&gt; 6                44895               700.23               2283.0</span>
<span class="co">#&gt;   Reporter.intensity.4 Reporter.intensity.5 Reporter.intensity.6</span>
<span class="co">#&gt; 1              1643.30               871.84               981.87</span>
<span class="co">#&gt; 2               494.32               349.26              1030.50</span>
<span class="co">#&gt; 3                 0.00              1945.40              1188.60</span>
<span class="co">#&gt; 4                 0.00                 0.00               318.74</span>
<span class="co">#&gt; 5               530.13               718.13              2204.50</span>
<span class="co">#&gt; 6              1109.60                 0.00               675.79</span>
<span class="co">#&gt;   Reporter.intensity.7 Reporter.intensity.8 Reporter.intensity.9</span>
<span class="co">#&gt; 1              1200.10               939.06              1457.50</span>
<span class="co">#&gt; 2                 0.00              1214.10               800.58</span>
<span class="co">#&gt; 3              1574.00              2302.10              2176.10</span>
<span class="co">#&gt; 4                 0.00               519.81                 0.00</span>
<span class="co">#&gt; 5               960.51               453.77              1188.40</span>
<span class="co">#&gt; 6                 0.00               809.38               668.88</span>
<span class="co">#&gt;   Reporter.intensity.10 Reporter.intensity.11 Reporter.intensity.12</span>
<span class="co">#&gt; 1               1329.80                981.83                    NA</span>
<span class="co">#&gt; 2                807.79                391.38                    NA</span>
<span class="co">#&gt; 3               1399.50               1307.50                2192.4</span>
<span class="co">#&gt; 4                507.23                370.79                    NA</span>
<span class="co">#&gt; 5                740.99                  0.00                    NA</span>
<span class="co">#&gt; 6               1467.50                901.38                    NA</span>
<span class="co">#&gt;   Reporter.intensity.13 Reporter.intensity.14 Reporter.intensity.15</span>
<span class="co">#&gt; 1                    NA                    NA                    NA</span>
<span class="co">#&gt; 2                    NA                    NA                    NA</span>
<span class="co">#&gt; 3                1791.4                1727.5                2157.3</span>
<span class="co">#&gt; 4                    NA                    NA                    NA</span>
<span class="co">#&gt; 5                    NA                    NA                    NA</span>
<span class="co">#&gt; 6                    NA                    NA                    NA</span>
<span class="co">#&gt;   Reporter.intensity.16</span>
<span class="co">#&gt; 1                    NA</span>
<span class="co">#&gt; 2                    NA</span>
<span class="co">#&gt; 3                  1398</span>
<span class="co">#&gt; 4                    NA</span>
<span class="co">#&gt; 5                    NA</span>
<span class="co">#&gt; 6                    NA</span></code></pre></div>
</div>
<div id="feature-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#feature-metadata" class="anchor" aria-hidden="true"></a>Feature metadata</h3>
<p>Most columns in the <code>mqScpData</code> table contain information used or generated during the identification of the MS spectra. For instance, you may find the charge of the parent ion, the score and probability of a correct match between the MS spectrum and a peptide sequence, the sequence of the best matching peptide, its length, its modifications, the retention time of the peptide on the LC, the protein(s) the peptide originates from and much more.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Charge"</span>, <span class="st">"Score"</span>, <span class="st">"PEP"</span>, <span class="st">"Sequence"</span>, <span class="st">"Length"</span>,
                   <span class="st">"Retention.time"</span>, <span class="st">"Proteins"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;   Charge  Score        PEP    Sequence Length Retention.time</span>
<span class="co">#&gt; 1      2 41.029 5.2636e-04   ATNFLAHEK      9         65.781</span>
<span class="co">#&gt; 2      2 44.349 5.8789e-04   ATNFLAHEK      9         63.787</span>
<span class="co">#&gt; 3      2 51.066 4.0315e-24 SHTILLVQPTK     11         71.884</span>
<span class="co">#&gt; 4      2 63.816 4.7622e-06 SHTILLVQPTK     11         68.633</span>
<span class="co">#&gt; 5      2 74.464 6.8709e-09 SHTILLVQPTK     11         71.946</span>
<span class="co">#&gt; 6      2 41.502 5.3705e-02     SLVIPEK      7         76.204</span>
<span class="co">#&gt;               Proteins</span>
<span class="co">#&gt; 1 sp|P29692|EF1D_HUMAN</span>
<span class="co">#&gt; 2 sp|P29692|EF1D_HUMAN</span>
<span class="co">#&gt; 3  sp|P84090|ERH_HUMAN</span>
<span class="co">#&gt; 4  sp|P84090|ERH_HUMAN</span>
<span class="co">#&gt; 5  sp|P84090|ERH_HUMAN</span>
<span class="co">#&gt; 6 sp|P62269|RS18_HUMAN</span></code></pre></div>
</div>
<div id="ms-run-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#ms-run-metadata" class="anchor" aria-hidden="true"></a>MS run metadata</h3>
<p>This type of metadata is related to the MS instrument. In MaxQuant, only the file name generated by the MS instrument is stored. There is one file for each MS run, hence the file name can be used as a batch identifier.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mqScpData</span><span class="op">$</span><span class="va">Raw.file</span><span class="op">)</span>
<span class="co">#&gt; [1] "190321S_LCA10_X_FP97AG"        "190222S_LCA9_X_FP94BM"        </span>
<span class="co">#&gt; [3] "190914S_LCB3_X_16plex_Set_21"  "190321S_LCA10_X_FP97_blank_01"</span></code></pre></div>
</div>
</div>
<div id="sample-data" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-data" class="anchor" aria-hidden="true"></a>Sample data</h1>
<p>Next to the quantification data and the feature data, sample metadata contains the experimental design generated by the researcher. The rows of sample metadata correspond to a sample in the experiment and the columns correspond to the available information about the sample. We will here use the second example table:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sampleAnnotation"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sampleAnnotation</span><span class="op">)</span>
<span class="co">#&gt;                Raw.file              Channel SampleType lcbatch sortday digest</span>
<span class="co">#&gt; 1 190222S_LCA9_X_FP94BM Reporter.intensity.1    Carrier    LCA9      s8      N</span>
<span class="co">#&gt; 2 190222S_LCA9_X_FP94BM Reporter.intensity.2  Reference    LCA9      s8      N</span>
<span class="co">#&gt; 3 190222S_LCA9_X_FP94BM Reporter.intensity.3     Unused    LCA9      s8      N</span>
<span class="co">#&gt; 4 190222S_LCA9_X_FP94BM Reporter.intensity.4   Monocyte    LCA9      s8      N</span>
<span class="co">#&gt; 5 190222S_LCA9_X_FP94BM Reporter.intensity.5      Blank    LCA9      s8      N</span>
<span class="co">#&gt; 6 190222S_LCA9_X_FP94BM Reporter.intensity.6   Monocyte    LCA9      s8      N</span></code></pre></div>
<p>This table may contain any information about the samples. For example, useful information could be the type of sample that is analysed, a phenotype known from the experimental design, the MS batch, the acquisition date, MS settings used to acquire the sample, the LC batch, the sample preparation batch, etc… However, <code>scp</code> <strong>requires</strong> 2 specific fields in the sample metadata:</p>
<ol style="list-style-type: decimal">
<li>One column containing the MS run names (<code>Raw.file</code> in this case). It must have the same name as the name of the column containing the MS run names in the quantification table. This will allow <code>scp</code> to correctly match and split data that were acquired separately. This is illustrated below by the linked circles.</li>
<li>One column that tells <code>scp</code> the names of the columns in the feature data holds the quantification of the corresponding sample. This is illustrated below by the arrow.</li>
</ol>
<div class="figure">
<img src="figures/readSCP_relation.png" alt="Linking the sample data to the feature data" width="196"><p class="caption">
Linking the sample data to the feature data
</p>
</div>
</div>
<div id="readscp" class="section level1">
<h1 class="hasAnchor">
<a href="#readscp" class="anchor" aria-hidden="true"></a><code>readSCP</code>
</h1>
<p><code>readSCP</code> is the function that converts the sample and the feature data into a <code>QFeatures</code> object following the data structure described above, that is storing the data belonging to each MS batch in a separate <code>SingleCellExperiment</code> object. We therefore provide the feature data, the sample data to the function as well as the name of the column that holds the batch name in both tables and the name of the column in the sample data that points to the quantification columns in the feature data.</p>
<div id="sample-names-and-suffix" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-names-and-suffix" class="anchor" aria-hidden="true"></a>Sample names and <code>suffix</code>
</h2>
<p><code>readSCP</code> automatically assigns names that are unique across all samples in all assays. This is performed by appending the name of the batch where a given sample is found in with the name of the quantification column for that sample. Suppose a sample belongs to batch <code>190222S_LCA9_X_FP94BM</code> and the quantification values in the feature data are found in the column called <code>Reporter.intensity.3</code>, then the sample name will become <code>190222S_LCA9_X_FP94BMReporter.intensity.3</code>. Optionally, to improve the readability of sample names, <code>readSCP</code> can take a suffix instead of the quantification column name. For instance, in the example below, we will provide a short suffix with the TMT index to remind that samples were multiplexed using TMT.</p>
</div>
<div id="special-case-empty-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#special-case-empty-samples" class="anchor" aria-hidden="true"></a>Special case: empty samples</h2>
<p>In some rare cases, it can be beneficial to remove empty samples (all quantifications are <code>NA</code>) from the assays. Such samples can occur when samples that were acquired with different multiplexing labels are merged in a single table. For instance, the SCoPE2 data we provide as an example contains runs that were acquired with two TMT protocols. The 3 first assays were acquired using the TMT-11 protocol and the last assay was acquired using a TMT-16 protocol. When exporting the table, the authors combined the data in a single table, were missing channels in the TMT-11 data are filled with <code>NA</code>. This is essential when working in table format, but since <code>scp</code> keeps the runs separated we can allow for different numbers of channels per run. When setting <code>removeEmptyCols = TRUE</code>, <code>readSCP</code> automatically detects and removes columns that contain only <code>NA</code>s,</p>
</div>
<div id="running-readscp" class="section level2">
<h2 class="hasAnchor">
<a href="#running-readscp" class="anchor" aria-hidden="true"></a>Running <code>readSCP</code>
</h2>
<p>We convert the sample and the feature data into a <code>QFeatures</code> object in a single command thanks to <code>readSCP</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCP.html">readSCP</a></span><span class="op">(</span>featureData <span class="op">=</span> <span class="va">mqScpData</span>,
               colData <span class="op">=</span> <span class="va">sampleAnnotation</span>,
               batchCol <span class="op">=</span> <span class="st">"Raw.file"</span>,
               channelCol <span class="op">=</span> <span class="st">"Channel"</span>,
               suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"_TMT"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>,
               removeEmptyCols <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Loading data as a 'SingleCellExperiment' object</span>
<span class="co">#&gt; Splitting data based on 'Raw.file'</span>
<span class="co">#&gt; Formatting sample metadata (colData)</span>
<span class="co">#&gt; Formatting data as a 'QFeatures' object</span></code></pre></div>
<p>As indicated by the output on the console, <code>readSCP</code> proceeds as follows:</p>
<ol style="list-style-type: decimal">
<li><p>If <code>featureData</code> is the path to a CSV file, it reads the file using <code>read.csv</code>. The table is converted to a <code>SingleCellExperiment</code> object. <code>readSCP</code> needs to know in which field(s) the quantitative data is stored. Those field name(s) is/are provided by the <code>channelCol</code> field in the <code>metaData</code> table. So in this example, the column names holding the quantitative data in <code>mqScpData</code> are stored in the column named <code>Channel</code> in <code>sampleAnnotation</code>.</p></li>
<li><p>The <code>SingleCellExperiment</code> object is then split according to batch. The split is performed depending on the <code>batchCol</code> field in <code>featureData</code>, in this case the data will be split according to the <code>Raw.file</code> column in <code>mqScpData</code>. <code>Raw.file</code> contains the names of the acquisition runs that was used by MaxQuant to retrieve the raw data files.</p></li>
<li><p>The sample metadata is generated from the supplied <code>colData</code>. Note that in order for <code>readSCP</code> to correctly match the feature data with the metadata, <code>colData</code> must contain the same <code>batchCol</code> field with batch names.</p></li>
<li><p>Finally, the split feature data and the sample metadata are stored in a single <code>QFeatures</code> object.</p></li>
</ol>
<p>Here is a compact overview of the data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scp</span>
<span class="co">#&gt; An instance of class QFeatures containing 4 assays:</span>
<span class="co">#&gt;  [1] 190222S_LCA9_X_FP94BM: SingleCellExperiment with 395 rows and 11 columns </span>
<span class="co">#&gt;  [2] 190321S_LCA10_X_FP97_blank_01: SingleCellExperiment with 109 rows and 11 columns </span>
<span class="co">#&gt;  [3] 190321S_LCA10_X_FP97AG: SingleCellExperiment with 487 rows and 11 columns </span>
<span class="co">#&gt;  [4] 190914S_LCB3_X_16plex_Set_21: SingleCellExperiment with 370 rows and 16 columns</span></code></pre></div>
<p>We can see that the object returned by <code>readSCP</code> is a <code>QFeatures</code> object containing 4 <code>SingleCellExperiment</code> assays that have been named after the 4 MS batches. Each assay contains either 11 or 16 columns (samples) depending on the TMT labelling strategy and a variable number of rows (quantified PSMs). Each piece of information can easily be retrieved thanks to the <code>Qfeatures</code> architectures. As mentioned in the previous vignette, sample data is retrieved from the <code>colData</code>. Note that unique sample names were automatically generated by combining the batch name and the suffix provided to <code>readSCP</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">scp</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 6 rows and 6 columns</span>
<span class="co">#&gt;                                 Raw.file       Channel  SampleType     lcbatch</span>
<span class="co">#&gt;                              &lt;character&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT1 190222S_LC... Reporter.i...     Carrier        LCA9</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT2 190222S_LC... Reporter.i...   Reference        LCA9</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT3 190222S_LC... Reporter.i...      Unused        LCA9</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT4 190222S_LC... Reporter.i...    Monocyte        LCA9</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT5 190222S_LC... Reporter.i...       Blank        LCA9</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT6 190222S_LC... Reporter.i...    Monocyte        LCA9</span>
<span class="co">#&gt;                                sortday      digest</span>
<span class="co">#&gt;                            &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT1          s8           N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT2          s8           N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT3          s8           N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT4          s8           N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT5          s8           N</span>
<span class="co">#&gt; 190222S_LCA9_X_FP94BM_TMT6          s8           N</span></code></pre></div>
<p>Notice that the sample names were suffixed with TMT indexes.</p>
<p>The feature metadata is retrieved from the <code>rowData</code>. Since the feature metadata is specific to each assay, we need to tell from which assay we want to get the <code>rowData</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">scp</span><span class="op">[[</span><span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt; DataFrame with 6 rows and 5 columns</span>
<span class="co">#&gt;                 uid      Sequence    Length Modifications Modified.sequence</span>
<span class="co">#&gt;         &lt;character&gt;   &lt;character&gt; &lt;integer&gt;   &lt;character&gt;       &lt;character&gt;</span>
<span class="co">#&gt; PSM2  _(Acetyl (...     ATNFLAHEK         9 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">#&gt; PSM4  _(Acetyl (... SHTILLVQPT...        11 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">#&gt; PSM6  _(Acetyl (...       SLVIPEK         7 Acetyl (Pr...     _(Acetyl (...</span>
<span class="co">#&gt; PSM9  _AAGLALK_ ...       AAGLALK         7    Unmodified         _AAGLALK_</span>
<span class="co">#&gt; PSM12 _AALSAGK_ ...       AALSAGK         7    Unmodified         _AALSAGK_</span>
<span class="co">#&gt; PSM15 _AAPEASGTP... AAPEASGTPS...        16    Unmodified     _AAPEASGTP...</span></code></pre></div>
<p>Finally, we can also retrieve the quantification matrix for an assay of interest:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">scp</span>, <span class="st">"190222S_LCA9_X_FP94BM"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT1 190222S_LCA9_X_FP94BM_TMT2</span>
<span class="co">#&gt; PSM2                     58648.0                    1099.80</span>
<span class="co">#&gt; PSM4                     27347.0                     405.90</span>
<span class="co">#&gt; PSM6                     44895.0                     700.23</span>
<span class="co">#&gt; PSM9                    122070.0                    1153.50</span>
<span class="co">#&gt; PSM12                    58605.0                     895.25</span>
<span class="co">#&gt; PSM15                     5006.5                     517.86</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT3 190222S_LCA9_X_FP94BM_TMT4</span>
<span class="co">#&gt; PSM2                     2837.70                     494.32</span>
<span class="co">#&gt; PSM4                     1525.20                       0.00</span>
<span class="co">#&gt; PSM6                     2283.00                    1109.60</span>
<span class="co">#&gt; PSM9                     7361.90                    1732.30</span>
<span class="co">#&gt; PSM12                    2763.80                     867.82</span>
<span class="co">#&gt; PSM15                     446.19                     458.17</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT5 190222S_LCA9_X_FP94BM_TMT6</span>
<span class="co">#&gt; PSM2                      349.26                    1030.50</span>
<span class="co">#&gt; PSM4                        0.00                     318.74</span>
<span class="co">#&gt; PSM6                        0.00                     675.79</span>
<span class="co">#&gt; PSM9                     1515.60                    2252.00</span>
<span class="co">#&gt; PSM12                    1050.30                    1268.70</span>
<span class="co">#&gt; PSM15                     467.90                     649.50</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT7 190222S_LCA9_X_FP94BM_TMT8</span>
<span class="co">#&gt; PSM2                        0.00                    1214.10</span>
<span class="co">#&gt; PSM4                        0.00                     519.81</span>
<span class="co">#&gt; PSM6                        0.00                     809.38</span>
<span class="co">#&gt; PSM9                      444.48                    2343.80</span>
<span class="co">#&gt; PSM12                     532.30                    1073.10</span>
<span class="co">#&gt; PSM15                     259.84                     533.55</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT9 190222S_LCA9_X_FP94BM_TMT10</span>
<span class="co">#&gt; PSM2                      800.58                      807.79</span>
<span class="co">#&gt; PSM4                        0.00                      507.23</span>
<span class="co">#&gt; PSM6                      668.88                     1467.50</span>
<span class="co">#&gt; PSM9                     3100.20                     1825.20</span>
<span class="co">#&gt; PSM12                     911.30                     1300.00</span>
<span class="co">#&gt; PSM15                     393.53                      463.26</span>
<span class="co">#&gt;       190222S_LCA9_X_FP94BM_TMT11</span>
<span class="co">#&gt; PSM2                       391.38</span>
<span class="co">#&gt; PSM4                       370.79</span>
<span class="co">#&gt; PSM6                       901.38</span>
<span class="co">#&gt; PSM9                      2372.50</span>
<span class="co">#&gt; PSM12                     1185.90</span>
<span class="co">#&gt; PSM15                      353.04</span></code></pre></div>
</div>
</div>
<div id="what-about-label-free-scp" class="section level1">
<h1 class="hasAnchor">
<a href="#what-about-label-free-scp" class="anchor" aria-hidden="true"></a>What about label-free SCP?</h1>
<p>The <code>scp</code> package is meant for both label-free and multiplexed SCP data. Like in the example above, the label-free data should contain the batch names in both the feature data and the sample data. The sample data must also contain a column that points to the columns of the feature data that contains the quantifications. Since label-free SCP acquires one single-cell per run, this sample data column should point the same column for all samples. Moreover, this means that each PSM assay will contain a single column.</p>
</div>
<div id="what-about-other-input-formats" class="section level1">
<h1 class="hasAnchor">
<a href="#what-about-other-input-formats" class="anchor" aria-hidden="true"></a>What about other input formats?</h1>
<p><code>readSCP</code> should work with any PSM quantification table that is output by a pre-processing software. For instance, you can easily import the PSM tables generated by ProteomeDiscoverer. The batch names are contained in the <code>File ID</code> column (that should be supplied as the <code>batchCol</code> argument to <code>readSCP</code>). The quantification columns are contained in the columns starting with <code>Abundance</code>, eventually followed by a multiplexing tag name. These columns should be stored in a dedicated column of the sample data to be supplied as <code>channelCol</code> to <code>readSCP</code>.</p>
<p>If your input cannot be loaded using the procedure described in this vignette, you can submit a feature request (see next section).</p>
</div>
<div id="need-help" class="section level1">
<h1 class="hasAnchor">
<a href="#need-help" class="anchor" aria-hidden="true"></a>Need help?</h1>
<p>You can open an issue on the <a href="https://github.com/UCLouvain-CBIO/scp/issues" class="external-link">GitHub repository</a> in case of troubles when loading your SCP data with <code>readSCP</code>. Any suggestion or feature request about the function or the documentation are also warmly welcome.</p>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor" aria-hidden="true"></a>Session information</h1>
<pre><code>R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scp_1.3.3                   QFeatures_1.3.7            
 [3] MultiAssayExperiment_1.19.5 SummarizedExperiment_1.23.1
 [5] Biobase_2.53.0              GenomicRanges_1.45.0       
 [7] GenomeInfoDb_1.29.3         IRanges_2.27.0             
 [9] S4Vectors_0.31.0            BiocGenerics_0.39.1        
[11] MatrixGenerics_1.5.3        matrixStats_0.60.0         
[13] BiocStyle_2.21.3           

loaded via a namespace (and not attached):
 [1] lattice_0.20-44             png_0.1-7                  
 [3] utf8_1.2.2                  rprojroot_2.0.2            
 [5] digest_0.6.27               SingleCellExperiment_1.15.1
 [7] R6_2.5.0                    evaluate_0.14              
 [9] highr_0.9                   pillar_1.6.2               
[11] zlibbioc_1.39.0             rlang_0.4.11               
[13] lazyeval_0.2.2              jquerylib_0.1.4            
[15] Matrix_1.3-4                rmarkdown_2.10             
[17] pkgdown_1.6.1.9001          textshaping_0.3.5          
[19] desc_1.3.0                  stringr_1.4.0              
[21] ProtGenerics_1.25.1         igraph_1.2.6               
[23] RCurl_1.98-1.3              DelayedArray_0.19.1        
[25] compiler_4.1.0              xfun_0.25                  
[27] pkgconfig_2.0.3             systemfonts_1.0.2          
[29] htmltools_0.5.1.1           tidyselect_1.1.1           
[31] tibble_3.1.3                GenomeInfoDbData_1.2.6     
[33] bookdown_0.22               fansi_0.5.0                
[35] crayon_1.4.1                dplyr_1.0.7                
[37] MASS_7.3-54                 bitops_1.0-7               
[39] grid_4.1.0                  jsonlite_1.7.2             
[41] lifecycle_1.0.0             AnnotationFilter_1.17.1    
[43] magrittr_2.0.1              MsCoreUtils_1.5.0          
[45] stringi_1.7.3               cachem_1.0.5               
[47] XVector_0.33.0              fs_1.5.0                   
[49] bslib_0.2.5.1               ellipsis_0.3.2             
[51] ragg_1.1.3                  vctrs_0.3.8                
[53] generics_0.1.0              tools_4.1.0                
[55] glue_1.4.2                  purrr_0.3.4                
[57] fastmap_1.1.0               yaml_2.2.1                 
[59] clue_0.3-59                 cluster_2.1.2              
[61] BiocManager_1.30.16         memoise_2.0.0              
[63] knitr_1.33                  sass_0.4.0                 </code></pre>
</div>
<div id="reference" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#reference" class="anchor" aria-hidden="true"></a>Reference</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Amezquita2019-bf" class="csl-entry">
Amezquita, Robert A, Aaron T L Lun, Etienne Becht, Vince J Carey, Lindsay N Carpp, Ludwig Geistlinger, Federico Martini, et al. 2019. <span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span> <em>Nat. Methods</em>, December, 1–9.
</div>
<div id="ref-Gatto2020-ry" class="csl-entry">
Gatto, Laurent. 2020. <span>“<span>QFeatures</span>: Quantitative Features for Mass Spectrometry Data.”</span>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
