<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell Proteomics data processing and analysis. • scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell Proteomics data processing and analysis.">
<meta property="og:description" content="scp">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scp_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single Cell Proteomics data processing and analysis.</h1>
                        <h4 class="author">Laurent Gatto</h4>
                                    <h4 class="author">Christophe Vanderaa</h4>
                        
            <h4 class="date">21 August 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/scp.Rmd"><code>vignettes/scp.Rmd</code></a></small>
      <div class="hidden name"><code>scp.Rmd</code></div>

    </div>

    
    
<div id="the-scp-package" class="section level1">
<h1 class="hasAnchor">
<a href="#the-scp-package" class="anchor"></a>The <code>scp</code> package</h1>
<p>The <code>scp</code> package is used to process and analyse mass spectrometry-based single cell proteomics (SCP) data. It relies on the <a href="https://rformassspectrometry.github.io/QFeatures/"><code>QFeatures</code></a> (<span class="citation">@QFeatures</span>) package to manage and process <a href="http://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> (<span class="citation">@sce</span>) objects.</p>
<p>This vignette will guide you through the common steps of mass spectrometry-based single-cell proteomics data analysis. To start, we load the <code>scp</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st"><a href="https://UCLouvain-CBIO.github.io/scp">"scp"</a></span>)
</pre></div>
<p>We also load <code>ggplot2</code> and <code>magrittr</code> for convenient data manipulation and plotting, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st"><a href="http://ggplot2.tidyverse.org">"ggplot2"</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"magrittr"</span>)
</pre></div>
</div>
<div id="read-in-scp-data" class="section level1">
<h1 class="hasAnchor">
<a href="#read-in-scp-data" class="anchor"></a>Read in SCP data</h1>
<p>The workflow starts with reading in the tabular quantification data generated by, for example, MaxQuant. We created a small example data by subsetting the MaxQuant table provided in the SCoPE2 preprint (<span class="citation">@Specht2019-jm</span>). The <code>mqFile</code> table is a typical example of what you would get after reading in a CSV file using <code>read.csv</code> or <code>read.table</code>. See <code><a href="../reference/mqFile.html">?mqFile</a></code> for more information about the table content.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"mqFile"</span>)
</pre></div>
<p>In order to convert this tabular data to a <code>scp</code>-compatible <code>QFeatures</code> object, we need to provide a metadata table. The rows contain sample information and must contain:</p>
<ul>
<li>The name of the batch the sample was acquired in</li>
<li>The name of the channel the sample was acquired in</li>
</ul>
<p>Any additional information will be stored in the <code>colData</code>.</p>
<p>We provide an example of such a data frame. It was formatted from the annotation table provided in the SCoPE2 preprint. See <code><a href="../reference/sampleAnnotation.html">?sampleAnnotation</a></code> for more information about the table content.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"sampleAnnotation"</span>)
</pre></div>
<p>The two tables are supplied to the <code>readSCP</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCP.html">readSCP</a></span>(quantTable = <span class="kw">mqFile</span>,
               metaTable = <span class="kw">sampleAnnotation</span>,
               channelCol = <span class="st">"Channel"</span>,
               batchCol = <span class="st">"Set"</span>)
</pre></div>
<p>As indicated by the output on the console, <code>readSCP</code> proceeds as follows:</p>
<ol style="list-style-type: decimal">
<li>If <code>quantTable</code> is the path to a CSV file, it reads the file using <code>read.csv</code>. The table is converted to a <code>SingleCellExperimentObject</code> object. <code>readSCP</code> needs to know in which field(s) the quantative data is stored. Those field name(s) is/are provided by the <code>channelCol</code> field in the <code>metaData</code> table.</li>
<li>The <code>SingleCellExperiment</code> object is then split according to batch. The split is performed depending on the <code>batchCol</code> field in <code>quantTable</code>.</li>
<li>The sample metadata is generated from the <code>metaTable</code>. Note that in order for <code>readSCP</code> to correctly match the feature data with the metadata, <code>metaTable</code> should also contain the <code>batchCol</code> field with batch names.</li>
<li>Finally, the split feature data and the sample metadata are stored in a single <code>QFeatures</code> object.</li>
</ol>
<p>Here is a compact overview of the data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw">scp</span>
</pre></div>
<p>We can see that the <code>scp</code> object we create is a <code>QFeatures</code> object containing 3 assays. Each assay has an associated name, this is the batch name that was used for splitting. We can also see that each assay is a <code>SingleCellExperiment</code> object. The rows represent the peptide to spectrum matches (PSMs), the number vary depending on the batch. Finally, all three assays contains 16 columns that correspond to the 16 TMT channels recorded during the 3 MS runs.</p>
</div>
<div id="clean-missing-data" class="section level1">
<h1 class="hasAnchor">
<a href="#clean-missing-data" class="anchor"></a>Clean missing data</h1>
<p>Single-cell (proteomics or transcriptomics) data contains many zeros. The zeros can be biological zeros or technical zeros. To avoid artefacts in dowstream steps, we replace the zeros by the missing value <code>NA</code>. The <code>zeroIsNA</code> function takes the <code>QFeatures</code> object and the name(s) or index/indices of the assay(s) to clean.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">zeroIsNA</span>(<span class="kw">scp</span>, i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>)
</pre></div>
</div>
<div id="filter-psms" class="section level1">
<h1 class="hasAnchor">
<a href="#filter-psms" class="anchor"></a>Filter PSMs</h1>
<p>A common steps in SCP is to filter out low-confidence PSMs. Each PSM assay contains feature meta-information that are stored in the assay’s <code>rowData</code>. The <code>QFeatures</code> package allows to quickly filter the rows of an assay by using these information. The available variables in the <code>rowData</code> are listed below for each assay.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu">rowDataNames</span>(<span class="kw">scp</span>)
</pre></div>
<div id="filter-features-based-on-feature-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-features-based-on-feature-metadata" class="anchor"></a>Filter features based on feature metadata</h2>
<p>Here are some examples of criteria that are used to identify low-confidence. The information is readily available since this was computed by MaxQuant:</p>
<ul>
<li>Remove PSMs that are matched to contaminants</li>
<li>Remove PSMs that are matched to the decoy database</li>
<li>Keep PSMs that exhibit a high PIF (parental ion fraction), indicative of the purity of a spectrum</li>
</ul>
<p>We can perform this filtering using the <code>filterFeatures</code> function. <code>filterFeatures</code> automatically accesses the feature metadata and selects the rows that meet the provided condition(s). For instance, <code>Reverse != "+"</code> keeps the rows for which the <code>Reverse</code> variable in the <code>rowData</code> is not <code>"+"</code> (<em>i.e.</em> the PSM is not matched to the decoy database).</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">filterFeatures</span>(<span class="kw">scp</span>, 
                      <span class="op">~</span> <span class="kw">Reverse</span> != <span class="st">"+"</span> <span class="op">&amp;</span>
                        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"REV|CON"</span>, <span class="kw">protein</span>) <span class="op">&amp;</span>
                        <span class="kw">Potential.contaminant</span> != <span class="st">"+"</span> <span class="op">&amp;</span>
                        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">PIF</span>) <span class="op">&amp;</span> <span class="kw">PIF</span> <span class="op">&gt;</span> <span class="fl">0.8</span>)
</pre></div>
</div>
<div id="filter-assays-based-on-detected-features" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-assays-based-on-detected-features" class="anchor"></a>Filter assays based on detected features</h2>
<p>To avoid proceedin with failed runs, another interesting filter is to remove assays with too few features. If a batch contains less than, for example, 150 features, then we can suspect something wrong happened in that batch and it should be removed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu">dims</span>(<span class="kw">scp</span>)
</pre></div>
<p>A <code>QFeatures</code> object can be seen as a three-order array: <span class="math inline">\(features \times samples \times assay\)</span>. Hence, <code>QFeatures</code> support three-order subsetting <code>x[rows, columns, assays]</code>. We first select the assay that have sufficient PSMs, and then subset the <code>scp</code> object for the assays that meet the criterion.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">keepAssay</span> <span class="op">&lt;-</span> <span class="fu">dims</span>(<span class="kw">scp</span>)[<span class="fl">1</span>, ] <span class="op">&gt;</span> <span class="fl">150</span>
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="kw">scp</span>[, , <span class="kw">keepAssay</span>]
<span class="kw">scp</span>
</pre></div>
<p>Note in our example, all assays have sufficient number of PSMs, hence none are removed.</p>
</div>
<div id="filter-features-based-on-scp-metrics" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-features-based-on-scp-metrics" class="anchor"></a>Filter features based on SCP metrics</h2>
<p>The third type of filtering is specific to SCP. For instance, the SCoPE2 script filters features based on the sample to carrier ratio (SCR), that is the reporter ion intensity of a single-cell sample divided by the reporter ion intensity of the carrier channel (200 cells) from the same batch. It is expected that the carrier intensities are much higher than the single-cell intensities.</p>
<p>The SCR can be computed using the <code>computeSCR</code> function from <code>scp</code>. The function must be told which channels are the samples that must be divided and which channel contains the carrier. This information is provided in the sample metadata and is accessed using the <code>colData</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu">colData</span>(<span class="kw">scp</span>)[, <span class="st">"SampleAnnotation"</span>] <span class="op">%&gt;%</span>
  <span class="kw">table</span>
</pre></div>
<p>In this dataset, <code>SampleType</code> gives the type of sample that is present in each TMT channel. The SCoPE2 protocole includes 5 types of samples: * The carrier channels (<code>carrier_mix</code>) contain 200 cell equivalents and are meant to boost the peptide identification rate. * The normalization channels (<code>norm</code> and <code>reference</code>) contain 5 cell equivalents and are used to partially correct for between-run variation. * The unused channels (<code>unused</code>) are channels that are left empty due to isotopic cross-contamination by the carrier channel. * The blank channels (<code>sc_0</code>) contain samples that do not contain any cell but are processed as single-cell samples. * The single-cell sample channels contain the single-cell samples of interest, that are macrophage (<code>sc_m0</code>) or monocyte (<code>sc_u</code>).</p>
<p>The <code>computeSCR</code> function expects the following input:</p>
<ul>
<li>The <code>QFeatures</code> dataset</li>
<li>The assay name(s) or index/indices for which the SCR should be computed</li>
<li>The sample metadata variable pointing to the channel annotation</li>
<li>A string pattern (following regular expression syntax) that uniquely identifies the carrier channel in each batch</li>
<li>A string pattern (following regular expression syntax) that identifies the samples to divide</li>
</ul>
<p>The function creates an field <code>meanSCR</code> and stores it in the <code>rowData</code> of each assay. Note that <code>^sc</code> means <em>starts with <code>sc</code></em>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeSCR.html">computeSCR</a></span>(<span class="kw">scp</span>,
                  i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,
                  colDataCol = <span class="st">"SampleAnnotation"</span>,
                  carrierPattern = <span class="st">"carrier"</span>,
                  samplePattern = <span class="st">"^sc"</span>)
</pre></div>
<p>Before applying the filter, we plot the distribution of the mean SCR. We can extract the <code>meanSCR</code> variable from the <code>rowData</code> of several assays using the <code>rowDataToDF</code>. It takes the <code>rowData</code> field(s) of interest and returns a <code>DataFrame</code> table.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">scp</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/rowDataToDF.html">rowDataToDF</a></span>(i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, 
              vars = <span class="st">"meanSCR"</span>) <span class="op">%&gt;%</span>
  <span class="kw">data.frame</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">meanSCR</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(xintercept = <span class="fl">0.1</span>)
</pre></div>
<p>We can see that most values are close to 0.02 as expected since the experimental ratio is 1/200. There are a few point that have higher signal than expected. We therefore filter those point using a cutoff of 0.1 using again the <code>filterFeatures</code> functions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">filterFeatures</span>(<span class="kw">scp</span>, 
                      <span class="op">~</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">meanSCR</span>) <span class="op">&amp;</span> 
                        <span class="kw">meanSCR</span> <span class="op">&lt;</span> <span class="fl">0.1</span>)
</pre></div>
</div>
<div id="filter-out-psms-with-high-false-discovery-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-out-psms-with-high-false-discovery-rate" class="anchor"></a>Filter out PSMs with high false discovery rate</h2>
<p>Finally, the last PSM filter criterion is the false discovery rate (FDR) for identification. Filtering on the PEP is too conservative (<span class="citation">@Kall2008-hb</span>) so we provide the <code>computeFDR</code> function to convert PEPs to FDR. Beside the dataset and the assay(s) for which to compute the FDR, we also need to give the feature grouping variable, here the peptide sequence, and the variable containing the PEPs. Those are contained in the feature metadata.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeFDR.html">computeFDR</a></span>(<span class="kw">scp</span>, 
                  i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, 
                  groupCol = <span class="st">"peptide"</span>,
                  pepCol = <span class="st">"dart_PEP"</span>)
</pre></div>
<p>Note that a new variable <code>.FDR</code> containing the computed FDRs is added to the <code>rowData</code>. We filter the PSMs that have an associated peptide FDR smaller than 1%.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">filterFeatures</span>(<span class="kw">scp</span>, 
                      <span class="op">~</span> <span class="kw">.FDR</span> <span class="op">&lt;</span> <span class="fl">0.01</span>)
</pre></div>
</div>
</div>
<div id="process-the-psm-data" class="section level1">
<h1 class="hasAnchor">
<a href="#process-the-psm-data" class="anchor"></a>Process the PSM data</h1>
<div id="relative-reporter-ion-intensity" class="section level2">
<h2 class="hasAnchor">
<a href="#relative-reporter-ion-intensity" class="anchor"></a>Relative reporter ion intensity</h2>
<p>In order to partialy correct for between-run variation, SCoPE2 suggests computing relative reporter ion intensities. This means that intensities measured for single-cells are divided by the reference channel containing 5-cell equivalents. We use the <code>divideByReference</code> function that divides channels of interest by the reference channel. Similarly to <code>computeSCR</code>, we can point to the samples and the reference columns in each assay using the annotation contained in the <code>colData</code>.</p>
<p>We here divide all columns (using the regular expression wildcard <code>.</code>) by the reference channel (either <code>reference</code> or <code>norm</code>).</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/divideByReference.html">divideByReference</a></span>(<span class="kw">scp</span>,
                         i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, 
                         colDataCol = <span class="st">"SampleAnnotation"</span>, 
                         samplePattern = <span class="st">"."</span>, 
                         refPattern = <span class="st">"reference|norm"</span>)
</pre></div>
</div>
</div>
<div id="aggregate-psm-data-to-peptide-data" class="section level1">
<h1 class="hasAnchor">
<a href="#aggregate-psm-data-to-peptide-data" class="anchor"></a>Aggregate PSM data to peptide data</h1>
<p>Now that the PSM assays are processed, we can aggregate them to peptides. This is performed using the <code>aggregateFeaturesOverAssays</code> function. For each assay, the function aggregates several PSMs into a unique peptide. This is best illustrated by the figure below.</p>
<p>The PSMs are aggregated over the <code>fcol</code> features variable, here peptides. We also need to supply an aggregating function that will tell how to combine the quantitative data of the PSMs to aggregate. We here use the median value. We name the aggregated assays using the original names and appending <code>peptides_</code> at the start.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span><span class="fu"><a href="../reference/aggregateFeaturesOverAssays.html">aggregateFeaturesOverAssays</a></span>(<span class="kw">scp</span>,
                                  i = <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, 
                                  fcol = <span class="st">"peptide"</span>, 
                                  name = <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"peptides_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">scp</span>)),
                                  fun = <span class="kw">matrixStats</span>::<span class="kw"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html">colMedians</a></span>, na.rm = <span class="fl">TRUE</span>)
</pre></div>
<p>Under the hood, the <code>QFeatures</code> architecture preserves the relationship between the aggregated assays. See <code>?AssayLinks</code> for more information on relationships between assays. Three new assays were added to the dataset:</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="kw">scp</span>
</pre></div>
</div>
<div id="join-the-scope2-sets-in-one-assay" class="section level1">
<h1 class="hasAnchor">
<a href="#join-the-scope2-sets-in-one-assay" class="anchor"></a>Join the SCoPE2 sets in one assay</h1>
<p>Up to now, we kept the data belonging to each MS run in separate assays. We now combine all batches into a single assay. This is done using the <code>joinAssays</code> function from the <code>QFeatures</code> package. Note that we now use the aggregated assays, so assay 4 to 6.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">joinAssays</span>(<span class="kw">scp</span>, 
                  i = <span class="fl">4</span><span class="op">:</span><span class="fl">6</span>, 
                  name = <span class="st">"peptides"</span>)
</pre></div>
</div>
<div id="filter-single-cells" class="section level1">
<h1 class="hasAnchor">
<a href="#filter-single-cells" class="anchor"></a>Filter single-cells</h1>
<p>Another common step in single-cell data analysis pipelines is to remove low-quality cells. After subseting for the samples of interest, we will use 2 metrics: the median relative intensities per cell and the median coefficient of variation (CV) per cell.</p>
<div id="filter-samples-of-interest" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-samples-of-interest" class="anchor"></a>Filter samples of interest</h2>
<p>We first subset the cells of interest, that is the blank samples (<code>sc_0</code>), the macrophages (<code>sc_m0</code>) and the monocytes (<code>sc_u</code>).</p>
<p>We extract the <code>SingleCellExperiment</code> assay with the joined peptides, subset the channels corresponding to blank, macrophage or monocytes and add it as a new assay in the <code>QFeatures</code> object. However, the sample annotation is contained in the <code>colData</code> of the <code>QFeatures</code> dataset, but we need to access it from the <code>SingeCellExperiment</code> object. Therefore, we provide the <code>transferColDataToAssay</code> to copy the sample metadata from the <code>QFeatures</code> to a target assay.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="fu">colData</span>(<span class="kw">scp</span>[[<span class="st">"peptides"</span>]])
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transferColDataToAssay.html">transferColDataToAssay</a></span>(<span class="kw">scp</span>, <span class="st">"peptides"</span>)
<span class="fu">colData</span>(<span class="kw">scp</span>[[<span class="st">"peptides"</span>]])
</pre></div>
<p>Once the metadata is transfered, we can subset the <code>SingleCellExperiment</code> assay.</p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">scp</span>[[<span class="st">"peptides"</span>]]
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">sce</span>[, <span class="kw">sce</span><span class="op">$</span><span class="kw">SampleAnnotation</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sc_0"</span>, <span class="st">"sc_m0"</span>, <span class="st">"sc_u"</span>)]
</pre></div>
<p>Then add this assay back into the <code>QFeatures</code> object. This is done using the <code>addAssay</code> function. We also preserve the links between dfeatures using the <code>addAssayLinkOneToOne</code>. Since the features did not change (only the samples did), one-to-one links between features are added.</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="kw">scp</span> <span class="op">%&gt;%</span>
  <span class="fu">addAssay</span>(y = <span class="kw">sce</span>, 
           name = <span class="st">"peptides_filter1"</span>) <span class="op">%&gt;%</span>
  <span class="fu">addAssayLinkOneToOne</span>(from = <span class="st">"peptides"</span>, 
                       to = <span class="st">"peptides_filter1"</span>) <span class="op">-&gt;</span>
  <span class="kw">scp</span>
</pre></div>
</div>
<div id="filter-based-on-the-median-relative-intensity" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-based-on-the-median-relative-intensity" class="anchor"></a>Filter based on the median relative intensity</h2>
<p>We compute the median relative reporter ion intensity for each cell separately and apply a filter based on this statistic. This procedure recalls that of library size filtering commonly performed in scRNA-Seq data analysis, where the library size is the sum of the counts in each single cell. We will store the median intensity in the <code>colData</code> of the <code>peptides_filter1</code> assay (so the <code>SingleCellExperiment</code> object) because this metric is specific to that assay. The medians are computed on the quantitative data using <code>colMedians</code> that requires a data matrix. The data matrix can be extracted from a <code>SingleCellExperiment</code> using the <code>assay</code> function.</p>
<div class="sourceCode" id="cb25"><pre class="downlit">
<span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]] <span class="op">%&gt;%</span>
  <span class="kw">assay</span> <span class="op">%&gt;%</span>
  <span class="kw">matrixStats</span>::<span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html">colMedians</a></span>(na.rm = <span class="fl">TRUE</span>) <span class="op">-&gt;</span>
  <span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]]<span class="op">$</span><span class="kw">MedianRI</span>
</pre></div>
<p>Looking at the distribution of the median per cell can highlight low-quality cells.</p>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]] <span class="op">%&gt;%</span>
  <span class="kw">colData</span> <span class="op">%&gt;%</span>
  <span class="kw">data.frame</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">MedianRI</span>, fill = <span class="kw">SampleAnnotation</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>()
</pre></div>
<p>Here all values seems reasonable so no filtering is needed. If it were the case, the same procedure as the previous section can be used for selecting the cells that pass the filter.</p>
</div>
<div id="filter-based-on-the-median-cv" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-based-on-the-median-cv" class="anchor"></a>Filter based on the median CV</h2>
<p>The median CV measures the consistency of quantification for a group of peptides that belong to a protein. We remove cells that exhibit high median CV over the different proteins. We compute the median CV per cell using the <code>computeMedianCV</code> function from the <code>scp</code> package. The function takes the <code>peptides_filter1</code> assay, protein and peptide information from the assay <code>rowData</code>, and the batch names in the <code>colData</code> of the <code>QFeatures</code> object. The peptide and batch information are required to perform normalization prior to the CV computation. Protein information is required since the CV are computed at the protein level.</p>
<div class="sourceCode" id="cb27"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeMedianCV.html">computeMedianCV</a></span>(<span class="kw">scp</span>,
                       i = <span class="st">"peptides_filter1"</span>, 
                       proteinCol = <span class="st">"protein"</span>, 
                       peptideCol = <span class="st">"peptide"</span>, 
                       batchCol = <span class="st">"Set"</span>)
</pre></div>
<p>The computed CVs are stored in the <code>colData</code> of the <code>peptides_filter1</code> assay and holds the median CV per cell computed using at least 5 observations (peptides). The main interest of computing the median CV per cell is to filter cells with reliable quantification. The blank samples are not expected to have reliable quantifications and hence can be used to estimate an empirical null distribution of the CV. This distribution helps defining a threshold that filters out single-cells that contain noisy quantification.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]] <span class="op">%&gt;%</span>
  <span class="kw">colData</span> <span class="op">%&gt;%</span>
  <span class="kw">data.frame</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">MedianCV</span>, 
             fill = <span class="kw">SampleType</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(xintercept = <span class="fl">0.4</span>)
</pre></div>
<p>We can see that the protein quantifiation for single-cells are much more consistent within cells than the blank channels. This is expected and can help defining a threshold. We keep the cells that have a a median CV lower than 0.4. We also remove the blanks since all QC is performed.</p>
<p>Again, we first compute the selection criterion.</p>
<div class="sourceCode" id="cb29"><pre class="downlit">
<span class="kw">keepSample</span> <span class="op">&lt;-</span> <span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]]<span class="op">$</span><span class="kw">MedianCV</span> <span class="op">&lt;</span> <span class="fl">0.4</span> <span class="op">&amp;</span>
  <span class="kw">scp</span>[[<span class="st">"peptides_filter1"</span>]]<span class="op">$</span><span class="kw">SampleAnnotation</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sc_m0"</span>, <span class="st">"sc_u"</span>)
<span class="kw">keepSample</span>[<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">keepSample</span>)] <span class="op">&lt;-</span> <span class="fl">FALSE</span>
</pre></div>
<p>Then we subset the assay for samples that pass the filter</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">specht2019v2</span>[[<span class="st">"peptides_filter1"</span>]]
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">sce</span>[, <span class="kw">keepSample</span>]
</pre></div>
<p>Finally, we add the filtered assay to the <code>QFeatures</code>object and created the required linking with the previous assay.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw">specht2019v2</span> <span class="op">%&gt;%</span>
  <span class="fu">addAssay</span>(y = <span class="kw">sce</span>, 
           name = <span class="st">"peptides_filter2"</span>) <span class="op">%&gt;%</span>
  <span class="fu">addAssayLinkOneToOne</span>(from = <span class="st">"peptides_filter1"</span>, 
                       to = <span class="st">"peptides_filter2"</span>) <span class="op">-&gt;</span>
  <span class="kw">specht2019v2</span>
</pre></div>
</div>
</div>
<div id="process-the-peptide-data" class="section level1">
<h1 class="hasAnchor">
<a href="#process-the-peptide-data" class="anchor"></a>Process the peptide data</h1>
<p>In this vignette, the peptide data are further processed before aggregation to proteins. The steps are: normalization, filter peptides based on missing data and log-transformation.</p>
<div id="normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h2>
<p>The columns (samples) of the peptide data are first normalized by dividing the relative intensities by the median relative intensities. Then, the rows (peptides) are normalized by dividing the relative intensities by the mean relative intensities. The normalized data is stored in a separate assay. This normalization procedure is suggested in the SCoPE2 analysis and is applied using the <code>sweep</code> method. Beside the dataset and the assay to normalize, the method expects a margin, that is either row-wise (<code>1</code>) or column-wise (<code>2</code>) transformation, the function to apply, and a vector of values to apply. More conventional normalization procedure can be found in <code><a href="https://rdrr.io/pkg/QFeatures/man/QFeatures-processing.html">?QFeatures::normalize</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="co">## Scale columns with median</span>
  <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span>(<span class="kw">scp</span>, 
        i = <span class="st">"peptides_filter2"</span>, 
        MARGIN = <span class="fl">2</span>, 
        FUN = <span class="st">"/"</span>, 
        STATS = <span class="fu">colMedians</span>(<span class="fu">assay</span>(<span class="kw">.</span>[[<span class="st">"peptides_filter2"</span>]]), 
                           na.rm = <span class="fl">TRUE</span>), 
        name = <span class="st">"peptides_norm_col"</span>) <span class="op">%&gt;%</span>
  <span class="co">## Scale rows with mean</span>
  <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span>(i = <span class="st">"peptides_norm_col"</span>, 
        MARGIN = <span class="fl">1</span>, 
        FUN = <span class="st">"/"</span>, 
        STATS = <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="fu">assay</span>(<span class="kw">.</span>[[<span class="st">"peptides_norm_col"</span>]]), 
                         na.rm = <span class="fl">TRUE</span>), 
        name = <span class="st">"peptides_norm"</span>) <span class="op">-&gt;</span>
  <span class="kw">scp</span>
</pre></div>
</div>
<div id="remove-peptides-with-high-missing-rate" class="section level2">
<h2 class="hasAnchor">
<a href="#remove-peptides-with-high-missing-rate" class="anchor"></a>Remove peptides with high missing rate</h2>
<p>Peptides that contain many missing values are not informative. Therefore, another common procedure is to remove higly missing data. In this example, we remove peptides with more than 99 % missing data. This is done using the <code>filterNA</code> function from <code>QFeatures</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">filterNA</span>(<span class="kw">scp</span>,
                i = <span class="st">"peptides_norm"</span>, 
                pNA = <span class="fl">0.99</span>)
</pre></div>
</div>
<div id="log-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#log-transformation" class="anchor"></a>Log-transformation</h2>
<p>In this vignette, we perform log2-transformation using the <code>logTransform</code> method from <code>QFeatures</code>. Other log-transformation can be applied by changing the <code>base</code> argument.</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">logTransform</span>(<span class="kw">scp</span>,
                    base = <span class="fl">2</span>,
                    i = <span class="st">"peptides_norm"</span>, 
                    name = <span class="st">"peptides_log"</span>)
</pre></div>
</div>
</div>
<div id="aggregate-peptide-data-to-protein-data" class="section level1">
<h1 class="hasAnchor">
<a href="#aggregate-peptide-data-to-protein-data" class="anchor"></a>Aggregate peptide data to protein data</h1>
<p>Similarly to aggregating PSM data to peptide data, we can aggregate peptide data to protein data using the <code>aggregateFeatures</code> function.</p>
<div class="sourceCode" id="cb35"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">aggregateFeatures</span>(i = <span class="st">"peptides_log"</span>,  
                         name = <span class="st">"proteins"</span>, 
                         fcol = <span class="st">"protein"</span>, 
                         fun = <span class="kw">matrixStats</span>::<span class="kw"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html">colMedians</a></span>, na.rm = <span class="fl">TRUE</span>)
</pre></div>
</div>
<div id="process-the-protein-data" class="section level1">
<h1 class="hasAnchor">
<a href="#process-the-protein-data" class="anchor"></a>Process the protein data</h1>
<p>The protein data is processed in three steps: normalization, imputation (using the KNN algorithm) and batch correction (using the <code>ComBat</code> algorithm).</p>
<div id="normalization-1" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization-1" class="anchor"></a>Normalization</h2>
<p>Normalization is performed similarly to peptide normalization. We use the same functions, but since the data were log-transformed at the peptide level, we subtract by the statistic (median or mean) instead of dividing.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="co">## Center columns with median</span>
<span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span>(<span class="kw">scp</span>, 
      i = <span class="st">"proteins"</span>, 
      MARGIN = <span class="fl">2</span>, 
      FUN = <span class="st">"-"</span>, 
      STATS = <span class="fu">colMedians</span>(<span class="fu">assay</span>(<span class="kw">.</span>[[<span class="st">"proteins"</span>]]), 
                         na.rm = <span class="fl">TRUE</span>), 
      name = <span class="st">"proteins_norm_col"</span>) <span class="op">%&gt;%</span>
  <span class="co">## Center rows with mean</span>
  <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span>(i = <span class="st">"proteins_norm_col"</span>, 
        MARGIN = <span class="fl">1</span>, 
        FUN = <span class="st">"-"</span>, 
        STATS = <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="fu">assay</span>(<span class="kw">.</span>[[<span class="st">"proteins_norm_col"</span>]]), 
                         na.rm = <span class="fl">TRUE</span>), 
        name = <span class="st">"proteins_norm"</span>) <span class="op">-&gt;</span>
  <span class="kw">scp</span>
</pre></div>
</div>
<div id="imputation" class="section level2">
<h2 class="hasAnchor">
<a href="#imputation" class="anchor"></a>Imputation</h2>
<p>The protein data contains a lot of missing values. The graph below shows the distribution of the proportion missingness in cells. Cells contain on average over 75 % missing values!</p>
<div class="sourceCode" id="cb37"><pre class="downlit">
<span class="fu">longFormat</span>(<span class="kw">specht2019v2</span>[, , <span class="st">"proteins_norm"</span>]) <span class="op">%&gt;%</span>
  <span class="kw">data.frame</span> <span class="op">%&gt;%</span>
  <span class="fu">group_by</span>(<span class="kw">colname</span>) <span class="op">%&gt;%</span>
  <span class="fu">summarize</span>(missingness = <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">value</span>))) <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">missingness</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>()
</pre></div>
<p>Whether imputation is beneficial or deleterious for the data will not be discussed in this vignette. Missing data can either be imputed by replacing the missing values by predicted values or can be taken into account during the statistical modeling. We here give an example of imputation using the K nearest neighbors algorithm, with k = 3, that is available from the <code>impute</code> mehtod. More details about the arguments can be found in <code>?impute::impute.knn</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="kw">scp</span> <span class="op">&lt;-</span> <span class="fu">impute</span>(<span class="kw">scp</span>, 
              i = <span class="st">"proteins_norm"</span>, 
              method = <span class="st">"knn"</span>, 
              k = <span class="fl">3</span>, rowmax = <span class="fl">1</span>, colmax= <span class="fl">1</span>, 
              maxp = <span class="fl">Inf</span>, rng.seed = <span class="fl">1234</span>)
</pre></div>
</div>
<div id="batch-correction" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-correction" class="anchor"></a>Batch correction</h2>
<p>A very important step for processing SCP data is to correct for batch effects. Batch effects are caused by technical variation occuring during different MS runs. Since only a small number of single-cells can be acquired at once, batch effects are unavoidable.</p>
<p>The <code>ComBat</code> function from the <code>sva</code> package can be used to perform batch correction as it is performed in the SCoPE2 analysis. We do not claim that <code>ComBat</code> is the best algorithm for batch correcting SCP data. The same steps as described below could be applied for any batch correcting method.</p>
<p>We first extract the assay to process.</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">scp</span>[[<span class="st">"proteins_impd"</span>]]
</pre></div>
<p>Next, we need to provide a design matrix and the batch annotation to <code>Combat</code>. The design matrix allows to protect variables of interest, in our case <code>SampleAnnotation</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="kw">batch</span> <span class="op">&lt;-</span> <span class="fu">colData</span>(<span class="kw">sce</span>)<span class="op">$</span><span class="kw">Set</span>
<span class="kw">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(<span class="op">~</span> <span class="kw">SampleType</span>, data = <span class="fu">colData</span>(<span class="kw">sce</span>))
</pre></div>
<p>We then load and call <code>ComBat</code> and overwrite the data matrix. Recall the data matrix can be accessed using the <code>assay</code> function.</p>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">sva</span>)
<span class="fu">assay</span>(<span class="kw">sce</span>) <span class="op">&lt;-</span> <span class="fu">ComBat</span>(dat = <span class="fu">assay</span>(<span class="kw">sce</span>), 
                     batch = <span class="kw">batch</span>, 
                     mod = <span class="kw">model</span>)
</pre></div>
<p>Finally, we add the batch corrected assay to the <code>QFeatures</code> object and create the feature links.</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="fu">addAssay</span>(<span class="kw">scp</span>, 
         y = <span class="kw">sce</span>,
         name = <span class="st">"proteins_batchC"</span>) <span class="op">%&gt;%</span>
  <span class="fu">addAssayLinkOneToOne</span>(from = <span class="st">"proteins_impd"</span>,
                       to = <span class="st">"proteins_batchC"</span>) <span class="op">-&gt;</span>
  <span class="kw">scp</span>
</pre></div>
</div>
</div>
<div id="dimension-reduction" class="section level1">
<h1 class="hasAnchor">
<a href="#dimension-reduction" class="anchor"></a>Dimension reduction</h1>
<p><strong><span class="citation">@todo</span></strong></p>
<div class="sourceCode" id="cb43"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">scater</span>)
</pre></div>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>As a side note, the <code>scater</code> packages provides a suite of functions that perform and visualize dimension reduction for <code>SingleCellExperiment</code> objects. Since our assays are all <code>SingleCellExperiment</code> objects, we can perform regular PCA on the protein data in just a few commands.</p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="co">## Transfer sample annotation to the protein assay</span>
<span class="kw">specht2019v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transferColDataToAssay.html">transferColDataToAssay</a></span>(<span class="kw">specht2019v2</span>, <span class="st">"proteins_batchC_norm"</span>) 
<span class="kw">specht2019v2</span>[[<span class="st">"proteins_batchC_norm"</span>]] <span class="op">%&gt;%</span>
  <span class="co">## Perform PCA, see ?runPCA for more info on arguments</span>
  <span class="fu">runPCA</span>(ncomponents = <span class="fl">50</span>, 
         ntop = <span class="fl">Inf</span>, 
         scale = <span class="fl">TRUE</span>, 
         exprs_values = <span class="fl">1</span>, 
         name = <span class="st">"PCA"</span>) <span class="op">%&gt;%</span>
  <span class="co">## Plotting is performed in a single line of code</span>
  <span class="fu">plotPCA</span>(colour_by = <span class="st">"SampleType"</span>)
</pre></div>
</div>
<div id="umap" class="section level2">
<h2 class="hasAnchor">
<a href="#umap" class="anchor"></a>UMAP</h2>
</div>
<div id="tsne" class="section level2">
<h2 class="hasAnchor">
<a href="#tsne" class="anchor"></a>tSNE</h2>
</div>
</div>
<div id="monitoring-data-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#monitoring-data-processing" class="anchor"></a>Monitoring data processing</h1>
<p>The <code>QFeatures</code> plot shows the quantitative data for a features at the different expression levels. For instance, suppose we are interested in the protein <em>VIM</em> (protein ID is <code>P08670</code>) from one of the SCoPE2 batches (<em>e.g.</em> <code>191110S_LCB7_X_APNOV16plex2_Set_9</code>). A useful QC is to monitor the data processing at the PSM, peptide and protein level. This can easily be done thanks to the <code>QFeatures</code> framework. Using the <code>subsetByAssay</code> and the <code>subsetByRow</code> functions, we can extract the batch and feature of interest, respectively. Then, the data is formated to a long format table that can easily be plugged in the <code>ggplot2</code> visualization tool.</p>
<div class="sourceCode" id="cb45"><pre class="downlit">
<span class="kw">scp</span> <span class="op">%&gt;%</span>
  <span class="co">## Get the features related to VIM (P08670)</span>
  <span class="fu">subsetByFeature</span>(<span class="st">"P08670"</span>) <span class="op">%&gt;%</span>
  <span class="co">## Format the `QFeatures` to a long format table</span>
  <span class="fu">longFormat</span>(colDataCols = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Set"</span>, <span class="st">"SampleType"</span>, <span class="st">"Channel"</span>)) <span class="op">%&gt;%</span>
  <span class="kw">data.frame</span> <span class="op">%&gt;%</span>
  <span class="co">## Keep only the SCoPE2 set of interest</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="kw">Set</span> <span class="op">==</span> <span class="st">"191110S_LCB7_X_APNOV16plex2_Set_9"</span>) <span class="op">%&gt;%</span>
  <span class="co">## This is used to preserve ordering of the samples and assays in ggplot2</span>
  <span class="fu">mutate</span>(assay = <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">assay</span>, levels = <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw">scp</span>)),
         Channel = <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"RI"</span>, <span class="st">"TMT-"</span>, <span class="kw">Channel</span>),
         Channel = <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw">Channel</span>, levels = <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">Channel</span>))) <span class="op">%&gt;%</span>
  <span class="co">## Start plotting</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">Channel</span>, y = <span class="kw">value</span>, group = <span class="kw">rowname</span>, col = <span class="kw">SampleType</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span>
  <span class="co">## Plot every assay in a separate facet</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(facets = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span>(<span class="kw">assay</span>), scales = <span class="st">"free_y"</span>) <span class="op">+</span>
  <span class="co">## Annotate plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Channels"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"Intensity (arbitrary units)"</span>) <span class="op">+</span>
  <span class="co">## Improve plot aspect</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.text.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(angle = <span class="fl">90</span>),
        strip.text = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(hjust = <span class="fl">0</span>),
        legend.position = <span class="st">"bottom"</span>)
</pre></div>
<p>This graph can help us track the processing of the data and help us detect anomalies in the data. For instance, we can see that 2 samples (TMT-13 and TMT-15) were removed during median filtering because all signal is 0. We can also see that no data was recorded for TMT-12, TMT-14, and TMT-16. Since TMT-12 to 16 are either macrophages or monocytes, those channels are not expected to be removed or have missing data. Thanks to this diagnostic plot, we notified Specht and colleagues there was an issue in their dataset and lead to version 3 of the SCoPE2 dataset.</p>
</div>
<div id="cell-type-dependent-missingness" class="section level1">
<h1 class="hasAnchor">
<a href="#cell-type-dependent-missingness" class="anchor"></a>Cell type dependent missingness</h1>
<p>We showed previously in this vignette that SCP data is characterized by a high rate of missing data. This missing data can be either technical or biological. To illustrate this, we plot the rate of missingness in macrophages against the rate of missingness in monocytes. Every point in the graph represents a peptide. We color the peptide by the amount of change in expression observed between the macrophages and the monocytes.</p>
<p>The code for plotting is rather elaborate, but notice the necessary data can easily be extracted using only few commands.</p>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="co">## Transfer the sample annotation from the `QFeatures` object </span>
<span class="co">## to the peptide assay</span>
<span class="kw">specht2019v2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transferColDataToAssay.html">transferColDataToAssay</a></span>(<span class="kw">specht2019v2</span>, <span class="st">"peptides_log"</span>)
<span class="kw">sce</span> <span class="op">&lt;-</span> <span class="kw">specht2019v2</span>[[<span class="st">"peptides_log"</span>]]
<span class="co">## Extract expression data from macrophages and monocytes separately</span>
<span class="kw">monoDat</span> <span class="op">&lt;-</span> <span class="fu">assay</span>(<span class="kw">sce</span>[, <span class="kw">sce</span><span class="op">$</span><span class="kw">SampleType</span> <span class="op">==</span> <span class="st">"Monocyte"</span>])
<span class="kw">m0Dat</span> <span class="op">&lt;-</span> <span class="fu">assay</span>(<span class="kw">sce</span>[, <span class="kw">sce</span><span class="op">$</span><span class="kw">SampleType</span> <span class="op">==</span> <span class="st">"Macrophage"</span>])
<span class="co">## Create a table with the rate of missingness for each cell type and </span>
<span class="co">## compute the log fold change in median expression</span>
<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(m0Missing = <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">m0Dat</span>)) <span class="op">*</span> <span class="fl">100</span>,
           monoMissing = <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="kw">monoDat</span>)) <span class="op">*</span> <span class="fl">100</span>,
           logFC = <span class="fu">rowMedians</span>(<span class="kw">m0Dat</span>, na.rm = <span class="fl">TRUE</span>) <span class="op">-</span> <span class="fu">rowMedians</span>(<span class="kw">monoDat</span>, na.rm = <span class="fl">TRUE</span>)) <span class="op">%&gt;%</span>
  <span class="co">## Trim log fold change to dampen the impact of outliers</span>
  <span class="fu">mutate</span>(logFC = <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw">logFC</span>) <span class="op">&gt;</span> <span class="fl">1</span>, <span class="fl">1</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span>(<span class="kw">logFC</span>), <span class="kw">logFC</span>)) <span class="op">%&gt;%</span>
  <span class="co">## Start plotting</span>
  <span class="kw">ggplot</span> <span class="op">-&gt;</span> 
  <span class="kw">p</span>
<span class="kw">p</span> <span class="op">+</span>
  <span class="co">## Plot rate of missingness in macrophages </span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">m0Missing</span>), bins = <span class="fl">100</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.title.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(), axis.text.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        axis.ticks.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>()) <span class="op">+</span>
  <span class="fu">plot_spacer</span>() <span class="op">+</span> 
  <span class="kw">p</span> <span class="op">+</span> 
  <span class="co">## Plot missingness in macrophages vs missingness in monocytes</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(x = <span class="kw">m0Missing</span>, y = <span class="kw">monoMissing</span>, col = <span class="kw">logFC</span>)) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(legend.position = <span class="st">"bottom"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Missingness (%) in macrophages"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"Missingness (%) in monocytes"</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradient2</a></span>(low = <span class="st">"#048ABF"</span>, mid = <span class="st">"bisque2"</span>, high = <span class="st">"#FF5733"</span>,
                        breaks = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>),
                        labels = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"monocyte"</span>, <span class="st">"macrophage"</span>)) <span class="op">+</span>
  <span class="kw">p</span> <span class="op">+</span>
  <span class="co">## Plot rate of missingness in monocytes</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(y = <span class="kw">monoMissing</span>), bins = <span class="fl">100</span>) <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>() <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(axis.title.y = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(), axis.text.y = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        axis.ticks.y = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(), 
        axis.text.x = <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(angle = <span class="op">-</span><span class="fl">90</span>, vjust = <span class="fl">0.5</span>, hjust=<span class="fl">0</span>)) <span class="op">+</span>
  <span class="co">## Adjust plotting layout</span>
  <span class="fu">plot_layout</span>(widths = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.9</span>, <span class="fl">0.1</span>), heights = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>))
</pre></div>
<p>We can see that most of the peptides have equal missingness in macrophages and monocytes. This is expected for technical missingness and this is further confirmed by the fact that that the slope 1 line is dominated by peptides with no large difference in log fold change (pale yellow). Interestingly, we can also observe that the missingness for some peptides can be explained by biological processes that are specific to one or the other cell type. Peptides more expressed in macrophages (red) tend to be less missing in monocytes. Conversely, peptides more expressed in monocytes tend to be less missing in monocytes.</p>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
</div>
<div id="reference" class="section level1">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
