<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single Cell Proteomics data modelling • scp</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single Cell Proteomics data modelling">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scp</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.17.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scp.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced.html">Advanced usage of `scp`</a>
    </li>
    <li>
      <a href="../articles/QFeatures_nutshell.html">QFeatures in a nutshell</a>
    </li>
    <li>
      <a href="../articles/read_scp.html">Load Single-Cell Proteomics data using `readSCP`</a>
    </li>
    <li>
      <a href="../articles/reporting_missing_values.html">Reporting missing values for Single Cell Proteomics</a>
    </li>
    <li>
      <a href="../articles/scp_data_modelling.html">Single Cell Proteomics data modelling</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/UCLouvain-CBIO/scp/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single Cell Proteomics data modelling</h1>
                        <h4 data-toc-skip class="author">Laurent
Gatto</h4>
                                    <h4 data-toc-skip class="author">Christophe
Vanderaa</h4>
                        
            <h4 data-toc-skip class="date">26 February 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/UCLouvain-CBIO/scp/blob/master/vignettes/scp_data_modelling.Rmd" class="external-link"><code>vignettes/scp_data_modelling.Rmd</code></a></small>
      <div class="hidden name"><code>scp_data_modelling.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this tutorial, we’ll walk you through the process of modelling
single-cell proteomics (SCP) data using the <em>scplainer</em> approach
(<span class="citation">Vanderaa and Gatto (2024)</span>). By the end of
this vignette, you will be able to:</p>
<ul>
<li>Define and estimate a model suitable for SCP data</li>
<li>Filter peptides based on the patterns of missing values</li>
<li>Exploring the model output through analysis of variance</li>
<li>Exploring the model output through differential abundance
analysis</li>
<li>Exploring the model output through component analysis</li>
<li>Perform batch correction to remove unwanted technical artefacts</li>
</ul>
<p>The last point will allow you to generate SCP data that is suitable
for downstream analysis, such as clustering or trajectory inference. The
figure below provides a roadmap of the workflow:</p>
<div class="float">
<img src="figures/ScpModel-class.png" alt="scplainer modelling workflow"><div class="figcaption">scplainer modelling workflow</div>
</div>
<p>The vignette will start with the processed data extracted as a
<code>SingleCellExperiment</code> object from a processed
<code>QFeatures</code> object. We will not cover data processing as it
is covered in another vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://UCLouvain-CBIO.github.io/scp" class="external-link">"scp"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://patchwork.data-imaginist.com" class="external-link">"patchwork"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-data-set">Example data set<a class="anchor" aria-label="anchor" href="#example-data-set"></a>
</h2>
<p>The example data set is a subset of the <code>leduc2022_pSCoPE</code>
data set (see <code><a href="https://rdrr.io/pkg/scpdata/man/leduc2022_pSCoPE.html" class="external-link">?scpdata::leduc2022_pSCoPE</a></code> for more info).
The data is acquired using TMT-18 multiplexing and data-dependent
acquisition (DDA). The data has been processed using a minimal
workflow:</p>
<ul>
<li>Any zero value has been replaced by NA.</li>
<li>Any peptide was removed from the data set if it matched to a decoy
or contaminant peptide, had low spectral purity, had low identification
confidence, or had a high sample to carrier ratio.</li>
<li>Any cell was removed from the data set if it had a high coefficient
of variation, had an abberrant median intensity, or had few identified
peptides.</li>
<li>PSM data were combined into peptide data. When multiple PSM match to
the same peptide, the median intensity was taken.</li>
<li>Intensities were log2 transformed</li>
<li>To limit intensive computation, we have limited the data set to 200
peptides in 73 cells.</li>
</ul>
<p>We suggest using this minimal processing workflow, although the
approach presented here is agnostic of previous processing and allows
for other custom workflows. The data processing was conducted with
<code>QFeatures</code> and <code>scp</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"leduc_minimal"</span><span class="op">)</span></span>
<span><span class="va">leduc_minimal</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 200 73 </span></span>
<span><span class="co">#&gt; metadata(1): model</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(200): SAVEDEGLK APNVVVTR ... FLLAVSRDR EASMVITESPAALQLR</span></span>
<span><span class="co">#&gt; rowData names(6): Sequence Reverse ... Leading.razor.protein.symbol</span></span>
<span><span class="co">#&gt;   gene</span></span>
<span><span class="co">#&gt; colnames(73): eAL00219RI5 eAL00219RI6 ... wAL00286RI17 wAL00286RI18</span></span>
<span><span class="co">#&gt; colData names(12): Set Channel ... MedianCV passQC</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>The data set is formatted as a <code>SingleCellExperiment</code>
object. The data set consists of 200 peptides and 73 cells. Peptide
annotations can be retrieved from the <code>rowData</code> and cell
annotations can be retrieved from the <code>colData</code>. The cell
annotation will be used during modelling.</p>
<p>A full reanalysis of Leduc’s nPOP dataset is also available <a href="https://uclouvain-cbio.github.io/SCP.replication/articles/scplainer_leduc2022.html" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="data-modelling">Data modelling<a class="anchor" aria-label="anchor" href="#data-modelling"></a>
</h2>
<p>The core of the approach relies on statistical modelling of the data
using linear regression. Under the hood, the model fetches as input the
intensity matrix stored in <code>assay(leduc_minimal)</code>. The cell
annotations are retrieved using <code>colData(leduc_minimal)</code>.
They describe known technical and biological variables that may
influence the acquired peptide intensities. The annotations are used to
build a regression model with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
parameters. Then, the model estimates the coefficients. Coefficients
provide the contributions of each parameter to the expression of each of
peptide as well as the uncertainty of the estimation. These will be
explored in the following section.</p>
<p>We’ll start by defining the variables to include in the model. Recall
that the example data set contains TMT-labeled cells. This means that
each MS acquisition run contains multiple cells. Each run is subject to
technical fluctuations that can lead to undesired variation, this is
known as a batch effect.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">$</span><span class="va">Set</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; eAL00219 eAL00220 eAL00221 wAL00284 wAL00285 wAL00286 </span></span>
<span><span class="co">#&gt;       13       12       12       12       12       12</span></span></code></pre></div>
<p>The labelling reagent (<code>Channel</code>) can also lead to
undesired systematic effects and will also be considered as a source for
batch effects.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">$</span><span class="va">Channel</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  TMT126 TMT127N TMT127C TMT128N TMT128C TMT129N TMT129C TMT130N TMT130C TMT131N </span></span>
<span><span class="co">#&gt;       0       0       0       0       6       5       6       6       5       4 </span></span>
<span><span class="co">#&gt; TMT131C TMT132N TMT132C TMT133N TMT133C TMT134N TMT134C TMT135N </span></span>
<span><span class="co">#&gt;       5       5       5       6       5       5       5       5</span></span></code></pre></div>
<p>Finally, each cell is processed individually and the amount of
peptide material recovered from each cell may lead to undesired
variation as well. This issue is usually solved through normalization,
such as removing the median intensity from each cell. Normalization was
internationally omitted in the minimal data processing so that we can
account for it during modelling. The median intensity were already
computed (<code>MedianIntensity</code>).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">$</span><span class="va">MedianIntensity</span>, breaks <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/technical_variables3-1.png" width="700"></p>
<p>Finally, the biological variable of interest in the example data set
is the cell type that is known because cells come from 2 cell lines
(<code>SampleType</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">$</span><span class="va">SampleType</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Melanoma Monocyte </span></span>
<span><span class="co">#&gt;       37       36</span></span></code></pre></div>
<p>We create a <code>formula</code> object that will define which
variable must be modelled in our analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">f</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fl">1</span> <span class="op">+</span> <span class="co">## intercept</span></span>
<span>    <span class="va">Channel</span> <span class="op">+</span> <span class="va">Set</span> <span class="op">+</span> <span class="co">## batch variables</span></span>
<span>    <span class="va">MedianIntensity</span> <span class="op">+</span> <span class="co">## normalization</span></span>
<span>    <span class="va">SampleType</span> <span class="co">## biological variable</span></span></code></pre></div>
<p>Note that the formula can be adapted to the data set. For instance,
no labelling reagents is used for LFQ experiments, so it can be dropped.
Similarly, each cell in an LFQ experiment is acquired in a single run so
MS run cannot be used as a batch effect variable. The day of acquisition
could be used instead.</p>
<p>Once a model is defined, we fit it with
<code><a href="../reference/ScpModel-Workflow.html">scpModelWorkflow()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc_minimal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-Workflow.html">scpModelWorkflow</a></span><span class="op">(</span><span class="va">leduc_minimal</span>, formula <span class="op">=</span> <span class="va">f</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in scpModelWorkflow(leduc_minimal, formula = f): An element called</span></span>
<span><span class="co">#&gt; 'model' is already present in the metadata. The associated content will be</span></span>
<span><span class="co">#&gt; overwritten.</span></span></code></pre></div>
<p>You can always retrieve the formula that was used to fit model
using</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelFormula</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; ~1 + Channel + Set + MedianIntensity + SampleType</span></span></code></pre></div>
<p>The data that is modelled by each variable are contained in the
so-called effect matrices.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelEffects</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 4</span></span>
<span><span class="co">#&gt; names(4): Channel Set MedianIntensity SampleType</span></span></code></pre></div>
<p>Similarly, the data that could not be captured by the model are
contained in the residual matrix.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelResiduals</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                eAL00219RI5 eAL00219RI6 eAL00219RI7 eAL00219RI8 eAL00219RI9</span></span>
<span><span class="co">#&gt; SAVEDEGLK               NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; APNVVVTR        -0.2112685   0.2569749 -0.08076257 -0.03317054  0.06679681</span></span>
<span><span class="co">#&gt; IVVVTAGVR               NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; GFQEVVTPNIFNSR   0.6375780  -0.3356473  0.34648412 -0.80982152  0.96378484</span></span>
<span><span class="co">#&gt; ENAYDLEANLAVLK          NA          NA          NA          NA          NA</span></span></code></pre></div>
<p>Finally, the input data used to model the can also be retrieved.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelInput</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                eAL00219RI5 eAL00219RI6 eAL00219RI7 eAL00219RI8 eAL00219RI9</span></span>
<span><span class="co">#&gt; SAVEDEGLK               NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; APNVVVTR          13.56997    13.85389   13.362218   13.110304    14.03041</span></span>
<span><span class="co">#&gt; IVVVTAGVR               NA          NA          NA          NA          NA</span></span>
<span><span class="co">#&gt; GFQEVVTPNIFNSR    10.78733     9.36700    9.815223    7.878296    11.01890</span></span>
<span><span class="co">#&gt; ENAYDLEANLAVLK          NA          NA          NA          NA          NA</span></span></code></pre></div>
<p>Note that the number of peptides changed. This is the consequence of
peptide filtering.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelInput</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 140  73</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="peptide-filtering">Peptide filtering<a class="anchor" aria-label="anchor" href="#peptide-filtering"></a>
</h2>
<p>The proportion of missing values for each features is high in
single-cell proteomics data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>missing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; missing</span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;  5954  8646</span></span></code></pre></div>
<p>Many features can typically contain more coefficients to estimate
than observed values. These features cannot be estimated and will be
ignored during further steps. These features are identified by computing
the ratio between the number of observed values and the number of
coefficients to estimate. We call it the n/p ratio. You can extract the
n/p ratio for each feature:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelFilterNPRatio</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      SAVEDEGLK       APNVVVTR      IVVVTAGVR GFQEVVTPNIFNSR ENAYDLEANLAVLK </span></span>
<span><span class="co">#&gt;       1.277778       3.476190       1.411765       2.333333       1.235294 </span></span>
<span><span class="co">#&gt;      IGPLGLSPK </span></span>
<span><span class="co">#&gt;       1.888889</span></span></code></pre></div>
<p>Once the model is estimated, use <code><a href="../reference/ScpModel-Workflow.html">scpModelFilterPlot()</a></code> to
explore the distribution of n/p ratios across the features.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-Workflow.html">scpModelFilterPlot</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; To change the threshold, use:</span></span>
<span><span class="co">#&gt; scpModelFilterThreshold(object, name) &lt;- threshold</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpModelFilterPlot-1.png" width="700"></p>
<p>By default, any feature that has an n/p ration greater than 1 is
included in the analysis. However, feature with an n/p ratio close to 1
may lead to unreliable outcome because there are not enough observed
data. You could consider the n/p ratio as the average number of
replicate per coefficient to estimate. Therefore, you may want to
increase the n/p threshold.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelFilterThreshold</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span> <span class="co">## default is 1</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelFilterThreshold</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1.5</span></span>
<span><span class="fu"><a href="../reference/ScpModel-class.html">scpModelFilterThreshold</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span> <span class="co">## threshold is now 1.5</span></span>
<span><span class="co">#&gt; [1] 1.5</span></span></code></pre></div>
<p>The plot is automatically updated.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-Workflow.html">scpModelFilterPlot</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; To change the threshold, use:</span></span>
<span><span class="co">#&gt; scpModelFilterThreshold(object, name) &lt;- threshold</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpModelFilterPlot2-1.png" width="700"></p>
<p>There is no guidelines for defining a suitable threshold. If too low,
you may include noisy peptides that have too few observations. If too
high, you may remove many informative peptides. The definition of the
threshold relies on a trade off between precision and sensitivity.</p>
</div>
<div class="section level2">
<h2 id="model-exploration-analysis-of-variance">Model exploration: analysis of variance<a class="anchor" aria-label="anchor" href="#model-exploration-analysis-of-variance"></a>
</h2>
<p>The variance analysis reports the relative amount of information that
is captured by each cell annotation included in the model. The model
also includes the residual information that is not captured by the
model. This offers a first glimpse into what information is contained in
the data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVarianceAnalysis</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrameList of length 5</span></span>
<span><span class="co">#&gt; names(5): Residuals Channel Set MedianIntensity SampleType</span></span></code></pre></div>
<p>The results are a list of tables, one table for each variable. Each
table reports for each peptide the variance captures (<code>SS</code>),
the residual degrees of freedom for estimating the variance
(<code>df</code>) and the percentage of total variance explained
(<code>percentExplainedVar</code>).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaRes</span><span class="op">$</span><span class="va">SampleType</span></span>
<span><span class="co">#&gt; DataFrame with 95 rows and 4 columns</span></span>
<span><span class="co">#&gt;           feature         SS        df percentExplainedVar</span></span>
<span><span class="co">#&gt;       &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1        APNVVVTR 27.8682309        52           33.052188</span></span>
<span><span class="co">#&gt; 2   GFQEVVTPNI...  0.1106189        28            0.392004</span></span>
<span><span class="co">#&gt; 3       IGPLGLSPK  0.0442532        16            0.242434</span></span>
<span><span class="co">#&gt; 4   VAETANEEEV...  0.4973533        11            2.694342</span></span>
<span><span class="co">#&gt; 5   IDATSASVLA...  0.9271610        36            2.174504</span></span>
<span><span class="co">#&gt; ...           ...        ...       ...                 ...</span></span>
<span><span class="co">#&gt; 91  SVPTSTVFYP...    1.88199        48             4.60836</span></span>
<span><span class="co">#&gt; 92      HIVENAVQK    1.05939        19             4.80442</span></span>
<span><span class="co">#&gt; 93  TDMDNQIVVS...    4.31709         9            24.56762</span></span>
<span><span class="co">#&gt; 94       AILGSVER    2.51736        16            15.90057</span></span>
<span><span class="co">#&gt; 95     LGAEVYHTLK    2.47027        15            10.14771</span></span></code></pre></div>
<p>By default, we explore the variance for all peptides combined:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span><span class="va">vaRes</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/unnamed-chunk-5-1.png" width="240"></p>
<p>We explore the top 20 peptides that are have the highest percentage
of variance explained by the biological variable (top) or by the batch
variable (bottom).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">10</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">10</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<p>We can also group the peptide by protein. To do so, we first need to
add the peptides annotations available from the
<code>rowData</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scpAnnotateResults.html">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">vaRes</span><span class="op">$</span><span class="va">SampleType</span></span>
<span><span class="co">#&gt; DataFrame with 95 rows and 9 columns</span></span>
<span><span class="co">#&gt;           feature         SS        df percentExplainedVar     Reverse</span></span>
<span><span class="co">#&gt;       &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1        APNVVVTR 27.8682309        52           33.052188            </span></span>
<span><span class="co">#&gt; 2   GFQEVVTPNI...  0.1106189        28            0.392004            </span></span>
<span><span class="co">#&gt; 3       IGPLGLSPK  0.0442532        16            0.242434            </span></span>
<span><span class="co">#&gt; 4   VAETANEEEV...  0.4973533        11            2.694342            </span></span>
<span><span class="co">#&gt; 5   IDATSASVLA...  0.9271610        36            2.174504            </span></span>
<span><span class="co">#&gt; ...           ...        ...       ...                 ...         ...</span></span>
<span><span class="co">#&gt; 91  SVPTSTVFYP...    1.88199        48             4.60836            </span></span>
<span><span class="co">#&gt; 92      HIVENAVQK    1.05939        19             4.80442            </span></span>
<span><span class="co">#&gt; 93  TDMDNQIVVS...    4.31709         9            24.56762            </span></span>
<span><span class="co">#&gt; 94       AILGSVER    2.51736        16            15.90057            </span></span>
<span><span class="co">#&gt; 95     LGAEVYHTLK    2.47027        15            10.14771            </span></span>
<span><span class="co">#&gt;     Potential.contaminant Leading.razor.protein.id Leading.razor.protein.symbol</span></span>
<span><span class="co">#&gt;               &lt;character&gt;              &lt;character&gt;                  &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1                                           P52566                        GDIR2</span></span>
<span><span class="co">#&gt; 2                                           P26639                         SYTC</span></span>
<span><span class="co">#&gt; 3                                           P30050                         RL12</span></span>
<span><span class="co">#&gt; 4                                           P61221                        ABCE1</span></span>
<span><span class="co">#&gt; 5                                           P13667                        PDIA4</span></span>
<span><span class="co">#&gt; ...                   ...                      ...                          ...</span></span>
<span><span class="co">#&gt; 91                                          P29401                          TKT</span></span>
<span><span class="co">#&gt; 92                                          O00410                         IPO5</span></span>
<span><span class="co">#&gt; 93                                          P50991                         TCPD</span></span>
<span><span class="co">#&gt; 94                                          P26639                         SYTC</span></span>
<span><span class="co">#&gt; 95                                          P09104                         ENOG</span></span>
<span><span class="co">#&gt;            gene</span></span>
<span><span class="co">#&gt;     &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1       ARHGDIB</span></span>
<span><span class="co">#&gt; 2          TARS</span></span>
<span><span class="co">#&gt; 3         RPL12</span></span>
<span><span class="co">#&gt; 4         ABCE1</span></span>
<span><span class="co">#&gt; 5         PDIA4</span></span>
<span><span class="co">#&gt; ...         ...</span></span>
<span><span class="co">#&gt; 91          TKT</span></span>
<span><span class="co">#&gt; 92         IPO5</span></span>
<span><span class="co">#&gt; 93         CCT4</span></span>
<span><span class="co">#&gt; 94         TARS</span></span>
<span><span class="co">#&gt; 95         ENO2</span></span></code></pre></div>
<p>Then, we draw the same plot, but this time we provide the
<code>fcol</code> argument.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">10</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaRes</span>, top <span class="op">=</span> <span class="fl">10</span>, by <span class="op">=</span> <span class="st">"percentExplainedVar"</span>, effect <span class="op">=</span> <span class="st">"Set"</span>,</span>
<span>    decreasing <span class="op">=</span> <span class="cn">TRUE</span>, combined <span class="op">=</span> <span class="cn">FALSE</span>, fcol <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpVariancePlot_protein-1.png" width="960"></p>
<p>In this example dataset, we retrieve peptides that all belong to a
different protein, however grouping becomes interesting when analyzing
real data sets.</p>
<p>Alternatively, we can generate protein level results by aggregating
peptide level results.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vaProtein</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVarianceAggregate</a></span><span class="op">(</span><span class="va">vaRes</span>, fcol <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ScpModel-VarianceAnalysis.html">scpVariancePlot</a></span><span class="op">(</span></span>
<span>    <span class="va">vaProtein</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span>, top <span class="op">=</span> <span class="fl">10</span>, combined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpVarianceAggregate-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="model-exploration-differential-abundance-analysis">Model exploration: differential abundance analysis<a class="anchor" aria-label="anchor" href="#model-exploration-differential-abundance-analysis"></a>
</h2>
<p>Differential abundance analysis dives deeper into the exploration of
the data, namely for exploring the biological effects. Given two groups
of interest, such as two cell types or two treatment groups, the
differential analysis derives estimated fold changes from the linear
model’s coefficients. This provides information, for each peptide or
protein, the amount of change between the two groups and the direction
of the change. Moreover, the model provides the uncertainty of the
differences, enabling the assessment of the statistical
significance.</p>
<p>The difference of interest is specified using the
<code>contrast</code> argument. The first element points to the variable
to test and the two following element are the groups of interest to
compare. You can provide multiple contrast in a list.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-DifferentialAnalysis.html">scpDifferentialAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc_minimal</span>,</span>
<span>    contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleType"</span>, <span class="st">"Melanoma"</span>, <span class="st">"Monocyte"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrameList of length 1</span></span>
<span><span class="co">#&gt; names(1): SampleType_Melanoma_vs_Monocyte</span></span></code></pre></div>
<p>Similarly to variance analysis, the results are a list of tables, one
table for each contrast. Each table reports for each peptide the
estimated difference between the two groups, the standard error
associated to the estimation, the degrees of freedom, the t-statistics,
the associated p-value and the p-value FDR-adjusted for multiple testing
across all peptides.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span><span class="op">$</span><span class="va">SampleType_Melanoma_vs_Monocyte</span></span>
<span><span class="co">#&gt; DataFrame with 95 rows and 7 columns</span></span>
<span><span class="co">#&gt;           feature   Estimate        SE        Df tstatistic      pvalue</span></span>
<span><span class="co">#&gt;       &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1        APNVVVTR  1.2357291  0.140225        52   8.812466 6.77782e-12</span></span>
<span><span class="co">#&gt; 2   GFQEVVTPNI... -0.0950269  0.312985        28  -0.303615 7.63666e-01</span></span>
<span><span class="co">#&gt; 3       IGPLGLSPK -0.0721544  0.375395        16  -0.192209 8.49997e-01</span></span>
<span><span class="co">#&gt; 4   VAETANEEEV... -0.2619169  0.486300        11  -0.538591 6.00898e-01</span></span>
<span><span class="co">#&gt; 5   IDATSASVLA... -0.2573437  0.249598        36  -1.031031 3.09404e-01</span></span>
<span><span class="co">#&gt; ...           ...        ...       ...       ...        ...         ...</span></span>
<span><span class="co">#&gt; 91  SVPTSTVFYP...  -0.330304  0.202807        48  -1.628658    0.109932</span></span>
<span><span class="co">#&gt; 92      HIVENAVQK   0.338421  0.495880        19   0.682466    0.503181</span></span>
<span><span class="co">#&gt; 93  TDMDNQIVVS...   0.814964  1.103851         9   0.738292    0.479146</span></span>
<span><span class="co">#&gt; 94       AILGSVER   0.544206  0.368422        16   1.477127    0.159054</span></span>
<span><span class="co">#&gt; 95     LGAEVYHTLK   0.547198  0.521144        15   1.049995    0.310344</span></span>
<span><span class="co">#&gt;            padj</span></span>
<span><span class="co">#&gt;       &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1   6.43893e-10</span></span>
<span><span class="co">#&gt; 2   8.91820e-01</span></span>
<span><span class="co">#&gt; 3   9.06232e-01</span></span>
<span><span class="co">#&gt; 4   8.64161e-01</span></span>
<span><span class="co">#&gt; 5   6.81383e-01</span></span>
<span><span class="co">#&gt; ...         ...</span></span>
<span><span class="co">#&gt; 91     0.521039</span></span>
<span><span class="co">#&gt; 92     0.810207</span></span>
<span><span class="co">#&gt; 93     0.789070</span></span>
<span><span class="co">#&gt; 94     0.521039</span></span>
<span><span class="co">#&gt; 95     0.681383</span></span></code></pre></div>
<p>We then visualize the results using a volcano plot. The function
below return a volcano plot for each contrast.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-DifferentialAnalysis.html">scpVolcanoPlot</a></span><span class="op">(</span><span class="va">daRes</span><span class="op">)</span></span>
<span><span class="co">#&gt; $SampleType_Melanoma_vs_Monocyte</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Since we subset the data set for only a few cell, we lack statistical
power. Still, two peptides come out as significant. Again, to better
explore the results, we add peptide annotations available from the
<code>rowData</code>, but we also add the n/p ratio as annotation.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scpAnnotateResults.html">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">np</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-class.html">scpModelFilterNPRatio</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="va">daRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scpAnnotateResults.html">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">np</span><span class="op">)</span>, npRatio <span class="op">=</span> <span class="va">np</span><span class="op">)</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"feature"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We plot the same volcano plot, but instead of labeling points with
the peptide sequence, we will show the associated gene symbol. Also, we
can control for point aesthetics by providing a list of
<code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">ggplot2::geom_point()</a></code> arguments. For example, we can colour
each point based on the n/p ratio, and adjust point size and shape.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-DifferentialAnalysis.html">scpVolcanoPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, top <span class="op">=</span> <span class="fl">30</span>, textBy <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">npRatio</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1.5</span>, shape <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; $SampleType_Melanoma_vs_Monocyte</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 14 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>We can also provide protein-level results. To do so, the
<code><a href="../reference/ScpModel-DifferentialAnalysis.html">scpDifferentialAggregate()</a></code> relies on the
<code>metapod</code> package. We here combine the statistical test
results for peptides that belong to the same protein using Simes’
method. Simes’ method will reject the combined null hypothesis (that is
the mean <strong>protein</strong> intensities are identical between two
groups) if any of the peptide nulls are rejected.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">byProteinDA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-DifferentialAnalysis.html">scpDifferentialAggregate</a></span><span class="op">(</span></span>
<span>    <span class="va">daRes</span>, fcol <span class="op">=</span> <span class="st">"gene"</span>, method <span class="op">=</span> <span class="st">"simes"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">byProteinDA</span><span class="op">$</span><span class="va">SampleType_Melanoma_vs_Monocyte</span></span>
<span><span class="co">#&gt; DataFrame with 86 rows and 10 columns</span></span>
<span><span class="co">#&gt;         feature   Estimate    pvalue      padj     Reverse</span></span>
<span><span class="co">#&gt;     &lt;character&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1         ABCE1 -0.2619169 0.6008984  0.856797            </span></span>
<span><span class="co">#&gt; 2          ACLY -1.0078958 0.0564024  0.404217            </span></span>
<span><span class="co">#&gt; 3         ACTN4 -0.0576889 0.8105808  0.910420            </span></span>
<span><span class="co">#&gt; 4         AIFM1  0.4300612 0.1488197  0.530259            </span></span>
<span><span class="co">#&gt; 5         APMAP  0.0683780 0.7445601  0.901862            </span></span>
<span><span class="co">#&gt; ...         ...        ...       ...       ...         ...</span></span>
<span><span class="co">#&gt; 82         TPP2 -0.2494580 0.4703385  0.812359            </span></span>
<span><span class="co">#&gt; 83       TUBA1C -0.6951016 0.0492061  0.384702            </span></span>
<span><span class="co">#&gt; 84         VAT1 -0.4073205 0.0691403  0.457390            </span></span>
<span><span class="co">#&gt; 85        XRCC6  0.0289353 0.8832055  0.915129            </span></span>
<span><span class="co">#&gt; 86        YWHAG -0.0979855 0.6550366  0.873363            </span></span>
<span><span class="co">#&gt;     Potential.contaminant Leading.razor.protein.id Leading.razor.protein.symbol</span></span>
<span><span class="co">#&gt;               &lt;character&gt;              &lt;character&gt;                  &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1                                           P61221                        ABCE1</span></span>
<span><span class="co">#&gt; 2                                           P53396                         ACLY</span></span>
<span><span class="co">#&gt; 3                                           O43707                        ACTN4</span></span>
<span><span class="co">#&gt; 4                                           O95831                        AIFM1</span></span>
<span><span class="co">#&gt; 5                                           Q9HDC9                        APMAP</span></span>
<span><span class="co">#&gt; ...                   ...                      ...                          ...</span></span>
<span><span class="co">#&gt; 82                                          P29144                         TPP2</span></span>
<span><span class="co">#&gt; 83                                          Q9BQE3                        TBA1C</span></span>
<span><span class="co">#&gt; 84                                          Q99536                         VAT1</span></span>
<span><span class="co">#&gt; 85                                          P12956                        XRCC6</span></span>
<span><span class="co">#&gt; 86                                          P61981                        1433G</span></span>
<span><span class="co">#&gt;            gene        .n</span></span>
<span><span class="co">#&gt;     &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; 1         ABCE1         1</span></span>
<span><span class="co">#&gt; 2          ACLY         1</span></span>
<span><span class="co">#&gt; 3         ACTN4         1</span></span>
<span><span class="co">#&gt; 4         AIFM1         1</span></span>
<span><span class="co">#&gt; 5         APMAP         1</span></span>
<span><span class="co">#&gt; ...         ...       ...</span></span>
<span><span class="co">#&gt; 82         TPP2         1</span></span>
<span><span class="co">#&gt; 83       TUBA1C         1</span></span>
<span><span class="co">#&gt; 84         VAT1         2</span></span>
<span><span class="co">#&gt; 85        XRCC6         1</span></span>
<span><span class="co">#&gt; 86        YWHAG         1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="model-exploration-component-analysis">Model exploration: component analysis<a class="anchor" aria-label="anchor" href="#model-exploration-component-analysis"></a>
</h2>
<p>Variance and differential analysis are not specific to single-cell
applications and explore the data without considering cellular
heterogeneity. The purpose of the component analysis is to dive into the
cellular heterogeneity by representing highly dimensional data in a few
informative dimensions for visual exploration. We integrate the
component analysis with the linear regression model thanks to the APCA+
(extended ANOVA-simultaneous component analysis) framework developed by
<a href="http://dx.doi.org/10.1002/cem.2895" class="external-link">Thiel et al. 2017</a>.
Briefly, APCA+ explores the reconstructed data that is captured by each
variable separately in the presence of the unmodelled data. The
advantage of this framework is it is generic and works for any linear
model. Also, this approach is well suited for single-cell applications
as it enables the visualization and exploration of the effects of a
known variable along the unmodelled information that contains cellular
heterogeneity.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">caRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentAnalysis</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc_minimal</span>, ncomp <span class="op">=</span> <span class="fl">20</span>, method <span class="op">=</span> <span class="st">"APCA"</span>, effect <span class="op">=</span> <span class="st">"SampleType"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 2</span></span>
<span><span class="co">#&gt; names(2): bySample byFeature</span></span></code></pre></div>
<p>The results are contained in a list with 2 elements.
<code>bySample</code> contains the PC scores, that is the component
results in cell space. <code>byFeature</code> contains the eigenvectors,
that is the component results in peptide space. Each of the two elements
contains components results for the data before modelling
(<code>unmodelled</code>), for the residuals or for the APCA on the
sample type variable (<code>APCA_SampleType</code>).</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">bySample</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 3</span></span>
<span><span class="co">#&gt; names(3): unmodelled residuals APCA_SampleType</span></span>
<span><span class="va">caResCells</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 73 rows and 21 columns</span></span>
<span><span class="co">#&gt;                    PC1       PC2       PC3       PC4        PC5       PC6</span></span>
<span><span class="co">#&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; eAL00219RI5   1.652635  3.711020  -4.21156 -1.079196  0.0666983  1.408647</span></span>
<span><span class="co">#&gt; eAL00219RI6   0.122763  5.156328  -5.49091  0.184137 -0.3458518  1.048142</span></span>
<span><span class="co">#&gt; eAL00219RI7  -1.356183  6.042980  -5.04466  0.372945 -2.1629620  0.234317</span></span>
<span><span class="co">#&gt; eAL00219RI8  -2.700162  4.483173  -5.34451  0.548601 -1.5492215 -0.345117</span></span>
<span><span class="co">#&gt; eAL00219RI9   2.546416  0.572361  -5.35620  0.723007 -0.1841323  4.479647</span></span>
<span><span class="co">#&gt; ...                ...       ...       ...       ...        ...       ...</span></span>
<span><span class="co">#&gt; wAL00286RI12  -7.98580 -1.856855  0.757797  2.716883   0.389962  1.670132</span></span>
<span><span class="co">#&gt; wAL00286RI14  -8.20192 -0.746896  1.535207 -1.595404   2.795661  0.498499</span></span>
<span><span class="co">#&gt; wAL00286RI16  -1.81072 -2.407762  3.963677 -1.732718   0.143329 -0.745458</span></span>
<span><span class="co">#&gt; wAL00286RI17  -2.86441 -3.393967 -1.515343 -1.529139  -2.434284  0.121065</span></span>
<span><span class="co">#&gt; wAL00286RI18  -2.25612 -4.270107  0.987672 -0.145896  -1.382700 -1.306674</span></span>
<span><span class="co">#&gt;                    PC7       PC8       PC9       PC10       PC11        PC12</span></span>
<span><span class="co">#&gt;              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; eAL00219RI5   -1.40664 -0.569028  1.816091  -0.335920  -3.425797   -0.261528</span></span>
<span><span class="co">#&gt; eAL00219RI6   -2.03264  0.242179  2.257062   0.808499  -1.271131    0.486000</span></span>
<span><span class="co">#&gt; eAL00219RI7   -1.26036 -1.921333 -0.241230   0.825305  -1.820185    2.704122</span></span>
<span><span class="co">#&gt; eAL00219RI8   -1.31727 -0.876115 -0.826323   2.179410   2.105049    0.760836</span></span>
<span><span class="co">#&gt; eAL00219RI9    1.28303  0.785622 -0.545883  -0.625327  -0.676794   -0.943669</span></span>
<span><span class="co">#&gt; ...                ...       ...       ...        ...        ...         ...</span></span>
<span><span class="co">#&gt; wAL00286RI12  1.117698 -0.812645 -0.223081 -0.4827100  1.1006363 -0.99808069</span></span>
<span><span class="co">#&gt; wAL00286RI14  0.979609 -0.959699  3.834839 -0.9982881  0.3246867 -0.00491314</span></span>
<span><span class="co">#&gt; wAL00286RI16  3.800138 -1.646333  1.374409 -1.5396094 -0.8875749  2.42672153</span></span>
<span><span class="co">#&gt; wAL00286RI17  1.894991 -0.681766 -0.612739  0.0177883 -0.4717603  1.93884428</span></span>
<span><span class="co">#&gt; wAL00286RI18  1.586430 -1.000224 -0.140993  0.8425007 -0.0447968  1.01835905</span></span>
<span><span class="co">#&gt;                    PC13       PC14      PC15      PC16       PC17      PC18</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; eAL00219RI5   1.3248564 -0.6213675  0.275210 -0.376301  0.2819763 -0.112223</span></span>
<span><span class="co">#&gt; eAL00219RI6  -0.0312991  0.1282374 -0.977545 -0.638003 -0.0905137  0.443276</span></span>
<span><span class="co">#&gt; eAL00219RI7   0.2210126 -0.8594984 -1.371305  0.293012 -1.2920848 -0.211974</span></span>
<span><span class="co">#&gt; eAL00219RI8  -0.1068177 -0.0299786 -0.293472 -1.618090  1.0066113  0.699695</span></span>
<span><span class="co">#&gt; eAL00219RI9  -0.1327679  0.8920529  0.332752 -1.424903  0.2110955  0.484973</span></span>
<span><span class="co">#&gt; ...                 ...        ...       ...       ...        ...       ...</span></span>
<span><span class="co">#&gt; wAL00286RI12   1.258691  -1.038123  0.715850  0.594017 -0.0320364  0.386167</span></span>
<span><span class="co">#&gt; wAL00286RI14  -1.463484  -1.115054  1.304641  0.471641 -1.2468862  1.135240</span></span>
<span><span class="co">#&gt; wAL00286RI16  -0.128652   0.438629  0.405675 -1.249913  0.8714631 -0.245134</span></span>
<span><span class="co">#&gt; wAL00286RI17  -2.385663  -0.696050 -0.636856 -0.639397 -0.5319913  0.677671</span></span>
<span><span class="co">#&gt; wAL00286RI18   0.286674   0.861364 -0.262798  1.071512 -0.0554080 -0.132846</span></span>
<span><span class="co">#&gt;                    PC19       PC20          cell</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;  &lt;numeric&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; eAL00219RI5  -0.2408148 -0.4442979 eAL00219RI...</span></span>
<span><span class="co">#&gt; eAL00219RI6  -0.0921259  0.0940622 eAL00219RI...</span></span>
<span><span class="co">#&gt; eAL00219RI7  -0.3802431  0.7556116 eAL00219RI...</span></span>
<span><span class="co">#&gt; eAL00219RI8   0.2000259 -0.5325636 eAL00219RI...</span></span>
<span><span class="co">#&gt; eAL00219RI9  -1.0358553  1.0556220 eAL00219RI...</span></span>
<span><span class="co">#&gt; ...                 ...        ...           ...</span></span>
<span><span class="co">#&gt; wAL00286RI12  -0.244162  1.4868339 wAL00286RI...</span></span>
<span><span class="co">#&gt; wAL00286RI14   0.602194  0.0939671 wAL00286RI...</span></span>
<span><span class="co">#&gt; wAL00286RI16  -1.108754  0.3540490 wAL00286RI...</span></span>
<span><span class="co">#&gt; wAL00286RI17   0.955012  0.7391520 wAL00286RI...</span></span>
<span><span class="co">#&gt; wAL00286RI18   2.147393  0.3731584 wAL00286RI...</span></span></code></pre></div>
<p>Let’s explore the component analysis in cell space. Similarly to the
previous explorations, we annotate the results.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc_minimal</span><span class="op">$</span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="va">caResCells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scpAnnotateResults.html">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"cell"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We then generate the component plot. Providing the
<code>pointParams</code> argument, we can shape the points by
<code>SampleType</code>. To assess the impact of batch effects, we also
colour the points according to the MS acquisition run.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">SampleType</span>, colour <span class="op">=</span> <span class="va">Set</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpComponentPlot_cell-1.png" width="480"></p>
<p>While the data before modelling is mainly driven by batch effects,
the APCA clearly separates the two cell populations. The plot can
however only show 2 components at a time. We can explore more components
using a subsequent dimension reduction, such as t-SNE. The
<code>scater</code> package offers a comprehensive set of tools for
dimension reduction on data contained in a
<code>SingleCellExperiment</code> object and requires the components to
be stored in the <code>reducedDim</code> slot. This is streamlined
thanks to <code><a href="../reference/addReducedDims.html">addReducedDims()</a></code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">leduc_minimal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addReducedDims.html">addReducedDims</a></span><span class="op">(</span><span class="va">leduc_minimal</span>, <span class="va">caResCells</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 3</span></span>
<span><span class="co">#&gt; names(3): unmodelled residuals APCA_SampleType</span></span></code></pre></div>
<p>We can now explore the <code>SampleType</code> effects for the 20
computed components through t-SNE.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="va">leduc_minimal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">leduc_minimal</span>, dimred <span class="op">=</span> <span class="st">"APCA_SampleType"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">leduc_minimal</span>, colour_by <span class="op">=</span> <span class="st">"Set"</span>, shape_by <span class="op">=</span> <span class="st">"SampleType"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"t-SNE on 20 APCA components"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/t-SNE-1.png" width="384"></p>
<p>The two cell populations remain clearly separated with an excellent
mixing of the acquisition runs, even when considering the 20 first APCA
components.</p>
<p>We use the same approach to explore the component results in peptide
space.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="va">caRes</span><span class="op">$</span><span class="va">byFeature</span></span>
<span><span class="va">caResPeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scpAnnotateResults.html">scpAnnotateResults</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">leduc_minimal</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"feature"</span>, by2 <span class="op">=</span> <span class="st">"Sequence"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentPlot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResPeps</span>, pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.8</span>, alpha <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/scpComponentPlot_peptide-1.png" width="480"></p>
<p>This exploration may identify groups of covarying peptides, although
no clear patterns appear in the example data set.</p>
<p>We can also combine the exploration of the components in cell and
peptide space. This is possible thanks to biplots.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentBiplot</a></span><span class="op">(</span></span>
<span>    <span class="va">caResCells</span>, <span class="va">caResPeps</span>,</span>
<span>    pointParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">SampleType</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    labelParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, max.overlaps <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>    textBy <span class="op">=</span> <span class="st">"gene"</span>, top <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scp_data_modelling_files/figure-html/biplot-1.png" width="480"></p>
<p>Finally, we offer functionality to aggregate the results at the
protein level instead of the peptide level.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">caResProts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentAggregate</a></span><span class="op">(</span><span class="va">caResPeps</span>, fcol <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Components may no longer be orthogonal after aggregation.</span></span>
<span><span class="va">caResProts</span><span class="op">$</span><span class="va">APCA_SampleType</span></span>
<span><span class="co">#&gt; DataFrame with 86 rows and 26 columns</span></span>
<span><span class="co">#&gt;                PC1        PC2        PC3        PC4         PC5          PC6</span></span>
<span><span class="co">#&gt;          &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ABCE1   -0.0523935 -0.0136777  0.0903348  0.1693285   0.0665345    0.0493461</span></span>
<span><span class="co">#&gt; ACLY    -0.2188357  0.0295186 -0.2475794 -0.0555763  -0.1122432    0.0431707</span></span>
<span><span class="co">#&gt; ACTN4   -0.0119062  0.0384471  0.2774156 -0.0797822  -0.0941029   -0.1243163</span></span>
<span><span class="co">#&gt; AIFM1    0.1006538 -0.0362258  0.1976476  0.0301861  -0.1747488    0.0622619</span></span>
<span><span class="co">#&gt; APMAP    0.0195111  0.0211236  0.0436172  0.0675338   0.0634203    0.1031632</span></span>
<span><span class="co">#&gt; ...            ...        ...        ...        ...         ...          ...</span></span>
<span><span class="co">#&gt; TPP2   -0.05528458 -0.0310194  0.0431430 -0.1027764 -0.08156241 -0.015980101</span></span>
<span><span class="co">#&gt; TUBA1C -0.14532912 -0.0163971  0.0198204  0.0211471 -0.02163806 -0.000503111</span></span>
<span><span class="co">#&gt; VAT1   -0.08475507 -0.0831661  0.1167588  0.0591425 -0.00581566 -0.013318437</span></span>
<span><span class="co">#&gt; XRCC6   0.00292548  0.0614277  0.1692137  0.0295988 -0.02905814 -0.000397610</span></span>
<span><span class="co">#&gt; YWHAG  -0.00908836  0.1134883 -0.2952287  0.0902165 -0.12814301  0.008616979</span></span>
<span><span class="co">#&gt;               PC7        PC8        PC9        PC10       PC11       PC12</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ABCE1   0.0288414  0.0154912 -0.0128954  -0.0168423  0.0485774 -0.0481859</span></span>
<span><span class="co">#&gt; ACLY    0.0716542 -0.1018500 -0.0276287  -0.0739296  0.0719484  0.1163738</span></span>
<span><span class="co">#&gt; ACTN4  -0.0886529  0.0746792 -0.0886262  -0.2615099 -0.0949204  0.2439664</span></span>
<span><span class="co">#&gt; AIFM1   0.0447698 -0.1740009  0.0274838   0.0931568 -0.0445237  0.1537508</span></span>
<span><span class="co">#&gt; APMAP  -0.0839194  0.0436465 -0.0291126  -0.1366302  0.0780494 -0.0467466</span></span>
<span><span class="co">#&gt; ...           ...        ...        ...         ...        ...        ...</span></span>
<span><span class="co">#&gt; TPP2   0.01002283  0.0613254 -0.0657707 -0.02449126 -0.0726705  0.0386026</span></span>
<span><span class="co">#&gt; TUBA1C 0.02632304 -0.0177832  0.0100838 -0.00880753 -0.0311755  0.0127917</span></span>
<span><span class="co">#&gt; VAT1   0.00580065 -0.0525127 -0.0188442 -0.13095970  0.1962062  0.0253535</span></span>
<span><span class="co">#&gt; XRCC6  0.10658824 -0.0676993 -0.0815302  0.23972712  0.1148027  0.0353619</span></span>
<span><span class="co">#&gt; YWHAG  0.01593590 -0.0617372 -0.0961599  0.18878619  0.0708525  0.1197667</span></span>
<span><span class="co">#&gt;                PC13       PC14        PC15       PC16        PC17       PC18</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ABCE1  -0.001663287  0.0343296   0.0343627  0.0990232  0.01865067 -0.0327103</span></span>
<span><span class="co">#&gt; ACLY    0.192738461  0.1474133  -0.1739438  0.1289705  0.05033950  0.0287668</span></span>
<span><span class="co">#&gt; ACTN4  -0.133406589  0.0132553  -0.0693790 -0.0808312 -0.08205540  0.2087751</span></span>
<span><span class="co">#&gt; AIFM1  -0.077020737 -0.0493539  -0.1162488  0.0365235  0.00510608 -0.1441099</span></span>
<span><span class="co">#&gt; APMAP  -0.000121253 -0.1390497  -0.0867093 -0.0410053 -0.11611495  0.1792133</span></span>
<span><span class="co">#&gt; ...             ...        ...         ...        ...         ...        ...</span></span>
<span><span class="co">#&gt; TPP2     0.07597089  0.0558818 -0.00337644 -0.0121018  0.04283655  0.0476116</span></span>
<span><span class="co">#&gt; TUBA1C  -0.02998746 -0.0217296 -0.00789350  0.0565058 -0.00373714 -0.0119916</span></span>
<span><span class="co">#&gt; VAT1    -0.05204097 -0.0852005 -0.02017030  0.0952139 -0.16910390 -0.0715073</span></span>
<span><span class="co">#&gt; XRCC6   -0.00257737  0.0259900  0.10724778 -0.0564281  0.08246124  0.0499968</span></span>
<span><span class="co">#&gt; YWHAG    0.03318137  0.0621676 -0.05404994  0.0792021 -0.12921509 -0.0384467</span></span>
<span><span class="co">#&gt;              PC19       PC20     Reverse Potential.contaminant</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt;  &lt;numeric&gt; &lt;character&gt;           &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCE1  -0.0365880 -0.0681561                                  </span></span>
<span><span class="co">#&gt; ACLY    0.1118103 -0.0208683                                  </span></span>
<span><span class="co">#&gt; ACTN4   0.0596604  0.0966149                                  </span></span>
<span><span class="co">#&gt; AIFM1  -0.2174666  0.1085721                                  </span></span>
<span><span class="co">#&gt; APMAP  -0.0331397  0.1518794                                  </span></span>
<span><span class="co">#&gt; ...           ...        ...         ...                   ...</span></span>
<span><span class="co">#&gt; TPP2   -0.0150711  0.0560492                                  </span></span>
<span><span class="co">#&gt; TUBA1C -0.0269349  0.0296168                                  </span></span>
<span><span class="co">#&gt; VAT1    0.0480006  0.1489793                                  </span></span>
<span><span class="co">#&gt; XRCC6   0.0529253 -0.0198865                                  </span></span>
<span><span class="co">#&gt; YWHAG  -0.1012201  0.1127549                                  </span></span>
<span><span class="co">#&gt;        Leading.razor.protein.id Leading.razor.protein.symbol        gene</span></span>
<span><span class="co">#&gt;                     &lt;character&gt;                  &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; ABCE1                    P61221                        ABCE1       ABCE1</span></span>
<span><span class="co">#&gt; ACLY                     P53396                         ACLY        ACLY</span></span>
<span><span class="co">#&gt; ACTN4                    O43707                        ACTN4       ACTN4</span></span>
<span><span class="co">#&gt; AIFM1                    O95831                        AIFM1       AIFM1</span></span>
<span><span class="co">#&gt; APMAP                    Q9HDC9                        APMAP       APMAP</span></span>
<span><span class="co">#&gt; ...                         ...                          ...         ...</span></span>
<span><span class="co">#&gt; TPP2                     P29144                         TPP2        TPP2</span></span>
<span><span class="co">#&gt; TUBA1C                   Q9BQE3                        TBA1C      TUBA1C</span></span>
<span><span class="co">#&gt; VAT1                     Q99536                         VAT1        VAT1</span></span>
<span><span class="co">#&gt; XRCC6                    P12956                        XRCC6       XRCC6</span></span>
<span><span class="co">#&gt; YWHAG                    P61981                        1433G       YWHAG</span></span>
<span><span class="co">#&gt;               .n</span></span>
<span><span class="co">#&gt;        &lt;integer&gt;</span></span>
<span><span class="co">#&gt; ABCE1          1</span></span>
<span><span class="co">#&gt; ACLY           1</span></span>
<span><span class="co">#&gt; ACTN4          1</span></span>
<span><span class="co">#&gt; AIFM1          1</span></span>
<span><span class="co">#&gt; APMAP          1</span></span>
<span><span class="co">#&gt; ...          ...</span></span>
<span><span class="co">#&gt; TPP2           1</span></span>
<span><span class="co">#&gt; TUBA1C         1</span></span>
<span><span class="co">#&gt; VAT1           2</span></span>
<span><span class="co">#&gt; XRCC6          1</span></span>
<span><span class="co">#&gt; YWHAG          1</span></span></code></pre></div>
<p>Note that the aggregated tables in <code>caResProts</code> can be
explored with the visualisation function
<code><a href="../reference/ScpModel-ComponentAnalysis.html">scpComponentPlot()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h2>
<p>Based on the estimated model, we generate batch-corrected data, that
is data with only the effect of cell type and the residual data. We also
remove the intercept.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">leduc_batchCorrect</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScpModel-DataCorrection.html">scpRemoveBatchEffect</a></span><span class="op">(</span></span>
<span>    <span class="va">leduc_minimal</span>, effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Set"</span>, <span class="st">"Channel"</span>, <span class="st">"MedianIntensity"</span><span class="op">)</span>,</span>
<span>    intercept <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 95 73 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): ''</span></span>
<span><span class="co">#&gt; rownames(95): APNVVVTR GFQEVVTPNIFNSR ... AILGSVER LGAEVYHTLK</span></span>
<span><span class="co">#&gt; rowData names(6): Sequence Reverse ... Leading.razor.protein.symbol</span></span>
<span><span class="co">#&gt;   gene</span></span>
<span><span class="co">#&gt; colnames(73): eAL00219RI5 eAL00219RI6 ... wAL00286RI17 wAL00286RI18</span></span>
<span><span class="co">#&gt; colData names(13): Set Channel ... passQC cell</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>Note that the batch-corrected data still contain missing values. The
<code>leduc_batchCorrect</code> object can be used for downstream
analysis.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code>R Under development (unstable) (2025-02-15 r87726)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 24.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] scater_1.35.1               scuttle_1.17.0             
 [3] ggplot2_3.5.1               patchwork_1.3.0            
 [5] SingleCellExperiment_1.29.1 scp_1.17.1                 
 [7] QFeatures_1.17.4            MultiAssayExperiment_1.33.9
 [9] SummarizedExperiment_1.37.0 Biobase_2.67.0             
[11] GenomicRanges_1.59.1        GenomeInfoDb_1.43.4        
[13] IRanges_2.41.3              S4Vectors_0.45.4           
[15] BiocGenerics_0.53.6         generics_0.1.3             
[17] MatrixGenerics_1.19.1       matrixStats_1.5.0          
[19] BiocStyle_2.35.0           

loaded via a namespace (and not attached):
 [1] gridExtra_2.3           fdrtool_1.2.18          rlang_1.1.5            
 [4] magrittr_2.0.3          clue_0.3-66             compiler_4.5.0         
 [7] systemfonts_1.2.1       vctrs_0.6.5             reshape2_1.4.4         
[10] nipals_1.0              stringr_1.5.1           ProtGenerics_1.39.2    
[13] pkgconfig_2.0.3         crayon_1.5.3            fastmap_1.2.0          
[16] XVector_0.47.2          labeling_0.4.3          rmarkdown_2.29         
[19] ggbeeswarm_0.7.2        UCSC.utils_1.3.1        ragg_1.3.3             
[22] purrr_1.0.4             xfun_0.51               cachem_1.1.0           
[25] beachmat_2.23.6         jsonlite_1.9.0          DelayedArray_0.33.6    
[28] BiocParallel_1.41.2     irlba_2.3.5.1           parallel_4.5.0         
[31] cluster_2.1.8           R6_2.6.1                bslib_0.9.0            
[34] stringi_1.8.4           RColorBrewer_1.1-3      jquerylib_0.1.4        
[37] Rcpp_1.0.14             bookdown_0.42           knitr_1.49             
[40] BiocBaseUtils_1.9.0     Matrix_1.7-2            igraph_2.1.4           
[43] tidyselect_1.2.1        viridis_0.6.5           abind_1.4-8            
[46] yaml_2.3.10             codetools_0.2-20        lattice_0.22-6         
[49] tibble_3.2.1            plyr_1.8.9              withr_3.0.2            
[52] Rtsne_0.17              evaluate_1.0.3          desc_1.4.3             
[55] IHW_1.35.0              pillar_1.10.1           BiocManager_1.30.25    
[58] lpsymphony_1.35.0       munsell_0.5.1           scales_1.3.0           
[61] glue_1.8.0              slam_0.1-55             metapod_1.15.0         
[64] lazyeval_0.2.2          tools_4.5.0             BiocNeighbors_2.1.2    
[67] ScaledMatrix_1.15.0     fs_1.6.5                grid_4.5.0             
[70] tidyr_1.3.1             MsCoreUtils_1.19.0      colorspace_2.1-1       
[73] GenomeInfoDbData_1.2.13 beeswarm_0.4.0          BiocSingular_1.23.0    
[76] vipor_0.4.7             rsvd_1.0.5              cli_3.6.4              
[79] textshaping_1.0.0       viridisLite_0.4.2       S4Arrays_1.7.3         
[82] dplyr_1.1.4             AnnotationFilter_1.31.0 gtable_0.3.6           
[85] sass_0.4.9              digest_0.6.37           SparseArray_1.7.6      
[88] ggrepel_0.9.6           htmlwidgets_1.6.4       farver_2.1.2           
[91] htmltools_0.5.8.1       pkgdown_2.1.1.9000      lifecycle_1.0.4        
[94] httr_1.4.7              MASS_7.3-64            </code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>This vignette is distributed under a <a href="https://creativecommons.org/licenses/by-sa/2.0/" class="external-link">CC BY-SA
license</a> license.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Vanderaa2023-scplainer" class="csl-entry">
Vanderaa, Christophe, and Laurent Gatto. 2024. <span>“<span class="nocase">scplainer</span>: Using Linear Models to Understand Mass
Spectrometry-Based Single-Cell Proteomics Data.”</span>
<em>bioRxiv</em>. <a href="https://doi.org/10.1101/2023.12.14.571792" class="external-link">https://doi.org/10.1101/2023.12.14.571792</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christophe Vanderaa, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
